<!doctype html><html lang=en dir=auto data-theme=auto><head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width,initial-scale=1,shrink-to-fit=no"><meta name=robots content="index, follow"><title>é‡æµ‹åºæ•™ç¨‹ | æˆ‘çš„åšå®¢</title><meta name=keywords content><meta name=description content="é‡æµ‹åºä¸ç¾¤ä½“é—ä¼ å­¦
äº‘å—å†œä¸šå¤§å­¦ äº‘å—ç”Ÿç‰©èµ„æºä¿æŠ¤ä¸åˆ©ç”¨å›½å®¶é‡ç‚¹å®éªŒå®¤
ï¼ˆæè¯¦ 2019å¹´12æœˆ31æ—¥ï¼‰
è½¯ä»¶å®‰è£…
è¯¥éƒ¨åˆ†è½¯ä»¶å¾ˆéš¾ç”¨condaç›´æ¥å®‰è£…ï¼Œå®‰è£…æ­¥éª¤æ¯”è¾ƒç‰¹æ®Šã€‚
lumpy-sv
1
2
3
4
5
6
7
lumpy-sv ä¾èµ– Python2.7ï¼Œåˆ‡æ¢åˆ°py27ç¯å¢ƒä¸‹è¿›è¡Œå®‰è£…ã€‚
conda activate py27
git clone &ndash;recursive https://github.com/arq5x/lumpy-sv.git
cd lumpy-sv
make
å®‰è£…svtyper
conda install svtyper
cnvnator
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
cnvnatorä¾èµ–äºrootè½¯ä»¶åŒ…åŠsamtoolsè½¯ä»¶åŒ…(åŒ…å«HTSlib)ï¼Œå› æ­¤å…ˆå®‰è£…SamtoolsåŠrootè½¯ä»¶åŒ…ã€‚
å®‰è£…samtools
wget https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2
tar -xvf samtools-1.9.tar.bz2
cd samtools-1.9/
./configure
make
å®‰è£…rootè½¯ä»¶, ç›´æ¥ä¸‹è½½åè§£å‹å³å¯
wget https://root.cern/download/root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz
tar -zxvf root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz"><meta name=author content><link rel=canonical href=https://Tdxxxx-student.github.io/posts/%E9%87%8D%E6%B5%8B%E5%BA%8F%E6%95%99%E7%A8%8B/><link crossorigin=anonymous href=/assets/css/stylesheet.da3211e5ef867bf2b75fd5a6515cfed7195c011e8ab735694e203810a827097b.css integrity="sha256-2jIR5e+Ge/K3X9WmUVz+1xlcAR6KtzVpTiA4EKgnCXs=" rel="preload stylesheet" as=style><link rel=icon href=https://Tdxxxx-student.github.io/favicon.ico><link rel=icon type=image/png sizes=16x16 href=https://Tdxxxx-student.github.io/favicon-16x16.png><link rel=icon type=image/png sizes=32x32 href=https://Tdxxxx-student.github.io/favicon-32x32.png><link rel=apple-touch-icon href=https://Tdxxxx-student.github.io/apple-touch-icon.png><link rel=mask-icon href=https://Tdxxxx-student.github.io/safari-pinned-tab.svg><meta name=theme-color content="#2e2e33"><meta name=msapplication-TileColor content="#2e2e33"><link rel=alternate hreflang=en href=https://Tdxxxx-student.github.io/posts/%E9%87%8D%E6%B5%8B%E5%BA%8F%E6%95%99%E7%A8%8B/><noscript><style>#theme-toggle,.top-link{display:none}</style><style>@media(prefers-color-scheme:dark){:root{--theme:rgb(29, 30, 32);--entry:rgb(46, 46, 51);--primary:rgb(218, 218, 219);--secondary:rgb(155, 156, 157);--tertiary:rgb(65, 66, 68);--content:rgb(196, 196, 197);--code-block-bg:rgb(46, 46, 51);--code-bg:rgb(55, 56, 62);--border:rgb(51, 51, 51);color-scheme:dark}.list{background:var(--theme)}.toc{background:var(--entry)}}@media(prefers-color-scheme:light){.list::-webkit-scrollbar-thumb{border-color:var(--code-bg)}}</style></noscript><script>localStorage.getItem("pref-theme")==="dark"?document.querySelector("html").dataset.theme="dark":localStorage.getItem("pref-theme")==="light"?document.querySelector("html").dataset.theme="light":window.matchMedia("(prefers-color-scheme: dark)").matches?document.querySelector("html").dataset.theme="dark":document.querySelector("html").dataset.theme="light"</script><meta property="og:url" content="https://Tdxxxx-student.github.io/posts/%E9%87%8D%E6%B5%8B%E5%BA%8F%E6%95%99%E7%A8%8B/"><meta property="og:site_name" content="æˆ‘çš„åšå®¢"><meta property="og:title" content="é‡æµ‹åºæ•™ç¨‹"><meta property="og:description" content="é‡æµ‹åºä¸ç¾¤ä½“é—ä¼ å­¦ äº‘å—å†œä¸šå¤§å­¦ äº‘å—ç”Ÿç‰©èµ„æºä¿æŠ¤ä¸åˆ©ç”¨å›½å®¶é‡ç‚¹å®éªŒå®¤ ï¼ˆæè¯¦ 2019å¹´12æœˆ31æ—¥ï¼‰ è½¯ä»¶å®‰è£… è¯¥éƒ¨åˆ†è½¯ä»¶å¾ˆéš¾ç”¨condaç›´æ¥å®‰è£…ï¼Œå®‰è£…æ­¥éª¤æ¯”è¾ƒç‰¹æ®Šã€‚
lumpy-sv 1 2 3 4 5 6 7 lumpy-sv ä¾èµ– Python2.7ï¼Œåˆ‡æ¢åˆ°py27ç¯å¢ƒä¸‹è¿›è¡Œå®‰è£…ã€‚ conda activate py27 git clone â€“recursive https://github.com/arq5x/lumpy-sv.git cd lumpy-sv make
å®‰è£…svtyper conda install svtyper cnvnator 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 cnvnatorä¾èµ–äºrootè½¯ä»¶åŒ…åŠsamtoolsè½¯ä»¶åŒ…(åŒ…å«HTSlib)ï¼Œå› æ­¤å…ˆå®‰è£…SamtoolsåŠrootè½¯ä»¶åŒ…ã€‚
å®‰è£…samtools wget https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2 tar -xvf samtools-1.9.tar.bz2 cd samtools-1.9/ ./configure make
å®‰è£…rootè½¯ä»¶, ç›´æ¥ä¸‹è½½åè§£å‹å³å¯ wget https://root.cern/download/root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz tar -zxvf root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz"><meta property="og:locale" content="zh-cn"><meta property="og:type" content="article"><meta property="article:section" content="posts"><meta property="article:published_time" content="2026-02-03T12:08:55+08:00"><meta property="article:modified_time" content="2026-02-03T12:08:55+08:00"><meta name=twitter:card content="summary"><meta name=twitter:title content="é‡æµ‹åºæ•™ç¨‹"><meta name=twitter:description content="é‡æµ‹åºä¸ç¾¤ä½“é—ä¼ å­¦
äº‘å—å†œä¸šå¤§å­¦ äº‘å—ç”Ÿç‰©èµ„æºä¿æŠ¤ä¸åˆ©ç”¨å›½å®¶é‡ç‚¹å®éªŒå®¤
ï¼ˆæè¯¦ 2019å¹´12æœˆ31æ—¥ï¼‰
è½¯ä»¶å®‰è£…
è¯¥éƒ¨åˆ†è½¯ä»¶å¾ˆéš¾ç”¨condaç›´æ¥å®‰è£…ï¼Œå®‰è£…æ­¥éª¤æ¯”è¾ƒç‰¹æ®Šã€‚
lumpy-sv
1
2
3
4
5
6
7
lumpy-sv ä¾èµ– Python2.7ï¼Œåˆ‡æ¢åˆ°py27ç¯å¢ƒä¸‹è¿›è¡Œå®‰è£…ã€‚
conda activate py27
git clone &ndash;recursive https://github.com/arq5x/lumpy-sv.git
cd lumpy-sv
make
å®‰è£…svtyper
conda install svtyper
cnvnator
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
cnvnatorä¾èµ–äºrootè½¯ä»¶åŒ…åŠsamtoolsè½¯ä»¶åŒ…(åŒ…å«HTSlib)ï¼Œå› æ­¤å…ˆå®‰è£…SamtoolsåŠrootè½¯ä»¶åŒ…ã€‚
å®‰è£…samtools
wget https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2
tar -xvf samtools-1.9.tar.bz2
cd samtools-1.9/
./configure
make
å®‰è£…rootè½¯ä»¶, ç›´æ¥ä¸‹è½½åè§£å‹å³å¯
wget https://root.cern/download/root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz
tar -zxvf root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz"><script type=application/ld+json>{"@context":"https://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Posts","item":"https://Tdxxxx-student.github.io/posts/"},{"@type":"ListItem","position":2,"name":"é‡æµ‹åºæ•™ç¨‹","item":"https://Tdxxxx-student.github.io/posts/%E9%87%8D%E6%B5%8B%E5%BA%8F%E6%95%99%E7%A8%8B/"}]}</script><script type=application/ld+json>{"@context":"https://schema.org","@type":"BlogPosting","headline":"é‡æµ‹åºæ•™ç¨‹","name":"é‡æµ‹åºæ•™ç¨‹","description":"é‡æµ‹åºä¸ç¾¤ä½“é—ä¼ å­¦ äº‘å—å†œä¸šå¤§å­¦ äº‘å—ç”Ÿç‰©èµ„æºä¿æŠ¤ä¸åˆ©ç”¨å›½å®¶é‡ç‚¹å®éªŒå®¤ ï¼ˆæè¯¦ 2019å¹´12æœˆ31æ—¥ï¼‰ è½¯ä»¶å®‰è£… è¯¥éƒ¨åˆ†è½¯ä»¶å¾ˆéš¾ç”¨condaç›´æ¥å®‰è£…ï¼Œå®‰è£…æ­¥éª¤æ¯”è¾ƒç‰¹æ®Šã€‚\nlumpy-sv 1 2 3 4 5 6 7 lumpy-sv ä¾èµ– Python2.7ï¼Œåˆ‡æ¢åˆ°py27ç¯å¢ƒä¸‹è¿›è¡Œå®‰è£…ã€‚ conda activate py27 git clone \u0026ndash;recursive https://github.com/arq5x/lumpy-sv.git cd lumpy-sv make\nå®‰è£…svtyper conda install svtyper cnvnator 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 cnvnatorä¾èµ–äºrootè½¯ä»¶åŒ…åŠsamtoolsè½¯ä»¶åŒ…(åŒ…å«HTSlib)ï¼Œå› æ­¤å…ˆå®‰è£…SamtoolsåŠrootè½¯ä»¶åŒ…ã€‚\nå®‰è£…samtools wget https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2 tar -xvf samtools-1.9.tar.bz2 cd samtools-1.9/ ./configure make\nå®‰è£…rootè½¯ä»¶, ç›´æ¥ä¸‹è½½åè§£å‹å³å¯ wget https://root.cern/download/root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz tar -zxvf root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz\n","keywords":[],"articleBody":"é‡æµ‹åºä¸ç¾¤ä½“é—ä¼ å­¦ äº‘å—å†œä¸šå¤§å­¦ äº‘å—ç”Ÿç‰©èµ„æºä¿æŠ¤ä¸åˆ©ç”¨å›½å®¶é‡ç‚¹å®éªŒå®¤ ï¼ˆæè¯¦ 2019å¹´12æœˆ31æ—¥ï¼‰ è½¯ä»¶å®‰è£… è¯¥éƒ¨åˆ†è½¯ä»¶å¾ˆéš¾ç”¨condaç›´æ¥å®‰è£…ï¼Œå®‰è£…æ­¥éª¤æ¯”è¾ƒç‰¹æ®Šã€‚\nlumpy-sv 1 2 3 4 5 6 7 lumpy-sv ä¾èµ– Python2.7ï¼Œåˆ‡æ¢åˆ°py27ç¯å¢ƒä¸‹è¿›è¡Œå®‰è£…ã€‚ conda activate py27 git clone â€“recursive https://github.com/arq5x/lumpy-sv.git cd lumpy-sv make\nå®‰è£…svtyper conda install svtyper cnvnator 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 cnvnatorä¾èµ–äºrootè½¯ä»¶åŒ…åŠsamtoolsè½¯ä»¶åŒ…(åŒ…å«HTSlib)ï¼Œå› æ­¤å…ˆå®‰è£…SamtoolsåŠrootè½¯ä»¶åŒ…ã€‚\nå®‰è£…samtools wget https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2 tar -xvf samtools-1.9.tar.bz2 cd samtools-1.9/ ./configure make\nå®‰è£…rootè½¯ä»¶, ç›´æ¥ä¸‹è½½åè§£å‹å³å¯ wget https://root.cern/download/root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz tar -zxvf root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz\nå®‰è£…cnvnator git clone https://github.com/abyzovlab/CNVnator.git cd CNVnator\né“¾æ¥samtoolsè½¯ä»¶ç›®å½•åˆ°cnvnatorç›®å½• ln -s ../samtools-1.9 ./samtools\nåŠ è½½rootè½¯ä»¶åˆ°ç¯å¢ƒå˜é‡ export ROOTSYS=/pub/software/root export PATH=/pub/software/root/bin:$PATH export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ROOTSYS/lib\nå®‰è£…cnvnator make LIBS=\"-lcrypto\"\nless ../data/sample.list | awk â€˜{ print â€œbwa mem -t 2 -R â€˜'â€™@RG\\tID:\"$1â€\\tSM:\"$1\"\\tPL:illuminaâ€™'â€™ ../01.ref/genome.fasta ../data/\"$1\"_1.fq.gz ../data/\"$1\"_2.fq.gz | /pub/software/samtools/samtools sort -@ 2 -m 1G -o â€œ$1â€.sort.bam -\" }â€™ å˜å¼‚æ£€æµ‹ æ„å»ºåŸºå› ç»„index 1 2 3 4 5 6 7 8 9 10 11\ncreate link ln -s ../data/genome.fasta ./genome.fasta\nfor gatk samtools faidx genome.fasta\nfor bwa bwa index genome.fasta\nfor picard picard CreateSequenceDictionary R=genome.fasta æ¯”å¯¹æ’åºå»é‡ 1 2 3 4 5 6 7 8 9 10 11\næ¯”å¯¹æ’åº bwa mem -t 2 -R â€˜@RG\\tID:S1\\tSM:S1\\tPL:illuminaâ€™ ../01.ref/genome.fasta ../data/S1_1.fq.gz ../data/S1_2.fq.gz | /pub/software/samtools/samtools sort -@ 2 -m 1G -o S1.sort.bam - bwa mem -t 2 -R â€˜@RG\\tID:S2\\tSM:S2\\tPL:illuminaâ€™ ../01.ref/genome.fasta ../data/S2_1.fq.gz ../data/S2_2.fq.gz | /pub/software/samtools/samtools sort -@ 2 -m 1G -o S2.sort.bam -\nå»é™¤é‡å¤ picard -Xmx4g MarkDuplicates I=S1.sort.bam O=S1.sort.rmdup.bam CREATE_INDEX=true REMOVE_DUPLICATES=true M=S1.marked_dup_metrics.txt picard -Xmx4g MarkDuplicates I=S2.sort.bam O=S2.sort.rmdup.bam CREATE_INDEX=true REMOVE_DUPLICATES=true M=S2.marked_dup_metrics.txt\næ¯”å¯¹ç‡ç»Ÿè®¡ /pub/software/samtools/samtools flagstat S1.sort.bam \u003e S1.sort.bam.flagstat /pub/software/samtools/samtools flagstat S2.sort.bam \u003e S2.sort.bam.flagstat SNPã€Indelæ£€æµ‹ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24\nstep1.HaplotypeCaller gatk â€“java-options â€œ-Xmx10g -Djava.io.tmpdir=./tmpâ€ HaplotypeCaller -R ../01.ref/genome.fasta -I ../02.mapping/S1.sort.rmdup.bam -ERC GVCF -O S1.g.vcf 1\u003eS1.HC.log 2\u003e\u00261 gatk â€“java-options â€œ-Xmx10g -Djava.io.tmpdir=./tmpâ€ HaplotypeCaller -R ../01.ref/genome.fasta -I ../02.mapping/S2.sort.rmdup.bam -ERC GVCF -O S2.g.vcf 1\u003eS2.HC.log 2\u003e\u00261\nstep2.CombineGVCFs ls ./S*.g.vcf \u003e gvcf.list gatk â€“java-options â€œ-Xmx4g -Djava.io.tmpdir=./tmpâ€ CombineGVCFs -R ../01.ref/genome.fasta -V gvcf.list -O all.merge.g.vcf\nstep3.GenotypeGVCFs gatk â€“java-options â€œ-Xmx4g -Djava.io.tmpdir=./tmpâ€ GenotypeGVCFs -R ../01.ref/genome.fasta â€“variant all.merge.g.vcf -O all.merge_raw.vcf\nstep4.SNP gatk â€“java-options â€œ-Xmx4g -Djava.io.tmpdir=./tmpâ€ SelectVariants -R ../01.ref/genome.fasta -V all.merge_raw.vcf â€“select-type SNP -O all.raw.snp.vcf\ngatk â€“java-options â€œ-Xmx4g -Djava.io.tmpdir=./tmpâ€ VariantFiltration -R ../01.ref/genome.fasta -V all.raw.snp.vcf â€“filter-expression â€œQD \u003c 2.0 || MQ \u003c 40.0 || FS \u003e 60.0 || SOR \u003e 3.0 || MQRankSum \u003c -12.5 || ReadPosRankSum \u003c -8.0â€ â€“filter-name â€˜SNP_filterâ€™ -O all.filter.snp.vcf\ngatk â€“java-options â€œ-Xmx4g -Djava.io.tmpdir=./tmpâ€ SelectVariants -R ../01.ref/genome.fasta -V all.filter.snp.vcf â€“exclude-filtered -O all.filtered.snp.vcf\nstep5.InDel gatk â€“java-options â€œ-Xmx4g -Djava.io.tmpdir=./tmpâ€ SelectVariants -R ../01.ref/genome.fasta -V all.merge_raw.vcf â€“select-type INDEL -O all.raw.indel.vcf\ngatk â€“java-options â€œ-Xmx4g -Djava.io.tmpdir=./tmpâ€ VariantFiltration -R ../01.ref/genome.fasta -V all.raw.indel.vcf â€“filter-expression â€œQD \u003c 2.0 || FS \u003e 200.0 || SOR \u003e 10.0 || MQRankSum \u003c -12.5 || ReadPosRankSum \u003c -8.0â€ â€“filter-name â€˜INDEL_filterâ€™ -O all.filter.indel.vcf\ngatk â€“java-options â€œ-Xmx4g -Djava.io.tmpdir=./tmpâ€ SelectVariants -R ../01.ref/genome.fasta -V all.filter.indel.vcf â€“exclude-filtered -O all.filtered.indel.vcf SVæ£€æµ‹ 1 2 3 4 5 6 7 8 9 10 11 12 13 14\nstep1.discordants_bam /pub/software/samtools/samtools view -b -F 1294 -@ 12 ../02.mapping/S1.sort.rmdup.bam \u003e S1.discordants.bam /pub/software/samtools/samtools view -b -F 1294 -@ 12 ../02.mapping/S2.sort.rmdup.bam \u003e S2.discordants.bam\nstep2.splitters_bam /pub/software/samtools/samtools view -h ../02.mapping/S1.sort.rmdup.bam | /pub/software/lumpy-sv/scripts/extractSplitReads_BwaMem -i stdin | /pub/software/samtools/samtools sort - \u003e S1.splitters.bam /pub/software/samtools/samtools view -h ../02.mapping/S2.sort.rmdup.bam | /pub/software/lumpy-sv/scripts/extractSplitReads_BwaMem -i stdin | /pub/software/samtools/samtools sort - \u003e S2.splitters.bam\nstep3.lumpy /pub/software/lumpy-sv/bin/lumpyexpress -B ../02.mapping/S1.sort.rmdup.bam,../02.mapping/S2.sort.rmdup.bam -S S1.splitters.bam,S2.splitters.bam -D S1.discordants.bam,S2.discordants.bam -o all.sv.lumpy.vcf\nstep4.ind_genotype vcftools â€“vcf all.sv.lumpy.vcf â€“indv S1 â€“recode â€“recode-INFO-all â€“out S1 \u0026\u0026 svtyper -i S1.recode.vcf -B ../02.mapping/S1.sort.rmdup.bam -o S1.genotype.vcf vcftools â€“vcf all.sv.lumpy.vcf â€“indv S2 â€“recode â€“recode-INFO-all â€“out S2 \u0026\u0026 svtyper -i S2.recode.vcf -B ../02.mapping/S2.sort.rmdup.bam -o S2.genotype.vcf CNVæ£€æµ‹ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37\nstep1.prepare_genome split genome into single sequence ln -s ../data/genome.fasta ./genome.fa\nseqkit split -i ./genome.fa -O split\nls ./split/genome.id_*.fa |sed â€™s/./split/genome.id_//â€™ |awk â€˜{print â€œmv ./split/genome.id_\"$aaâ€ ./split/\"$aa}â€™ \u003e mv_name.sh\nsh mv_name.sh\nstep2.cnvnator export ROOTSYS=/pub/software/root export PATH=/pub/software/root/bin:$PATH export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ROOTSYS/lib\né“¾æ¥æ–‡ä»¶åˆ°å½“å‰ç›®å½• ln -s ../02.mapping/S1.sort.rmdup.bam\næå–æ¯”å¯¹ç»“æœ /pub/software/CNVnator/cnvnator -genome genome.fa -root S1.root -tree S1.sort.rmdup.bam\n##ç”Ÿæˆæ·±åº¦åˆ†å¸ƒ /pub/software/CNVnator/cnvnator -genome genome.fa -root S1.root -his 500 -d split\nè¿›è¡Œç»Ÿè®¡è®¡ç®— /pub/software/CNVnator/cnvnator -root S1.root -stat 500\næ£€æŸ¥bin sizeæ˜¯å¦åˆé€‚ /pub/software/CNVnator/cnvnator -root S1.root -eval 500 \u003e S1.eval.ratio\nRDä¿¡å·åˆ†å‰² /pub/software/CNVnator/cnvnator -root S1.root -partition 500\nè¿›è¡ŒCNVæ£€æµ‹ /pub/software/CNVnator/cnvnator -root S1.root -call 500 \u003e S1.cnv SNP INDELæ³¨é‡Š 1 2 3 4 5 6 7 8 9 10 11 12 13\nstep1.prepare_genome ln -s ../data/genome.gtf ln -s ../data/genome.fasta\ngtfToGenePred -genePredExt genome.gtf genome_refGene.txt /pub/software/annovar/retrieve_seq_from_fasta.pl â€“format refGene â€“seqfile genome.fasta genome_refGene.txt â€“out genome_refGeneMrna.fa\n#step2.ann_vcf ln -s ../03.SNP_indel/all.filtered.snp.vcf\n/pub/software/annovar/convert2annovar.pl -format vcf4old all.filtered.snp.vcf \u003eall.snp.vcf.annovar.input\n/pub/software/annovar/annotate_variation.pl -geneanno â€“neargene 2000 -buildver genome -dbtype refGene -outfile all.anno -exonsort all.snp.vcf.annovar.input ./ CNVå’ŒSVæ³¨é‡Š 1 2 3 4 5 6 7 less -S ../04.SV/all.sv.lumpy.vcf|awk -F â€œ;â€ â€˜{ if($1!/#/){print $1\"\\t\"$2\"\\t\"$3\"\\t\"$4} }â€™|awk â€˜{print $1\"\\t\"$2\"\\t\"$11\"\\t\"$3\"\\t\"$8}â€™|sed â€™s/END=//â€™ |sed â€™s/SVTYPE=//â€™| awk â€˜$3\u003c$2â€™ \u003e SV.bed less -S ../04.SV/all.sv.lumpy.vcf|awk -F â€œ;â€ â€˜{ if($1!/#/){print $1\"\\t\"$2\"\\t\"$3\"\\t\"$4} }â€™|awk â€˜{print $1\"\\t\"$2\"\\t\"$11\"\\t\"$3\"\\t\"$8}â€™|sed â€™s/END=//â€™ |sed â€™s/SVTYPE=//â€™ \u003e SV.bed awk â€˜$5 !~ /BND/â€™ SV.bed \u003e SV_filter.bed bedtools intersect -wo -a SV_filter.bed -b gene.gtf \u003e SV_filter.Anno less -S ../05.CNV/S1.cnv | awk â€˜{print $2 â€œ\\tâ€ $1}â€™|sed â€™s/:/\\t/â€™|sed â€™s/-/\\t/â€™ \u003e CNV.bed bedtools intersect -wo -a CNV.bed -b gene.gtf \u003e CNV.Anno\nç¾¤ä½“ç»“æ„ è¿‡æ»¤ 1 2 3 4 5 6 7 8 9 vcf=../data/all.vcf\nfilter missing and maf plink â€“vcf $vcf â€“geno 0.1 â€“maf 0.01 â€“out all.missing_maf â€“recode vcf-iid â€“allow-extra-chr â€“set-missing-var-ids @:# â€“keep-allele-order\nfilter LD plink â€“vcf all.missing_maf.vcf â€“indep-pairwise 50 10 0.2 â€“out tmp.ld â€“allow-extra-chr â€“set-missing-var-ids @:# plink â€“vcf all.missing_maf.vcf â€“make-bed â€“extract tmp.ld.prune.in â€“out all.LDfilter â€“recode vcf-iid â€“keep-allele-order â€“allow-extra-chr â€“set-missing-var-ids @:#\nå»ºæ ‘ NJ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 perl ../../script/vcf2phy.pl ../../00.filter/all.LDfilter.vcf \u003e sequences.dat\necho -e â€œsequences.dat\\nYâ€ \u003e dnadist.cfg\ndnadist \u003c dnadist.cfg \u003ednadist.log\nmv outfile infile.dist\necho -e â€œinfile.dist\\nYâ€ \u003e neighbor.cfg\nneighbor \u003c neighbor.cfg \u003enj.log\nless infile.dist | tr â€˜\\nâ€™ â€˜|â€™| sed â€™s/| / /gâ€™ | tr â€˜|â€™ â€˜\\nâ€™ \u003einfile.dist.table less outtree | tr â€˜\\nâ€™ â€™ â€˜|sed â€™s/ //gâ€™ \u003e outtree.nwk\ntree_snphylo 1 2 /pub/software/SNPhylo/snphylo.sh -v ../../00.filter/all.LDfilter.vcf -r -l 1 -m 0.01 -o GA0001 -b -B 2 -P tree_snphylo\nPCA 1 2 3 4 plink â€“vcf ../00.filter/all.LDfilter.vcf â€“pca 10 â€“out PCA_out â€“allow-extra-chr â€“set-missing-var-ids @:#\nRscript ../script/draw_PCA.R PCA_out.eigenvec 1 2 ../data/sample.pop PCA_out_figure\nç¾¤ä½“ç»“æ„ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\nstep0.vcf2bed plink â€“vcf ../00.filter/all.LDfilter.vcf â€“make-bed â€“out all â€“allow-extra-chr â€“keep-allele-order â€“set-missing-var-ids @:#\nstep1.get_Admix_shell seq 2 4 | awk â€˜{print â€œadmixture â€“cv -j2 all.bed â€œ$1â€ 1\u003eadmix.\"$1â€.log 2\u003e\u00261\"}â€™ \u003e admixture.sh\nstep2.run_admix sh admixture.sh\nstep3.draw_structure mkdir result cp ./*.Q result/ Rscript ../script/draw_admixture.R result all.nosex structure\nadmixture admixture â€“cv -j2 all.bed 2 1\u003eadmix.2.log 2\u003e\u00261 admixture â€“cv -j2 all.bed 3 1\u003eadmix.3.log 2\u003e\u00261 admixture â€“cv -j2 all.bed 4 1\u003eadmix.4.log 2\u003e\u00261\né—ä¼ è¡°å‡ 1 2 3 4 5 6 7 8 9 10\nstep1.LD_calcu /pub/software/PopLDdecay/bin/PopLDdecay -InVCF ../data/all.vcf -SubPop ../data/pop.SC.table -MaxDist 500 -OutStat pop.SC.stat /pub/software/PopLDdecay/bin/PopLDdecay -InVCF ../data/all.vcf -SubPop ../data/pop.YZR.table -MaxDist 500 -OutStat pop.YZR.stat\nstep2.draw_singlePlot perl /pub/software/PopLDdecay/bin/Plot_OnePop.pl -inFile pop.SC.stat.gz -output pop.SC.ld perl /pub/software/PopLDdecay/bin/Plot_OnePop.pl -inFile pop.YZR.stat.gz -output pop.YZR.ld\nstep3.draw_multiPlotare/PopLDdecay/bin/Plot_MultiPop.pl -inList ld_stat.list -output ld_stat.multi é€‰æ‹©åˆ†æ i_ROD 1 2 3 4 5 6 7 8 9 10 11 12\nstep1.pi vcf=../../data/all.vcf window=20000 step=2000 vcftools â€“vcf $vcf â€“window-pi $window â€“window-pi-step $step â€“keep ../../data/pop.SC.table â€“out ./Pi.pop1 vcftools â€“vcf $vcf â€“window-pi $window â€“window-pi-step $step â€“keep ../../data/pop.YZR.table â€“out ./Pi.pop2\nstep2.ROD pi1=Pi.pop1.windowed.pi pi2=Pi.pop2.windowed.pi perl ../../script/merge2pi_ROD.pl $pi1 $pi2 \u003ePi.ROD\ntajimaD 1 2 3 4 5 6 vcf=../../data/all.vcf window=20000\nvcftools â€“vcf $vcf â€“TajimaD $window â€“keep ../../data/pop.SC.table â€“out TajimaD.pop1 vcftools â€“vcf $vcf â€“TajimaD $window â€“keep ../../data/pop.YZR.table â€“out TajimaD.pop2\nFst 1 2 3 4 5 vcf=../../data/all.vcf window=20000 step=2000 vcftools â€“vcf $vcf â€“fst-window-size $window â€“fst-window-step $step â€“weir-fst-pop ../../data/pop.SC.table â€“weir-fst-pop ../../data/pop.YZR.table â€“out ./Fst.pop1.pop2\nROD_Fst 1 2 3 4 5 6 7 8 9 10 11 12 13 fst=../03.Fst/Fst.pop1.pop2.windowed.weir.fst ROD=../01.pi_ROD/Pi.ROD gff=../../data/genome.gff\nperl ../../script/top_Fst_ROD_S.pl $fst $ROD 0.05 0.05 Fst_ROD.table Fst_ROD.stat\nawk â€˜$NF==â€œtopâ€â€™ Fst_ROD.table \u003e Fst_ROD.table.top\nawk â€˜$3==â€œgeneâ€â€™ $gff \u003e gene.gff\nbedtools intersect -wo -F 0.1 -a Fst_ROD.table.top -b gene.gff | awk -F â€œ\\tâ€ â€˜{print $7\"\\t\"$10\"\\t\"$11\"\\t\"$13\"\\t\"$15}â€™ | sort -u \u003e Fst_ROD.table.top.gene\nGWAS mpute 1 2 beagle -Xmx4g -Djava.io.tmpdir=./TMP gt=../data/all.vcf out=all.impute impute=true window=10 nthreads=2\nGWAS 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\nstep1.prepare prepare vcf plink â€“vcf ../data/all.vcf â€“maf 0.05 â€“geno 0.1 â€“recode12 â€“output-missing-genotype 0 â€“transpose â€“out snp_filter â€“set-missing-var-ids @:# â€“allow-extra-chr\nprepare phenotype perl ../script/sort_pheno.pl snp_filter.tfam ../data/trait.table \u003e trait.sort.txt\nstep2.kinship /pub/software/emmax/emmax-kin-intel64 -v -d 10 -o ./pop.kinship snp_filter\nstep3.gwas /pub/software/emmax/emmax-intel64 -v -d 10 -t snp_filter -p trait.sort.txt -k pop.kinship -o emmax.out 1\u003e emmax.log 2\u003eemmax.err\ndraw_manhattan paste snp_filter.map emmax.out.ps | awk â€˜BEGIN{print â€œSNP\\tCHR\\tBP\\tPâ€}{if($2==$5){print $2\"\\t\"$1\"\\t\"$4\"\\t\"$NF}}â€™ \u003e emmax.out.ps.manht_input Rscript ../script/manhattan.R emmax.out.ps.manht_input emmax.out.ps.manht_figure\nGWAS-Q 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17\nstep3.gwas ln -s ../01.GWAS/snp_filter.tped ln -s ../01.GWAS/snp_filter.tfam ln -s ../01.GWAS/snp_filter.nosex ln -s ../01.GWAS/snp_filter.map\nln -s ../01.GWAS/trait.sort.txt ln -s ../../02.population_genetics/03.structure/all.3.Q\npaste snp_filter.nosex all.3.Q |awk â€˜{print $1\" â€œ$1â€ 1 â€œ$3â€ â€œ$4}â€™ \u003e pop.Qmatrix\n/pub/software/emmax/emmax-intel64 -v -d 10 -t snp_filter -p trait.sort.txt -k pop.kinship -c pop.Qmatrix -o emmax.out 1\u003e emmax.log 2\u003eemmax.err\nstep4.draw_manhattan paste snp_filter.map emmax.out.ps | awk â€˜BEGIN{print â€œSNP\\tCHR\\tBP\\tPâ€}{if($2==$5){print $2â€\\t\"$1\"\\t\"$4\"\\t\"$NF}}â€™ \u003e emmax.out.ps.manht_input Rscript ../script/manhattan.R emmax.out.ps.manht_input emmax.out.ps.manht_figure_Q\nGWAS_realData 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19\nstep1.prepare prepare vcf plink â€“vcf ../data_real/Sample215.M5M8H3.2allel.vcf â€“maf 0.05 â€“geno 0.1 â€“recode12 â€“output-missing-genotype 0 â€“transpose â€“out snp_filter â€“set-missing-var-ids @:# â€“allow-extra-chr\nprepare phenotype perl ../script/sort_pheno.pl snp_filter.tfam ../data_real/trait.C16_0.table \u003e trait.sort.txt\nstep2.kinship /pub/software/emmax/emmax-kin-intel64 -v -d 10 -o ./pop.kinship snp_filter\nstep3.gwas /pub/software/emmax/emmax-intel64 -v -d 10 -t snp_filter -p trait.sort.txt -k pop.kinship -o emmax.out 1\u003e emmax.log 2\u003eemmax.err\nstep4.draw_manhattan paste snp_filter.map emmax.out.ps | awk â€˜BEGIN{print â€œSNP\\tCHR\\tBP\\tPâ€}{if($2==$5){print $2\"\\t\"$1\"\\t\"$4\"\\t\"$NF}}â€™ \u003e emmax.out.ps.manht_input Rscript ../script/manhattan.R emmax.out.ps.manht_input emmax.out.ps.manht_figure_readData\nQTL-seq change_VCF2Table 1 2 3 4 5 ref= vcf= outfile= gatk VariantsToTable -R $ref -V $vcf -F CHROM -F POS -F REF -F ALT -GF AD -GF DP -GF GQ -GF PL -O $outfile\nQTLSeq_Analysis 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 library(QTLseqr) library(â€œggplot2â€)\nHighBulk \u003c- â€œSRR834931â€ ## é«˜å€¼æ··æ± åç§° LowBulk \u003c- â€œSRR834927â€ ## ä½å€¼æ··æ± åç§° Chroms \u003c- paste0(rep(â€œChrâ€, 12), 1:12) ## æŸ“è‰²ä½“åˆ—è¡¨\nSNPæ•°æ®è¯»å– df \u003c- importFromGATK(file = â€œ../data/Yang_et_al_2013.tableâ€ , highBulk = HighBulk, # æŒ‡å®šé«˜å€¼æ··æ±  lowBulk = LowBulk, # æŒ‡å®šä½è‡³æ··æ±  chromList = Chroms # æŒ‡å®šåˆ†æç”¨æŸ“è‰²ä½“åˆ—è¡¨ )\nç»˜åˆ¶æ·±åº¦åˆ†å¸ƒå›¾ p1 \u003c- ggplot(data = df) + geom_histogram(aes(x = DP.HIGH + DP.LOW)) + xlim(0,800)\npdf(file=â€œSNP_depth.pdfâ€) p1 dev.off()\nç»˜åˆ¶refç­‰ä½åŸºå› é¢‘ç‡åˆ†å¸ƒå›¾ p2 \u003c- ggplot(data = df) + geom_histogram(aes(x = REF_FRQ))\npdf(file=â€œref_allele_frequency.pdfâ€) p2 dev.off()\nç»˜åˆ¶é«˜å€¼æ··æ± SNP-indexåˆ†å¸ƒ p3 \u003c- ggplot(data = df) + geom_histogram(aes(x = SNPindex.HIGH)) pdf(file=â€œSNPindex.HIGH.dis.pdfâ€) p3 dev.off()\nç»˜åˆ¶ä½å€¼æ··æ± SNP-indexåˆ†å¸ƒ p4 \u003c- ggplot(data = df) + geom_histogram(aes(x = SNPindex.LOW))\npdf(file=â€œSNPindex.LOW.dis.pdfâ€) p4 dev.off()\nSNPè¿‡æ»¤ df_filt \u003c- filterSNPs( SNPset = df, refAlleleFreq = 0.20, ## ref alleleé¢‘ç‡è¿‡æ»¤ï¼Œ0.2è¡¨ç¤º0.2~0.8ä¹‹é—´ minTotalDepth = 100, ## æœ€å°æ·±åº¦è¿‡æ»¤ maxTotalDepth = 400, ## æœ€å¤§æ·±åº¦è¿‡æ»¤ minSampleDepth = 40, ## å•ä¸ªæ ·å“æœ€å°æ·±åº¦ minGQ = 99, ## genotype quality è¿‡æ»¤ verbose = TRUE ## è¾“å‡ºæ—¥å¿— )\nè¿›è¡ŒdeltaSNPindex è®¡ç®— df_filt \u003c- runQTLseqAnalysis(df_filt, windowSize = 1e6, ## çª—å£å¤§å° popStruc = â€œF2â€, ## ç¾¤ä½“ç±»å‹ï¼ŒF2 æˆ–è€…RIL bulkSize = c(385, 430), ## æ··æ ·ä¸ªæ•°ï¼Œç¬¬ä¸€ä¸ªé«˜å€¼ç»„ replications = 10000, ## bootstrapæ¬¡æ•° intervals = c(95, 99) ## ç½®ä¿¡åŒºé—´ )\nSNP æ²¿æŸ“è‰²ä½“åˆ†å¸ƒå›¾ p5 \u003c- plotQTLStats(SNPset = df_filt, var = â€œnSNPsâ€) pdf(file=â€œSNP_filter.window.pdfâ€, width = 20, height = 4) p5 dev.off()\ndeltaSNPindexæ²¿æŸ“è‰²ä½“åˆ†å¸ƒ p6 \u003c- plotQTLStats(SNPset = df_filt, var = â€œdeltaSNPâ€, plotIntervals = TRUE) pdf(file=â€œdeltaSNPindex.pdfâ€, width = 20, height = 4) p6 dev.off()\næå–æ˜¾è‘—æ€§åŒºåŸŸ QTL \u003c- getSigRegions(SNPset = df_filt, method = â€œQTLseqâ€, interval = 95)\nè¾“å‡ºåˆ°æ–‡ä»¶ write.table(QTL[[1]], â€œQTLseq_result.SigRegions.tableâ€, sep=\"\\t\",quote=F)\næå–QTLseqç»“æœ results99 \u003c- getQTLTable( SNPset = df_filt,\nmethod = â€œQTLseqâ€, interval = 99, ## ç»“æœé˜ˆå€¼ export = FALSE)\nè¾“å‡ºåˆ°æ–‡ä»¶ write.table(results99, file=â€œQTLseq_result_deltaSNPindex_CI99.tableâ€ , sep=\"\\t\", quote = F)\nğŸ’Œlixiang117423@gmail.com\nğŸ’Œlixiang117423@foxmail.com\nç”Ÿç‰©ä¿¡æ¯å­¦ - ç¾¤ä½“é—ä¼ å­¦ - Linux #ç”Ÿç‰©ä¿¡æ¯å­¦ #Linux\n","wordCount":"1932","inLanguage":"en","datePublished":"2026-02-03T12:08:55+08:00","dateModified":"2026-02-03T12:08:55+08:00","mainEntityOfPage":{"@type":"WebPage","@id":"https://Tdxxxx-student.github.io/posts/%E9%87%8D%E6%B5%8B%E5%BA%8F%E6%95%99%E7%A8%8B/"},"publisher":{"@type":"Organization","name":"æˆ‘çš„åšå®¢","logo":{"@type":"ImageObject","url":"https://Tdxxxx-student.github.io/favicon.ico"}}}</script></head><body id=top><header class=header><nav class=nav><div class=logo><a href=https://Tdxxxx-student.github.io/ accesskey=h title="æˆ‘çš„åšå®¢ (Alt + H)">æˆ‘çš„åšå®¢</a><div class=logo-switches><button id=theme-toggle accesskey=t title="(Alt + T)" aria-label="Toggle theme">
<svg id="moon" width="24" height="18" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><path d="M21 12.79A9 9 0 1111.21 3 7 7 0 0021 12.79z"/></svg>
<svg id="sun" width="24" height="18" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><circle cx="12" cy="12" r="5"/><line x1="12" y1="1" x2="12" y2="3"/><line x1="12" y1="21" x2="12" y2="23"/><line x1="4.22" y1="4.22" x2="5.64" y2="5.64"/><line x1="18.36" y1="18.36" x2="19.78" y2="19.78"/><line x1="1" y1="12" x2="3" y2="12"/><line x1="21" y1="12" x2="23" y2="12"/><line x1="4.22" y1="19.78" x2="5.64" y2="18.36"/><line x1="18.36" y1="5.64" x2="19.78" y2="4.22"/></svg></button></div></div><ul id=menu><li><a href=https://Tdxxxx-student.github.io/posts title=ä¸ªäºº><span>ä¸ªäºº</span></a></li><li><a href=https://Tdxxxx-student.github.io/docs title=æ•™ç¨‹><span>æ•™ç¨‹</span></a></li><li><a href=https://Tdxxxx-student.github.io/study title=å­¦ä¹ ><span>å­¦ä¹ </span></a></li></ul></nav></header><main class=main><article class=post-single><header class=post-header><h1 class="post-title entry-hint-parent">é‡æµ‹åºæ•™ç¨‹</h1><div class=post-meta><span title='2026-02-03 12:08:55 +0800 +0800'>February 3, 2026</span></div></header><div class=post-content><p>é‡æµ‹åºä¸ç¾¤ä½“é—ä¼ å­¦
äº‘å—å†œä¸šå¤§å­¦ äº‘å—ç”Ÿç‰©èµ„æºä¿æŠ¤ä¸åˆ©ç”¨å›½å®¶é‡ç‚¹å®éªŒå®¤
ï¼ˆæè¯¦ 2019å¹´12æœˆ31æ—¥ï¼‰
è½¯ä»¶å®‰è£…
è¯¥éƒ¨åˆ†è½¯ä»¶å¾ˆéš¾ç”¨condaç›´æ¥å®‰è£…ï¼Œå®‰è£…æ­¥éª¤æ¯”è¾ƒç‰¹æ®Šã€‚</p><p>lumpy-sv
1
2
3
4
5
6
7
lumpy-sv ä¾èµ– Python2.7ï¼Œåˆ‡æ¢åˆ°py27ç¯å¢ƒä¸‹è¿›è¡Œå®‰è£…ã€‚
conda activate py27
git clone &ndash;recursive <a href=https://github.com/arq5x/lumpy-sv.git>https://github.com/arq5x/lumpy-sv.git</a>
cd lumpy-sv
make</p><h2 id=å®‰è£…svtyper>å®‰è£…svtyper<a hidden class=anchor aria-hidden=true href=#å®‰è£…svtyper>#</a></h2><p>conda install svtyper
cnvnator
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
cnvnatorä¾èµ–äºrootè½¯ä»¶åŒ…åŠsamtoolsè½¯ä»¶åŒ…(åŒ…å«HTSlib)ï¼Œå› æ­¤å…ˆå®‰è£…SamtoolsåŠrootè½¯ä»¶åŒ…ã€‚</p><h2 id=å®‰è£…samtools>å®‰è£…samtools<a hidden class=anchor aria-hidden=true href=#å®‰è£…samtools>#</a></h2><p>wget <a href=https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2>https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2</a>
tar -xvf samtools-1.9.tar.bz2
cd samtools-1.9/
./configure
make</p><h2 id=å®‰è£…rootè½¯ä»¶-ç›´æ¥ä¸‹è½½åè§£å‹å³å¯>å®‰è£…rootè½¯ä»¶, ç›´æ¥ä¸‹è½½åè§£å‹å³å¯<a hidden class=anchor aria-hidden=true href=#å®‰è£…rootè½¯ä»¶-ç›´æ¥ä¸‹è½½åè§£å‹å³å¯>#</a></h2><p>wget <a href=https://root.cern/download/root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz>https://root.cern/download/root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz</a>
tar -zxvf root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz</p><h2 id=å®‰è£…cnvnator>å®‰è£…cnvnator<a hidden class=anchor aria-hidden=true href=#å®‰è£…cnvnator>#</a></h2><p>git clone <a href=https://github.com/abyzovlab/CNVnator.git>https://github.com/abyzovlab/CNVnator.git</a>
cd CNVnator</p><h2 id=é“¾æ¥samtoolsè½¯ä»¶ç›®å½•åˆ°cnvnatorç›®å½•>é“¾æ¥samtoolsè½¯ä»¶ç›®å½•åˆ°cnvnatorç›®å½•<a hidden class=anchor aria-hidden=true href=#é“¾æ¥samtoolsè½¯ä»¶ç›®å½•åˆ°cnvnatorç›®å½•>#</a></h2><p>ln -s ../samtools-1.9 ./samtools</p><h2 id=åŠ è½½rootè½¯ä»¶åˆ°ç¯å¢ƒå˜é‡>åŠ è½½rootè½¯ä»¶åˆ°ç¯å¢ƒå˜é‡<a hidden class=anchor aria-hidden=true href=#åŠ è½½rootè½¯ä»¶åˆ°ç¯å¢ƒå˜é‡>#</a></h2><p>export ROOTSYS=/pub/software/root
export PATH=/pub/software/root/bin:$PATH
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ROOTSYS/lib</p><h2 id=å®‰è£…cnvnator-1>å®‰è£…cnvnator<a hidden class=anchor aria-hidden=true href=#å®‰è£…cnvnator-1>#</a></h2><p>make LIBS="-lcrypto"</p><p>less ../data/sample.list | awk &lsquo;{ print &ldquo;bwa mem -t 2 -R &lsquo;'&rsquo;@RG\tID:"$1&rdquo;\tSM:"$1"\tPL:illumina&rsquo;'&rsquo; ../01.ref/genome.fasta ../data/"$1"_1.fq.gz ../data/"$1"_2.fq.gz | /pub/software/samtools/samtools sort -@ 2 -m 1G -o &ldquo;$1&rdquo;.sort.bam -" }&rsquo;
å˜å¼‚æ£€æµ‹
æ„å»ºåŸºå› ç»„index
1
2
3
4
5
6
7
8
9
10
11</p><h1 id=create-link>create link<a hidden class=anchor aria-hidden=true href=#create-link>#</a></h1><p>ln -s ../data/genome.fasta ./genome.fasta</p><h1 id=for-gatk>for gatk<a hidden class=anchor aria-hidden=true href=#for-gatk>#</a></h1><p>samtools faidx genome.fasta</p><h1 id=for-bwa>for bwa<a hidden class=anchor aria-hidden=true href=#for-bwa>#</a></h1><p>bwa index genome.fasta</p><h1 id=for-picard>for picard<a hidden class=anchor aria-hidden=true href=#for-picard>#</a></h1><p>picard CreateSequenceDictionary R=genome.fasta
æ¯”å¯¹æ’åºå»é‡
1
2
3
4
5
6
7
8
9
10
11</p><h1 id=æ¯”å¯¹æ’åº>æ¯”å¯¹æ’åº<a hidden class=anchor aria-hidden=true href=#æ¯”å¯¹æ’åº>#</a></h1><p>bwa mem -t 2 -R &lsquo;@RG\tID:S1\tSM:S1\tPL:illumina&rsquo; ../01.ref/genome.fasta ../data/S1_1.fq.gz ../data/S1_2.fq.gz | /pub/software/samtools/samtools sort -@ 2 -m 1G -o S1.sort.bam -
bwa mem -t 2 -R &lsquo;@RG\tID:S2\tSM:S2\tPL:illumina&rsquo; ../01.ref/genome.fasta ../data/S2_1.fq.gz ../data/S2_2.fq.gz | /pub/software/samtools/samtools sort -@ 2 -m 1G -o S2.sort.bam -</p><h1 id=å»é™¤é‡å¤>å»é™¤é‡å¤<a hidden class=anchor aria-hidden=true href=#å»é™¤é‡å¤>#</a></h1><p>picard -Xmx4g MarkDuplicates I=S1.sort.bam O=S1.sort.rmdup.bam CREATE_INDEX=true REMOVE_DUPLICATES=true M=S1.marked_dup_metrics.txt
picard -Xmx4g MarkDuplicates I=S2.sort.bam O=S2.sort.rmdup.bam CREATE_INDEX=true REMOVE_DUPLICATES=true M=S2.marked_dup_metrics.txt</p><h1 id=æ¯”å¯¹ç‡ç»Ÿè®¡>æ¯”å¯¹ç‡ç»Ÿè®¡<a hidden class=anchor aria-hidden=true href=#æ¯”å¯¹ç‡ç»Ÿè®¡>#</a></h1><p>/pub/software/samtools/samtools flagstat S1.sort.bam > S1.sort.bam.flagstat
/pub/software/samtools/samtools flagstat S2.sort.bam > S2.sort.bam.flagstat
SNPã€Indelæ£€æµ‹
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24</p><h1 id=step1haplotypecaller>step1.HaplotypeCaller<a hidden class=anchor aria-hidden=true href=#step1haplotypecaller>#</a></h1><p>gatk &ndash;java-options &ldquo;-Xmx10g -Djava.io.tmpdir=./tmp&rdquo; HaplotypeCaller -R ../01.ref/genome.fasta -I ../02.mapping/S1.sort.rmdup.bam -ERC GVCF -O S1.g.vcf 1>S1.HC.log 2>&amp;1
gatk &ndash;java-options &ldquo;-Xmx10g -Djava.io.tmpdir=./tmp&rdquo; HaplotypeCaller -R ../01.ref/genome.fasta -I ../02.mapping/S2.sort.rmdup.bam -ERC GVCF -O S2.g.vcf 1>S2.HC.log 2>&amp;1</p><h1 id=step2combinegvcfs>step2.CombineGVCFs<a hidden class=anchor aria-hidden=true href=#step2combinegvcfs>#</a></h1><p>ls ./S*.g.vcf > gvcf.list
gatk &ndash;java-options &ldquo;-Xmx4g -Djava.io.tmpdir=./tmp&rdquo; CombineGVCFs -R ../01.ref/genome.fasta -V gvcf.list -O all.merge.g.vcf</p><h1 id=step3genotypegvcfs>step3.GenotypeGVCFs<a hidden class=anchor aria-hidden=true href=#step3genotypegvcfs>#</a></h1><p>gatk &ndash;java-options &ldquo;-Xmx4g -Djava.io.tmpdir=./tmp&rdquo; GenotypeGVCFs -R ../01.ref/genome.fasta &ndash;variant all.merge.g.vcf -O all.merge_raw.vcf</p><h1 id=step4snp>step4.SNP<a hidden class=anchor aria-hidden=true href=#step4snp>#</a></h1><p>gatk &ndash;java-options &ldquo;-Xmx4g -Djava.io.tmpdir=./tmp&rdquo; SelectVariants -R ../01.ref/genome.fasta -V all.merge_raw.vcf &ndash;select-type SNP -O all.raw.snp.vcf</p><p>gatk &ndash;java-options &ldquo;-Xmx4g -Djava.io.tmpdir=./tmp&rdquo; VariantFiltration -R ../01.ref/genome.fasta -V all.raw.snp.vcf &ndash;filter-expression &ldquo;QD &lt; 2.0 || MQ &lt; 40.0 || FS > 60.0 || SOR > 3.0 || MQRankSum &lt; -12.5 || ReadPosRankSum &lt; -8.0&rdquo; &ndash;filter-name &lsquo;SNP_filter&rsquo; -O all.filter.snp.vcf</p><p>gatk &ndash;java-options &ldquo;-Xmx4g -Djava.io.tmpdir=./tmp&rdquo; SelectVariants -R ../01.ref/genome.fasta -V all.filter.snp.vcf &ndash;exclude-filtered -O all.filtered.snp.vcf</p><h1 id=step5indel>step5.InDel<a hidden class=anchor aria-hidden=true href=#step5indel>#</a></h1><p>gatk &ndash;java-options &ldquo;-Xmx4g -Djava.io.tmpdir=./tmp&rdquo; SelectVariants -R ../01.ref/genome.fasta -V all.merge_raw.vcf &ndash;select-type INDEL -O all.raw.indel.vcf</p><p>gatk &ndash;java-options &ldquo;-Xmx4g -Djava.io.tmpdir=./tmp&rdquo; VariantFiltration -R ../01.ref/genome.fasta -V all.raw.indel.vcf &ndash;filter-expression &ldquo;QD &lt; 2.0 || FS > 200.0 || SOR > 10.0 || MQRankSum &lt; -12.5 || ReadPosRankSum &lt; -8.0&rdquo; &ndash;filter-name &lsquo;INDEL_filter&rsquo; -O all.filter.indel.vcf</p><p>gatk &ndash;java-options &ldquo;-Xmx4g -Djava.io.tmpdir=./tmp&rdquo; SelectVariants -R ../01.ref/genome.fasta -V all.filter.indel.vcf &ndash;exclude-filtered -O all.filtered.indel.vcf
SVæ£€æµ‹
1
2
3
4
5
6
7
8
9
10
11
12
13
14</p><h1 id=step1discordants_bam>step1.discordants_bam<a hidden class=anchor aria-hidden=true href=#step1discordants_bam>#</a></h1><p>/pub/software/samtools/samtools view -b -F 1294 -@ 12 ../02.mapping/S1.sort.rmdup.bam > S1.discordants.bam
/pub/software/samtools/samtools view -b -F 1294 -@ 12 ../02.mapping/S2.sort.rmdup.bam > S2.discordants.bam</p><h1 id=step2splitters_bam>step2.splitters_bam<a hidden class=anchor aria-hidden=true href=#step2splitters_bam>#</a></h1><p>/pub/software/samtools/samtools view -h ../02.mapping/S1.sort.rmdup.bam | /pub/software/lumpy-sv/scripts/extractSplitReads_BwaMem -i stdin | /pub/software/samtools/samtools sort - > S1.splitters.bam
/pub/software/samtools/samtools view -h ../02.mapping/S2.sort.rmdup.bam | /pub/software/lumpy-sv/scripts/extractSplitReads_BwaMem -i stdin | /pub/software/samtools/samtools sort - > S2.splitters.bam</p><h1 id=step3lumpy>step3.lumpy<a hidden class=anchor aria-hidden=true href=#step3lumpy>#</a></h1><p>/pub/software/lumpy-sv/bin/lumpyexpress -B ../02.mapping/S1.sort.rmdup.bam,../02.mapping/S2.sort.rmdup.bam -S S1.splitters.bam,S2.splitters.bam -D S1.discordants.bam,S2.discordants.bam -o all.sv.lumpy.vcf</p><h1 id=step4ind_genotype>step4.ind_genotype<a hidden class=anchor aria-hidden=true href=#step4ind_genotype>#</a></h1><p>vcftools &ndash;vcf all.sv.lumpy.vcf &ndash;indv S1 &ndash;recode &ndash;recode-INFO-all &ndash;out S1 && svtyper -i S1.recode.vcf -B ../02.mapping/S1.sort.rmdup.bam -o S1.genotype.vcf
vcftools &ndash;vcf all.sv.lumpy.vcf &ndash;indv S2 &ndash;recode &ndash;recode-INFO-all &ndash;out S2 && svtyper -i S2.recode.vcf -B ../02.mapping/S2.sort.rmdup.bam -o S2.genotype.vcf
CNVæ£€æµ‹
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37</p><h1 id=step1prepare_genome>step1.prepare_genome<a hidden class=anchor aria-hidden=true href=#step1prepare_genome>#</a></h1><h2 id=split-genome-into-single-sequence>split genome into single sequence<a hidden class=anchor aria-hidden=true href=#split-genome-into-single-sequence>#</a></h2><p>ln -s ../data/genome.fasta ./genome.fa</p><p>seqkit split -i ./genome.fa -O split</p><p>ls ./split/genome.id_*.fa |sed &rsquo;s/./split/genome.id_//&rsquo; |awk &lsquo;{print &ldquo;mv ./split/genome.id_"$aa&rdquo; ./split/"$aa}&rsquo; > mv_name.sh</p><p>sh mv_name.sh</p><h1 id=step2cnvnator>step2.cnvnator<a hidden class=anchor aria-hidden=true href=#step2cnvnator>#</a></h1><p>export ROOTSYS=/pub/software/root
export PATH=/pub/software/root/bin:$PATH
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ROOTSYS/lib</p><h2 id=é“¾æ¥æ–‡ä»¶åˆ°å½“å‰ç›®å½•>é“¾æ¥æ–‡ä»¶åˆ°å½“å‰ç›®å½•<a hidden class=anchor aria-hidden=true href=#é“¾æ¥æ–‡ä»¶åˆ°å½“å‰ç›®å½•>#</a></h2><p>ln -s ../02.mapping/S1.sort.rmdup.bam</p><h2 id=æå–æ¯”å¯¹ç»“æœ>æå–æ¯”å¯¹ç»“æœ<a hidden class=anchor aria-hidden=true href=#æå–æ¯”å¯¹ç»“æœ>#</a></h2><p>/pub/software/CNVnator/cnvnator -genome genome.fa -root S1.root -tree S1.sort.rmdup.bam</p><p>##ç”Ÿæˆæ·±åº¦åˆ†å¸ƒ
/pub/software/CNVnator/cnvnator -genome genome.fa -root S1.root -his 500 -d split</p><h2 id=è¿›è¡Œç»Ÿè®¡è®¡ç®—>è¿›è¡Œç»Ÿè®¡è®¡ç®—<a hidden class=anchor aria-hidden=true href=#è¿›è¡Œç»Ÿè®¡è®¡ç®—>#</a></h2><p>/pub/software/CNVnator/cnvnator -root S1.root -stat 500</p><h2 id=æ£€æŸ¥bin-sizeæ˜¯å¦åˆé€‚>æ£€æŸ¥bin sizeæ˜¯å¦åˆé€‚<a hidden class=anchor aria-hidden=true href=#æ£€æŸ¥bin-sizeæ˜¯å¦åˆé€‚>#</a></h2><p>/pub/software/CNVnator/cnvnator -root S1.root -eval 500 > S1.eval.ratio</p><h2 id=rdä¿¡å·åˆ†å‰²>RDä¿¡å·åˆ†å‰²<a hidden class=anchor aria-hidden=true href=#rdä¿¡å·åˆ†å‰²>#</a></h2><p>/pub/software/CNVnator/cnvnator -root S1.root -partition 500</p><h2 id=è¿›è¡Œcnvæ£€æµ‹>è¿›è¡ŒCNVæ£€æµ‹<a hidden class=anchor aria-hidden=true href=#è¿›è¡Œcnvæ£€æµ‹>#</a></h2><p>/pub/software/CNVnator/cnvnator -root S1.root -call 500 > S1.cnv
SNP INDELæ³¨é‡Š
1
2
3
4
5
6
7
8
9
10
11
12
13</p><h1 id=step1prepare_genome-1>step1.prepare_genome<a hidden class=anchor aria-hidden=true href=#step1prepare_genome-1>#</a></h1><p>ln -s ../data/genome.gtf
ln -s ../data/genome.fasta</p><p>gtfToGenePred -genePredExt genome.gtf genome_refGene.txt
/pub/software/annovar/retrieve_seq_from_fasta.pl &ndash;format refGene &ndash;seqfile genome.fasta genome_refGene.txt &ndash;out genome_refGeneMrna.fa</p><p>#step2.ann_vcf
ln -s ../03.SNP_indel/all.filtered.snp.vcf</p><p>/pub/software/annovar/convert2annovar.pl -format vcf4old all.filtered.snp.vcf >all.snp.vcf.annovar.input</p><p>/pub/software/annovar/annotate_variation.pl -geneanno &ndash;neargene 2000 -buildver genome -dbtype refGene -outfile all.anno -exonsort all.snp.vcf.annovar.input ./
CNVå’ŒSVæ³¨é‡Š
1
2
3
4
5
6
7
less -S ../04.SV/all.sv.lumpy.vcf|awk -F &ldquo;;&rdquo; &lsquo;{ if($1!<del>/#/){print $1"\t"$2"\t"$3"\t"$4} }&rsquo;|awk &lsquo;{print $1"\t"$2"\t"$11"\t"$3"\t"$8}&rsquo;|sed &rsquo;s/END=//&rsquo; |sed &rsquo;s/SVTYPE=//&rsquo;| awk &lsquo;$3&lt;$2&rsquo; > SV.bed
less -S ../04.SV/all.sv.lumpy.vcf|awk -F &ldquo;;&rdquo; &lsquo;{ if($1!</del>/#/){print $1"\t"$2"\t"$3"\t"$4} }&rsquo;|awk &lsquo;{print $1"\t"$2"\t"$11"\t"$3"\t"$8}&rsquo;|sed &rsquo;s/END=//&rsquo; |sed &rsquo;s/SVTYPE=//&rsquo; > SV.bed
awk &lsquo;$5 !~ /BND/&rsquo; SV.bed > SV_filter.bed
bedtools intersect -wo -a SV_filter.bed -b gene.gtf > SV_filter.Anno
less -S ../05.CNV/S1.cnv | awk &lsquo;{print $2 &ldquo;\t&rdquo; $1}&rsquo;|sed &rsquo;s/:/\t/&rsquo;|sed &rsquo;s/-/\t/&rsquo; > CNV.bed
bedtools intersect -wo -a CNV.bed -b gene.gtf > CNV.Anno</p><p>ç¾¤ä½“ç»“æ„
è¿‡æ»¤
1
2
3
4
5
6
7
8
9
vcf=../data/all.vcf</p><h2 id=filter-missing-and-maf>filter missing and maf<a hidden class=anchor aria-hidden=true href=#filter-missing-and-maf>#</a></h2><p>plink &ndash;vcf $vcf &ndash;geno 0.1 &ndash;maf 0.01 &ndash;out all.missing_maf &ndash;recode vcf-iid &ndash;allow-extra-chr &ndash;set-missing-var-ids @:# &ndash;keep-allele-order</p><h2 id=filter-ld>filter LD<a hidden class=anchor aria-hidden=true href=#filter-ld>#</a></h2><p>plink &ndash;vcf all.missing_maf.vcf &ndash;indep-pairwise 50 10 0.2 &ndash;out tmp.ld &ndash;allow-extra-chr &ndash;set-missing-var-ids @:#
plink &ndash;vcf all.missing_maf.vcf &ndash;make-bed &ndash;extract tmp.ld.prune.in &ndash;out all.LDfilter &ndash;recode vcf-iid &ndash;keep-allele-order &ndash;allow-extra-chr &ndash;set-missing-var-ids @:#</p><p>å»ºæ ‘
NJ
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
perl ../../script/vcf2phy.pl ../../00.filter/all.LDfilter.vcf > sequences.dat</p><p>echo -e &ldquo;sequences.dat\nY&rdquo; > dnadist.cfg</p><p>dnadist &lt; dnadist.cfg >dnadist.log</p><p>mv outfile infile.dist</p><p>echo -e &ldquo;infile.dist\nY&rdquo; > neighbor.cfg</p><p>neighbor &lt; neighbor.cfg >nj.log</p><p>less infile.dist | tr &lsquo;\n&rsquo; &lsquo;|&rsquo;| sed &rsquo;s/| / /g&rsquo; | tr &lsquo;|&rsquo; &lsquo;\n&rsquo; >infile.dist.table
less outtree | tr &lsquo;\n&rsquo; &rsquo; &lsquo;|sed &rsquo;s/ //g&rsquo; > outtree.nwk</p><p>tree_snphylo
1
2
/pub/software/SNPhylo/snphylo.sh -v ../../00.filter/all.LDfilter.vcf -r -l 1 -m 0.01 -o GA0001 -b -B 2 -P tree_snphylo</p><p>PCA
1
2
3
4
plink &ndash;vcf ../00.filter/all.LDfilter.vcf &ndash;pca 10 &ndash;out PCA_out &ndash;allow-extra-chr &ndash;set-missing-var-ids @:#</p><p>Rscript ../script/draw_PCA.R PCA_out.eigenvec 1 2 ../data/sample.pop PCA_out_figure</p><p>ç¾¤ä½“ç»“æ„
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19</p><h1 id=step0vcf2bed>step0.vcf2bed<a hidden class=anchor aria-hidden=true href=#step0vcf2bed>#</a></h1><p>plink &ndash;vcf ../00.filter/all.LDfilter.vcf &ndash;make-bed &ndash;out all &ndash;allow-extra-chr &ndash;keep-allele-order &ndash;set-missing-var-ids @:#</p><h1 id=step1get_admix_shell>step1.get_Admix_shell<a hidden class=anchor aria-hidden=true href=#step1get_admix_shell>#</a></h1><p>seq 2 4 | awk &lsquo;{print &ldquo;admixture &ndash;cv -j2 all.bed &ldquo;$1&rdquo; 1>admix."$1&rdquo;.log 2>&amp;1"}&rsquo; > admixture.sh</p><h1 id=step2run_admix>step2.run_admix<a hidden class=anchor aria-hidden=true href=#step2run_admix>#</a></h1><p>sh admixture.sh</p><h1 id=step3draw_structure>step3.draw_structure<a hidden class=anchor aria-hidden=true href=#step3draw_structure>#</a></h1><p>mkdir result
cp ./*.Q result/
Rscript ../script/draw_admixture.R result all.nosex structure</p><h1 id=admixture>admixture<a hidden class=anchor aria-hidden=true href=#admixture>#</a></h1><p>admixture &ndash;cv -j2 all.bed 2 1>admix.2.log 2>&amp;1
admixture &ndash;cv -j2 all.bed 3 1>admix.3.log 2>&amp;1
admixture &ndash;cv -j2 all.bed 4 1>admix.4.log 2>&amp;1</p><p>é—ä¼ è¡°å‡
1
2
3
4
5
6
7
8
9
10</p><h1 id=step1ld_calcu>step1.LD_calcu<a hidden class=anchor aria-hidden=true href=#step1ld_calcu>#</a></h1><p>/pub/software/PopLDdecay/bin/PopLDdecay -InVCF ../data/all.vcf -SubPop ../data/pop.SC.table -MaxDist 500 -OutStat pop.SC.stat
/pub/software/PopLDdecay/bin/PopLDdecay -InVCF ../data/all.vcf -SubPop ../data/pop.YZR.table -MaxDist 500 -OutStat pop.YZR.stat</p><h1 id=step2draw_singleplot>step2.draw_singlePlot<a hidden class=anchor aria-hidden=true href=#step2draw_singleplot>#</a></h1><p>perl /pub/software/PopLDdecay/bin/Plot_OnePop.pl -inFile pop.SC.stat.gz -output pop.SC.ld
perl /pub/software/PopLDdecay/bin/Plot_OnePop.pl -inFile pop.YZR.stat.gz -output pop.YZR.ld</p><h1 id=step3draw_multiplotarepoplddecaybinplot_multipoppl--inlist-ld_statlist--output-ld_statmulti>step3.draw_multiPlotare/PopLDdecay/bin/Plot_MultiPop.pl -inList ld_stat.list -output ld_stat.multi<a hidden class=anchor aria-hidden=true href=#step3draw_multiplotarepoplddecaybinplot_multipoppl--inlist-ld_statlist--output-ld_statmulti>#</a></h1><p>é€‰æ‹©åˆ†æ
i_ROD
1
2
3
4
5
6
7
8
9
10
11
12</p><h1 id=step1pi>step1.pi<a hidden class=anchor aria-hidden=true href=#step1pi>#</a></h1><p>vcf=../../data/all.vcf
window=20000
step=2000
vcftools &ndash;vcf $vcf &ndash;window-pi $window &ndash;window-pi-step $step &ndash;keep ../../data/pop.SC.table &ndash;out ./Pi.pop1
vcftools &ndash;vcf $vcf &ndash;window-pi $window &ndash;window-pi-step $step &ndash;keep ../../data/pop.YZR.table &ndash;out ./Pi.pop2</p><h1 id=step2rod>step2.ROD<a hidden class=anchor aria-hidden=true href=#step2rod>#</a></h1><p>pi1=Pi.pop1.windowed.pi
pi2=Pi.pop2.windowed.pi
perl ../../script/merge2pi_ROD.pl $pi1 $pi2 >Pi.ROD</p><p>tajimaD
1
2
3
4
5
6
vcf=../../data/all.vcf
window=20000</p><p>vcftools &ndash;vcf $vcf &ndash;TajimaD $window &ndash;keep ../../data/pop.SC.table &ndash;out TajimaD.pop1
vcftools &ndash;vcf $vcf &ndash;TajimaD $window &ndash;keep ../../data/pop.YZR.table &ndash;out TajimaD.pop2</p><p>Fst
1
2
3
4
5
vcf=../../data/all.vcf
window=20000
step=2000
vcftools &ndash;vcf $vcf &ndash;fst-window-size $window &ndash;fst-window-step $step &ndash;weir-fst-pop ../../data/pop.SC.table &ndash;weir-fst-pop ../../data/pop.YZR.table &ndash;out ./Fst.pop1.pop2</p><p>ROD_Fst
1
2
3
4
5
6
7
8
9
10
11
12
13
fst=../03.Fst/Fst.pop1.pop2.windowed.weir.fst
ROD=../01.pi_ROD/Pi.ROD
gff=../../data/genome.gff</p><p>perl ../../script/top_Fst_ROD_S.pl $fst $ROD 0.05 0.05 Fst_ROD.table Fst_ROD.stat</p><p>awk &lsquo;$NF==&ldquo;top&rdquo;&rsquo; Fst_ROD.table > Fst_ROD.table.top</p><p>awk &lsquo;$3==&ldquo;gene&rdquo;&rsquo; $gff > gene.gff</p><p>bedtools intersect -wo -F 0.1 -a Fst_ROD.table.top -b gene.gff | awk -F &ldquo;\t&rdquo; &lsquo;{print $7"\t"$10"\t"$11"\t"$13"\t"$15}&rsquo; | sort -u > Fst_ROD.table.top.gene</p><p>GWAS
mpute
1
2
beagle -Xmx4g -Djava.io.tmpdir=./TMP gt=../data/all.vcf out=all.impute impute=true window=10 nthreads=2</p><p>GWAS
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19</p><h1 id=step1prepare>step1.prepare<a hidden class=anchor aria-hidden=true href=#step1prepare>#</a></h1><h2 id=prepare-vcf>prepare vcf<a hidden class=anchor aria-hidden=true href=#prepare-vcf>#</a></h2><p>plink &ndash;vcf ../data/all.vcf &ndash;maf 0.05 &ndash;geno 0.1 &ndash;recode12 &ndash;output-missing-genotype 0 &ndash;transpose &ndash;out snp_filter &ndash;set-missing-var-ids @:# &ndash;allow-extra-chr</p><h2 id=prepare-phenotype>prepare phenotype<a hidden class=anchor aria-hidden=true href=#prepare-phenotype>#</a></h2><p>perl ../script/sort_pheno.pl snp_filter.tfam ../data/trait.table > trait.sort.txt</p><h1 id=step2kinship>step2.kinship<a hidden class=anchor aria-hidden=true href=#step2kinship>#</a></h1><p>/pub/software/emmax/emmax-kin-intel64 -v -d 10 -o ./pop.kinship snp_filter</p><h1 id=step3gwas>step3.gwas<a hidden class=anchor aria-hidden=true href=#step3gwas>#</a></h1><p>/pub/software/emmax/emmax-intel64 -v -d 10 -t snp_filter -p trait.sort.txt -k pop.kinship -o emmax.out 1> emmax.log 2>emmax.err</p><h1 id=draw_manhattan>draw_manhattan<a hidden class=anchor aria-hidden=true href=#draw_manhattan>#</a></h1><p>paste snp_filter.map emmax.out.ps | awk &lsquo;BEGIN{print &ldquo;SNP\tCHR\tBP\tP&rdquo;}{if($2==$5){print $2"\t"$1"\t"$4"\t"$NF}}&rsquo; > emmax.out.ps.manht_input
Rscript ../script/manhattan.R emmax.out.ps.manht_input emmax.out.ps.manht_figure</p><p>GWAS-Q
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17</p><h1 id=step3gwas-1>step3.gwas<a hidden class=anchor aria-hidden=true href=#step3gwas-1>#</a></h1><p>ln -s ../01.GWAS/snp_filter.tped
ln -s ../01.GWAS/snp_filter.tfam
ln -s ../01.GWAS/snp_filter.nosex
ln -s ../01.GWAS/snp_filter.map</p><p>ln -s ../01.GWAS/trait.sort.txt
ln -s ../../02.population_genetics/03.structure/all.3.Q</p><p>paste snp_filter.nosex all.3.Q |awk &lsquo;{print $1" &ldquo;$1&rdquo; 1 &ldquo;$3&rdquo; &ldquo;$4}&rsquo; > pop.Qmatrix</p><p>/pub/software/emmax/emmax-intel64 -v -d 10 -t snp_filter -p trait.sort.txt -k pop.kinship -c pop.Qmatrix -o emmax.out 1> emmax.log 2>emmax.err</p><h1 id=step4draw_manhattan>step4.draw_manhattan<a hidden class=anchor aria-hidden=true href=#step4draw_manhattan>#</a></h1><p>paste snp_filter.map emmax.out.ps | awk &lsquo;BEGIN{print &ldquo;SNP\tCHR\tBP\tP&rdquo;}{if($2==$5){print $2&rdquo;\t"$1"\t"$4"\t"$NF}}&rsquo; > emmax.out.ps.manht_input
Rscript ../script/manhattan.R emmax.out.ps.manht_input emmax.out.ps.manht_figure_Q</p><p>GWAS_realData
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19</p><h1 id=step1prepare-1>step1.prepare<a hidden class=anchor aria-hidden=true href=#step1prepare-1>#</a></h1><h2 id=prepare-vcf-1>prepare vcf<a hidden class=anchor aria-hidden=true href=#prepare-vcf-1>#</a></h2><p>plink &ndash;vcf ../data_real/Sample215.M5M8H3.2allel.vcf &ndash;maf 0.05 &ndash;geno 0.1 &ndash;recode12 &ndash;output-missing-genotype 0 &ndash;transpose &ndash;out snp_filter &ndash;set-missing-var-ids @:# &ndash;allow-extra-chr</p><h2 id=prepare-phenotype-1>prepare phenotype<a hidden class=anchor aria-hidden=true href=#prepare-phenotype-1>#</a></h2><p>perl ../script/sort_pheno.pl snp_filter.tfam ../data_real/trait.C16_0.table > trait.sort.txt</p><h1 id=step2kinship-1>step2.kinship<a hidden class=anchor aria-hidden=true href=#step2kinship-1>#</a></h1><p>/pub/software/emmax/emmax-kin-intel64 -v -d 10 -o ./pop.kinship snp_filter</p><h1 id=step3gwas-2>step3.gwas<a hidden class=anchor aria-hidden=true href=#step3gwas-2>#</a></h1><p>/pub/software/emmax/emmax-intel64 -v -d 10 -t snp_filter -p trait.sort.txt -k pop.kinship -o emmax.out 1> emmax.log 2>emmax.err</p><h1 id=step4draw_manhattan-1>step4.draw_manhattan<a hidden class=anchor aria-hidden=true href=#step4draw_manhattan-1>#</a></h1><p>paste snp_filter.map emmax.out.ps | awk &lsquo;BEGIN{print &ldquo;SNP\tCHR\tBP\tP&rdquo;}{if($2==$5){print $2"\t"$1"\t"$4"\t"$NF}}&rsquo; > emmax.out.ps.manht_input
Rscript ../script/manhattan.R emmax.out.ps.manht_input emmax.out.ps.manht_figure_readData</p><p>QTL-seq
change_VCF2Table
1
2
3
4
5
ref=
vcf=
outfile=
gatk VariantsToTable -R $ref -V $vcf -F CHROM -F POS -F REF -F ALT -GF AD -GF DP -GF GQ -GF PL -O $outfile</p><p>QTLSeq_Analysis
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
library(QTLseqr)
library(&ldquo;ggplot2&rdquo;)</p><p>HighBulk &lt;- &ldquo;SRR834931&rdquo; ## é«˜å€¼æ··æ± åç§°
LowBulk &lt;- &ldquo;SRR834927&rdquo; ## ä½å€¼æ··æ± åç§°
Chroms &lt;- paste0(rep(&ldquo;Chr&rdquo;, 12), 1:12) ## æŸ“è‰²ä½“åˆ—è¡¨</p><h2 id=snpæ•°æ®è¯»å–>SNPæ•°æ®è¯»å–<a hidden class=anchor aria-hidden=true href=#snpæ•°æ®è¯»å–>#</a></h2><p>df &lt;- importFromGATK(file = &ldquo;../data/Yang_et_al_2013.table&rdquo; ,
highBulk = HighBulk, # æŒ‡å®šé«˜å€¼æ··æ± 
lowBulk = LowBulk, # æŒ‡å®šä½è‡³æ··æ± 
chromList = Chroms # æŒ‡å®šåˆ†æç”¨æŸ“è‰²ä½“åˆ—è¡¨
)</p><h2 id=ç»˜åˆ¶æ·±åº¦åˆ†å¸ƒå›¾>ç»˜åˆ¶æ·±åº¦åˆ†å¸ƒå›¾<a hidden class=anchor aria-hidden=true href=#ç»˜åˆ¶æ·±åº¦åˆ†å¸ƒå›¾>#</a></h2><p>p1 &lt;- ggplot(data = df) +
geom_histogram(aes(x = DP.HIGH + DP.LOW)) +
xlim(0,800)</p><p>pdf(file=&ldquo;SNP_depth.pdf&rdquo;)
p1
dev.off()</p><h2 id=ç»˜åˆ¶refç­‰ä½åŸºå› é¢‘ç‡åˆ†å¸ƒå›¾>ç»˜åˆ¶refç­‰ä½åŸºå› é¢‘ç‡åˆ†å¸ƒå›¾<a hidden class=anchor aria-hidden=true href=#ç»˜åˆ¶refç­‰ä½åŸºå› é¢‘ç‡åˆ†å¸ƒå›¾>#</a></h2><p>p2 &lt;- ggplot(data = df) +
geom_histogram(aes(x = REF_FRQ))</p><p>pdf(file=&ldquo;ref_allele_frequency.pdf&rdquo;)
p2
dev.off()</p><h2 id=ç»˜åˆ¶é«˜å€¼æ··æ± snp-indexåˆ†å¸ƒ>ç»˜åˆ¶é«˜å€¼æ··æ± SNP-indexåˆ†å¸ƒ<a hidden class=anchor aria-hidden=true href=#ç»˜åˆ¶é«˜å€¼æ··æ± snp-indexåˆ†å¸ƒ>#</a></h2><p>p3 &lt;- ggplot(data = df) +
geom_histogram(aes(x = SNPindex.HIGH))
pdf(file=&ldquo;SNPindex.HIGH.dis.pdf&rdquo;)
p3
dev.off()</p><h2 id=ç»˜åˆ¶ä½å€¼æ··æ± snp-indexåˆ†å¸ƒ>ç»˜åˆ¶ä½å€¼æ··æ± SNP-indexåˆ†å¸ƒ<a hidden class=anchor aria-hidden=true href=#ç»˜åˆ¶ä½å€¼æ··æ± snp-indexåˆ†å¸ƒ>#</a></h2><p>p4 &lt;- ggplot(data = df) +
geom_histogram(aes(x = SNPindex.LOW))</p><p>pdf(file=&ldquo;SNPindex.LOW.dis.pdf&rdquo;)
p4
dev.off()</p><h2 id=snpè¿‡æ»¤>SNPè¿‡æ»¤<a hidden class=anchor aria-hidden=true href=#snpè¿‡æ»¤>#</a></h2><p>df_filt &lt;- filterSNPs(
SNPset = df,
refAlleleFreq = 0.20, ## ref alleleé¢‘ç‡è¿‡æ»¤ï¼Œ0.2è¡¨ç¤º0.2~0.8ä¹‹é—´
minTotalDepth = 100, ## æœ€å°æ·±åº¦è¿‡æ»¤
maxTotalDepth = 400, ## æœ€å¤§æ·±åº¦è¿‡æ»¤
minSampleDepth = 40, ## å•ä¸ªæ ·å“æœ€å°æ·±åº¦
minGQ = 99, ## genotype quality è¿‡æ»¤
verbose = TRUE ## è¾“å‡ºæ—¥å¿—
)</p><h2 id=è¿›è¡Œdeltasnpindex-è®¡ç®—>è¿›è¡ŒdeltaSNPindex è®¡ç®—<a hidden class=anchor aria-hidden=true href=#è¿›è¡Œdeltasnpindex-è®¡ç®—>#</a></h2><p>df_filt &lt;- runQTLseqAnalysis(df_filt,
windowSize = 1e6, ## çª—å£å¤§å°
popStruc = &ldquo;F2&rdquo;, ## ç¾¤ä½“ç±»å‹ï¼ŒF2 æˆ–è€…RIL
bulkSize = c(385, 430), ## æ··æ ·ä¸ªæ•°ï¼Œç¬¬ä¸€ä¸ªé«˜å€¼ç»„
replications = 10000, ## bootstrapæ¬¡æ•°
intervals = c(95, 99) ## ç½®ä¿¡åŒºé—´
)</p><h2 id=snp-æ²¿æŸ“è‰²ä½“åˆ†å¸ƒå›¾>SNP æ²¿æŸ“è‰²ä½“åˆ†å¸ƒå›¾<a hidden class=anchor aria-hidden=true href=#snp-æ²¿æŸ“è‰²ä½“åˆ†å¸ƒå›¾>#</a></h2><p>p5 &lt;- plotQTLStats(SNPset = df_filt, var = &ldquo;nSNPs&rdquo;)
pdf(file=&ldquo;SNP_filter.window.pdf&rdquo;, width = 20, height = 4)
p5
dev.off()</p><h2 id=deltasnpindexæ²¿æŸ“è‰²ä½“åˆ†å¸ƒ>deltaSNPindexæ²¿æŸ“è‰²ä½“åˆ†å¸ƒ<a hidden class=anchor aria-hidden=true href=#deltasnpindexæ²¿æŸ“è‰²ä½“åˆ†å¸ƒ>#</a></h2><p>p6 &lt;- plotQTLStats(SNPset = df_filt, var = &ldquo;deltaSNP&rdquo;, plotIntervals = TRUE)
pdf(file=&ldquo;deltaSNPindex.pdf&rdquo;, width = 20, height = 4)
p6
dev.off()</p><h2 id=æå–æ˜¾è‘—æ€§åŒºåŸŸ>æå–æ˜¾è‘—æ€§åŒºåŸŸ<a hidden class=anchor aria-hidden=true href=#æå–æ˜¾è‘—æ€§åŒºåŸŸ>#</a></h2><p>QTL &lt;- getSigRegions(SNPset = df_filt,
method = &ldquo;QTLseq&rdquo;, interval = 95)</p><h2 id=è¾“å‡ºåˆ°æ–‡ä»¶>è¾“å‡ºåˆ°æ–‡ä»¶<a hidden class=anchor aria-hidden=true href=#è¾“å‡ºåˆ°æ–‡ä»¶>#</a></h2><p>write.table(QTL[[1]],
&ldquo;QTLseq_result.SigRegions.table&rdquo;,
sep="\t",quote=F)</p><h2 id=æå–qtlseqç»“æœ>æå–QTLseqç»“æœ<a hidden class=anchor aria-hidden=true href=#æå–qtlseqç»“æœ>#</a></h2><p>results99 &lt;- getQTLTable(
SNPset = df_filt,<br>method = &ldquo;QTLseq&rdquo;,
interval = 99, ## ç»“æœé˜ˆå€¼
export = FALSE)</p><h2 id=è¾“å‡ºåˆ°æ–‡ä»¶-1>è¾“å‡ºåˆ°æ–‡ä»¶<a hidden class=anchor aria-hidden=true href=#è¾“å‡ºåˆ°æ–‡ä»¶-1>#</a></h2><p>write.table(results99,
file=&ldquo;QTLseq_result_deltaSNPindex_CI99.table&rdquo; ,
sep="\t", quote = F)</p><p>ğŸ’Œlixiang117423@gmail.com</p><p>ğŸ’Œlixiang117423@foxmail.com</p><p>ç”Ÿç‰©ä¿¡æ¯å­¦ - ç¾¤ä½“é—ä¼ å­¦ - Linux
#ç”Ÿç‰©ä¿¡æ¯å­¦ #Linux</p></div><footer class=post-footer><ul class=post-tags></ul></footer></article></main><footer class=footer><span>&copy; 2026 <a href=https://Tdxxxx-student.github.io/>æˆ‘çš„åšå®¢</a></span> Â·
<span>Powered by
<a href=https://gohugo.io/ rel="noopener noreferrer" target=_blank>Hugo</a> &
<a href=https://github.com/adityatelange/hugo-PaperMod/ rel=noopener target=_blank>PaperMod</a></span></footer><a href=#top aria-label="go to top" title="Go to Top (Alt + G)" class=top-link id=top-link accesskey=g><svg viewBox="0 0 12 6" fill="currentColor"><path d="M12 6H0l6-6z"/></svg>
</a><script>let menu=document.getElementById("menu");if(menu){const e=localStorage.getItem("menu-scroll-position");e&&(menu.scrollLeft=parseInt(e,10)),menu.onscroll=function(){localStorage.setItem("menu-scroll-position",menu.scrollLeft)}}document.querySelectorAll('a[href^="#"]').forEach(e=>{e.addEventListener("click",function(e){e.preventDefault();var t=this.getAttribute("href").substr(1);window.matchMedia("(prefers-reduced-motion: reduce)").matches?document.querySelector(`[id='${decodeURIComponent(t)}']`).scrollIntoView():document.querySelector(`[id='${decodeURIComponent(t)}']`).scrollIntoView({behavior:"smooth"}),t==="top"?history.replaceState(null,null," "):history.pushState(null,null,`#${t}`)})})</script><script>var mybutton=document.getElementById("top-link");window.onscroll=function(){document.body.scrollTop>800||document.documentElement.scrollTop>800?(mybutton.style.visibility="visible",mybutton.style.opacity="1"):(mybutton.style.visibility="hidden",mybutton.style.opacity="0")}</script><script>document.getElementById("theme-toggle").addEventListener("click",()=>{const e=document.querySelector("html");e.dataset.theme==="dark"?(e.dataset.theme="light",localStorage.setItem("pref-theme","light")):(e.dataset.theme="dark",localStorage.setItem("pref-theme","dark"))})</script></body></html>