+++
date = '2026-02-03T12:08:55+08:00'
draft = false
title = 'äº‘å—å†œä¸šå¤§å­¦é‡æµ‹åºé¡¹ç›®æ•™ç¨‹'
categories = ["é‡æµ‹åºåˆ†æ"]
tags = ["ç¾¤ä½“ç»“æ„åˆ†æ","å˜å¼‚æ£€æµ‹","é‡æµ‹åº","GWAS","QTL-seq"]
 
+++
äº‘å—å†œä¸šå¤§å­¦ äº‘å—ç”Ÿç‰©èµ„æºä¿æŠ¤ä¸åˆ©ç”¨å›½å®¶é‡ç‚¹å®éªŒå®¤

æè¯¦ | 2019 å¹´ 12 æœˆ 31 æ—¥

## ç›®å½•

- [è½¯ä»¶å®‰è£…](#%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85)
- [å˜å¼‚æ£€æµ‹](#%E5%8F%98%E5%BC%82%E6%A3%80%E6%B5%8B)
- [ç¾¤ä½“ç»“æ„åˆ†æ](#%E7%BE%A4%E4%BD%93%E7%BB%93%E6%9E%84%E5%88%86%E6%9E%90)
- [é€‰æ‹©åˆ†æ](#%E9%80%89%E6%8B%A9%E5%88%86%E6%9E%90)
- [GWAS åˆ†æ](#gwas%E5%88%86%E6%9E%90)
- [QTL-seq åˆ†æ](#qtl-seq%E5%88%86%E6%9E%90)
- [é™„å½•ï¼šåˆ†ææµç¨‹å›¾](#%E9%99%84%E5%BD%95%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B%E5%9B%BE)

ï¼ˆå…¶ä¸­åŒ…å«äº†ä¸€äº›æˆ‘å¤ç”¨æ—¶å¯èƒ½é‡åˆ°çš„é—®é¢˜å’Œè§£å†³åŠæ³•ï¼Œæ¯”è¾ƒæ··ä¹±ï¼Œç”„åˆ«ä½¿ç”¨ï¼‰

## è½¯ä»¶å®‰è£…


> âš ï¸ æ³¨æ„ï¼šè¯¥éƒ¨åˆ†è½¯ä»¶å¾ˆéš¾ç”¨ conda ç›´æ¥å®‰è£…ï¼Œå®‰è£…æ­¥éª¤æ¯”è¾ƒç‰¹æ®Šã€‚

### 1. lumpy-sv

ğŸ’¡ lumpy-sv ä¾èµ– Python2.7ï¼Œéœ€åˆ‡æ¢åˆ° py27 ç¯å¢ƒä¸‹è¿›è¡Œå®‰è£…ã€‚

```bash
conda activate py27
git clone --recursive https://github.com/arq5x/lumpy-sv.git
cd lumpy-sv
make
```

### 2. svtyper

```bash
conda install svtyper
```

### 3. cnvnator

ğŸ’¡ cnvnator ä¾èµ–äº root è½¯ä»¶åŒ…åŠ samtools è½¯ä»¶åŒ…ï¼ˆåŒ…å« HTSlibï¼‰ï¼Œå› æ­¤å…ˆå®‰è£…ä¾èµ–ã€‚

#### 3.1 å®‰è£… samtools

```bash
wget https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2
tar -xvf samtools-1.9.tar.bz2
cd samtools-1.9/
./configure
make
```

#### 3.2 å®‰è£… root è½¯ä»¶

```bash
# ç›´æ¥ä¸‹è½½åè§£å‹å³å¯
wget https://root.cern/download/root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz
tar -zxvf root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz
```

#### 3.3 å®‰è£… cnvnator

```bash
git clone https://github.com/abyzovlab/CNVnator.git
cd CNVnator

# é“¾æ¥samtoolsè½¯ä»¶ç›®å½•åˆ°cnvnatorç›®å½•
ln -s ../samtools-1.9 ./samtools

# åŠ è½½rootè½¯ä»¶åˆ°ç¯å¢ƒå˜é‡
export ROOTSYS=/pub/software/root
export PATH=/pub/software/root/bin:$PATH
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ROOTSYS/lib

# å®‰è£…cnvnator
make LIBS="-lcrypto"
```



## å˜å¼‚æ£€æµ‹



### 1. æ„å»ºåŸºå› ç»„ index

```bash
# åˆ›å»ºè½¯é“¾æ¥
ln -s ../data/genome.fasta ./genome.fasta

# for gatk
samtools faidx genome.fasta

# for bwa
bwa index genome.fasta

# for picard
picard CreateSequenceDictionary R=genome.fasta
```

### 2. æ¯”å¯¹æ’åºå»é‡

#### 2.1 æ¯”å¯¹æ’åº

```bash
bwa mem -t 2 -R '@RG\tID:S1\tSM:S1\tPL:illumina' \
    ../01.ref/genome.fasta \
    ../data/S1_1.fq.gz ../data/S1_2.fq.gz | \
    /pub/software/samtools/samtools sort -@ 2 -m 1G -o S1.sort.bam -

bwa mem -t 2 -R '@RG\tID:S2\tSM:S2\tPL:illumina' \
    ../01.ref/genome.fasta \
    ../data/S2_1.fq.gz ../data/S2_2.fq.gz | \
    /pub/software/samtools/samtools sort -@ 2 -m 1G -o S2.sort.bam -
```

#### 2.2 å»é™¤é‡å¤

```bash
picard -Xmx4g MarkDuplicates \
    I=S1.sort.bam \
    O=S1.sort.rmdup.bam \
    CREATE_INDEX=true \
    REMOVE_DUPLICATES=true \
    M=S1.marked_dup_metrics.txt

picard -Xmx4g MarkDuplicates \
    I=S2.sort.bam \
    O=S2.sort.rmdup.bam \
    CREATE_INDEX=true \
    REMOVE_DUPLICATES=true \
    M=S2.marked_dup_metrics.txt
```

#### 2.3 æ¯”å¯¹ç‡ç»Ÿè®¡

```bash
/pub/software/samtools/samtools flagstat S1.sort.bam > S1.sort.bam.flagstat
/pub/software/samtools/samtools flagstat S2.sort.bam > S2.sort.bam.flagstat
```

#### 2.4 æ‰¹é‡æ¯”å¯¹è„šæœ¬ç”Ÿæˆ

```bash
less ../data/sample.list | awk '{ 
    print "bwa mem -t 2 -R '\''@RG\\tID:"$1"\\tSM:"$1"\\tPL:illumina'\'' \
    ../01.ref/genome.fasta ../data/"$1"_1.fq.gz ../data/"$1"_2.fq.gz | \
    /pub/software/samtools/samtools sort -@ 2 -m 1G -o "$1".sort.bam -" 
}'
```

### 3. SNPã€Indel æ£€æµ‹

#### 3.1 HaplotypeCaller

```bash
gatk --java-options "-Xmx10g -Djava.io.tmpdir=./tmp" HaplotypeCaller \
    -R ../01.ref/genome.fasta \
    -I ../02.mapping/S1.sort.rmdup.bam \
    -ERC GVCF \
    -O S1.g.vcf 1>S1.HC.log 2>&1

gatk --java-options "-Xmx10g -Djava.io.tmpdir=./tmp" HaplotypeCaller \
    -R ../01.ref/genome.fasta \
    -I ../02.mapping/S2.sort.rmdup.bam \
    -ERC GVCF \
    -O S2.g.vcf 1>S2.HC.log 2>&1
```

#### 3.2 CombineGVCFs

```bash
ls ./S*.g.vcf > gvcf.list

gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" CombineGVCFs \
    -R ../01.ref/genome.fasta \
    -V gvcf.list \
    -O all.merge.g.vcf
```

#### 3.3GenotpeGVCFs

```bash
gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" GenotypeGVCFs \
    -R ../01.ref/genome.fasta \
    --variant all.merge.g.vcf \
    -O all.merge_raw.vcf
```

#### 3.4 SNP è¿‡æ»¤

```bash
# æå–SNP
gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" SelectVariants \
    -R ../01.ref/genome.fasta \
    -V all.merge_raw.vcf \
    --select-type SNP \
    -O all.raw.snp.vcf

# æ ‡è®°è¿‡æ»¤
gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" VariantFiltration \
    -R ../01.ref/genome.fasta \
    -V all.raw.snp.vcf \
    --filter-expression "QD < 2.0 || MQ < 40.0 || FS > 60.0 || SOR > 3.0 || MQRankSum < -12.5 || ReadPosRankSum < -8.0" \
    --filter-name 'SNP_filter' \
    -O all.filter.snp.vcf

# æå–è¿‡æ»¤åçš„SNP
gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" SelectVariants \
    -R ../01.ref/genome.fasta \
    -V all.filter.snp.vcf \
    --exclude-filtered \
    -O all.filtered.snp.vcf
```

#### 3.5 InDel è¿‡æ»¤

```bash
# æå–InDel
gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" SelectVariants \
    -R ../01.ref/genome.fasta \
    -V all.merge_raw.vcf \
    --select-type INDEL \
    -O all.raw.indel.vcf

# æ ‡è®°è¿‡æ»¤
gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" VariantFiltration \
    -R ../01.ref/genome.fasta \
    -V all.raw.indel.vcf \
    --filter-expression "QD < 2.0 || FS > 200.0 || SOR > 10.0 || MQRankSum < -12.5 || ReadPosRankSum < -8.0" \
    --filter-name 'INDEL_filter' \
    -O all.filter.indel.vcf

# æå–è¿‡æ»¤åçš„InDel
gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" SelectVariants \
    -R ../01.ref/genome.fasta \
    -V all.filter.indel.vcf \
    --exclude-filtered \
    -O all.filtered.indel.vcf
```

### 4. SV æ£€æµ‹

```bash
# Step 1: æå– discordants reads
/pub/software/samtools/samtools view -b -F 1294 -@ 12 \
    ../02.mapping/S1.sort.rmdup.bam > S1.discordants.bam
/pub/software/samtools/samtools view -b -F 1294 -@ 12 \
    ../02.mapping/S2.sort.rmdup.bam > S2.discordants.bam

# Step 2: æå– split reads
/pub/software/samtools/samtools view -h ../02.mapping/S1.sort.rmdup.bam | \
    /pub/software/lumpy-sv/scripts/extractSplitReads_BwaMem -i stdin | \
    /pub/software/samtools/samtools sort - > S1.splitters.bam

/pub/software/samtools/samtools view -h ../02.mapping/S2.sort.rmdup.bam | \
    /pub/software/lumpy-sv/scripts/extractSplitReads_BwaMem -i stdin | \
    /pub/software/samtools/samtools sort - > S2.splitters.bam

# Step 3: è¿è¡Œ lumpy
/pub/software/lumpy-sv/bin/lumpyexpress \
    -B ../02.mapping/S1.sort.rmdup.bam,../02.mapping/S2.sort.rmdup.bam \
    -S S1.splitters.bam,S2.splitters.bam \
    -D S1.discordants.bam,S2.discordants.bam \
    -o all.sv.lumpy.vcf

# Step 4: ä¸ªä½“åŸºå› å‹åˆ†å‹
vcftools --vcf all.sv.lumpy.vcf --indv S1 --recode --recode-INFO-all --out S1 && \
    svtyper -i S1.recode.vcf -B ../02.mapping/S1.sort.rmdup.bam -o S1.genotype.vcf

vcftools --vcf all.sv.lumpy.vcf --indv S2 --recode --recode-INFO-all --out S2 && \
    svtyper -i S2.recode.vcf -B ../02.mapping/S2.sort.rmdup.bam -o S2.genotype.vcf
```

### 5. CNV æ£€æµ‹

#### 5.1 å‡†å¤‡åŸºå› ç»„

```bash
# å°†åŸºå› ç»„æ‹†åˆ†ä¸ºå•æ¡æŸ“è‰²ä½“
ln -s ../data/genome.fasta ./genome.fa
seqkit split -i ./genome.fa -O split

# é‡å‘½åæ–‡ä»¶
ls ./split/genome.id_*.fa | sed 's/./split/genome.id_//' | \
    awk '{print "mv ./split/genome.id_"$aa" ./split/"$aa}' > mv_name.sh
sh mv_name.sh
```

#### 5.2 è¿è¡Œ cnvnator

```bash
# è®¾ç½®ç¯å¢ƒå˜é‡
export ROOTSYS=/pub/software/root
export PATH=/pub/software/root/bin:$PATH
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ROOTSYS/lib

# é“¾æ¥æ–‡ä»¶
ln -s ../02.mapping/S1.sort.rmdup.bam

# æå–æ¯”å¯¹ç»“æœ
/pub/software/CNVnator/cnvnator -genome genome.fa -root S1.root -tree S1.sort.rmdup.bam

# ç”Ÿæˆæ·±åº¦åˆ†å¸ƒ
/pub/software/CNVnator/cnvnator -genome genome.fa -root S1.root -his 500 -d split

# è¿›è¡Œç»Ÿè®¡è®¡ç®—
/pub/software/CNVnator/cnvnator -root S1.root -stat 500

# æ£€æŸ¥bin sizeæ˜¯å¦åˆé€‚
/pub/software/CNVnator/cnvnator -root S1.root -eval 500 > S1.eval.ratio

# RDä¿¡å·åˆ†å‰²
/pub/software/CNVnator/cnvnator -root S1.root -partition 500

# è¿›è¡ŒCNVæ£€æµ‹
/pub/software/CNVnator/cnvnator -root S1.root -call 500 > S1.cnv
```

### 6. SNP/InDel æ³¨é‡Š

#### 6.1 å‡†å¤‡æ³¨é‡Šæ•°æ®åº“

```bash
ln -s ../data/genome.gtf
ln -s ../data/genome.fasta

gtfToGenePred -genePredExt genome.gtf genome_refGene.txt

/pub/software/annovar/retrieve_seq_from_fasta.pl \
    --format refGene \
    --seqfile genome.fasta \
    genome_refGene.txt \
    --out genome_refGeneMrna.fa
```

#### 6.2 æ³¨é‡Š VCF æ–‡ä»¶

```bash
ln -s ../03.SNP_indel/all.filtered.snp.vcf

# æ ¼å¼è½¬æ¢
/pub/software/annovar/convert2annovar.pl \
    -format vcf4old \
    all.filtered.snp.vcf > all.snp.vcf.annovar.input

# è¿›è¡Œæ³¨é‡Š
/pub/software/annovar/annotate_variation.pl \
    -geneanno \
    --neargene 2000 \
    -buildver genome \
    -dbtype refGene \
    -outfile all.anno \
    -exonsort \
    all.snp.vcf.annovar.input ./
```

### 7. CNV å’Œ SV æ³¨é‡Š

```bash
# å¤„ç†SVæ–‡ä»¶
less -S ../04.SV/all.sv.lumpy.vcf | \
    awk -F ";" '{ if($1!~/#/){print $1"\t"$2"\t"$3"\t"$4} }' | \
    awk '{print $1"\t"$2"\t"$11"\t"$3"\t"$8}' | \
    sed 's/END=//' | sed 's/SVTYPE=//' > SV.bed

# è¿‡æ»¤BNDç±»å‹
awk '$5 !~ /BND/' SV.bed > SV_filter.bed

# SVæ³¨é‡Š
bedtools intersect -wo -a SV_filter.bed -b gene.gtf > SV_filter.Anno

# å¤„ç†CNVæ–‡ä»¶
less -S ../05.CNV/S1.cnv | \
    awk '{print $2 "\t" $1}' | \
    sed 's/:/\t/' | sed 's/-/\t/' > CNV.bed

# CNVæ³¨é‡Š
bedtools intersect -wo -a CNV.bed -b gene.gtf > CNV.Anno
```

## ç¾¤ä½“ç»“æ„åˆ†æ

### 1. æ•°æ®è¿‡æ»¤

```bash
vcf=../data/all.vcf

# è¿‡æ»¤ç¼ºå¤±ç‡å’Œæœ€å°ç­‰ä½åŸºå› é¢‘ç‡
plink --vcf $vcf \
    --geno 0.1 \
    --maf 0.01 \
    --out all.missing_maf \
    --recode vcf-iid \
    --allow-extra-chr \
    --set-missing-var-ids @:# \
    --keep-allele-order

# è¿‡æ»¤è¿é”ä¸å¹³è¡¡
plink --vcf all.missing_maf.vcf \
    --indep-pairwise 50 10 0.2 \
    --out tmp.ld \
    --allow-extra-chr \
    --set-missing-var-ids @:#

plink --vcf all.missing_maf.vcf \
    --make-bed \
    --extract tmp.ld.prune.in \
    --out all.LDfilter \
    --recode vcf-iid \
    --keep-allele-order \
    --allow-extra-chr \
    --set-missing-var-ids @:#
```

### 2. å»ºæ ‘åˆ†æ

#### 2.1 NJ æ ‘

```bash
# ä¸‹è½½
git clone https://gitcode.com/gh_mirrors/vc/vcf2phylip
# 1 æ ¼å¼è½¬æ¢ï¼ˆnewï¼‰
/home2/xtc/pub/vcf2phylip/vcf2phylip.py -i all.LDfilter.vcf -o phylip_file
# 2 æˆ–ä½¿ç”¨seqtk
seqtk subseq reference.fasta region_list.txt > alignment.fasta
# åŸºæœ¬åˆ†æ
iqtree -s phylip_file -m GTR+G -bb 1000 -nt AUTO

# å‚æ•°è¯´æ˜ï¼š
# -s: è¾“å…¥æ–‡ä»¶ï¼ˆPHYLIPæˆ–FASTAæ ¼å¼ï¼‰
# -m: æ›¿ä»£æ¨¡å‹ï¼ˆGTR+Gå¸¸ç”¨ï¼Œæˆ–ä½¿ç”¨MFPè®©è½¯ä»¶è‡ªåŠ¨é€‰æ‹©ï¼‰
# -bb: è‡ªä¸¾é‡å¤æ¬¡æ•°ï¼ˆ1000ä¸ºæ ‡å‡†ï¼‰
# -nt: çº¿ç¨‹æ•°ï¼ˆAUTOè‡ªåŠ¨æ£€æµ‹ï¼‰

# è¾“å‡ºæ–‡ä»¶ï¼š
# - *.treefile: æœ€ä½³ç³»ç»Ÿå‘è‚²æ ‘ï¼ˆNewickæ ¼å¼ï¼‰
# - *.log: è¿è¡Œæ—¥å¿—
# - *.iqtree: è¯¦ç»†ç»“æœ
# 3 æ ¼å¼è½¬æ¢ï¼ˆæ›´æ–°ç”¨pyè„šæœ¬ï¼‰

#-----------------------------------------------------------------------------
# æ ¼å¼è½¬æ¢
perl ../../script/vcf2phy.pl ../../00.filter/all.LDfilter.vcf > sequences.dat

# è®¡ç®—é—ä¼ è·ç¦»
echo -e "sequences.dat\nY" > dnadist.cfg
dnadist < dnadist.cfg > dnadist.log
mv outfile infile.dist

# æ„å»ºNJæ ‘
echo -e "infile.dist\nY" > neighbor.cfg
neighbor < neighbor.cfg > nj.log

# æ ¼å¼æ•´ç†
less infile.dist | tr '\n' '|' | sed 's/| / /g' | tr '|' '\n' > infile.dist.table
less outtree | tr '\n' ' ' | sed 's/ //g' > outtree.nwk
```

#### 2.2 SNPhylo å»ºæ ‘ï¼ˆæ…¢ï¼‰

```bash
/pub/software/SNPhylo/snphylo.sh \
    -v ../../00.filter/all.LDfilter.vcf \
    -r \
    -l 1 \
    -m 0.01 \
    -o GA0001 \
    -b \
    -B 2 \
    -P tree_snphylo
```
treefileæ–‡ä»¶è®©å¦‚ITOLç”»è¿›åŒ–æ ‘ï¼ŒåŒæ—¶æ ¹æ®éœ€æ±‚è¿›è¡Œç¾åŒ–ï¼Œä¸Šä¼ æ³¨é‡Šæ–‡ä»¶ã€‚
```text
# itol_colorstrip.txt ,ç»™ä¸åŒç§ç¾¤å¤–åœˆæ·»åŠ é¢œè‰²åŒºåˆ†
DATASET_COLORSTRIP
SEPARATOR TAB
DATASET_LABEL	population
COLOR	#000000
STRIP_WIDTH	50

LEGEND_TITLE	Population
LEGEND_SHAPES	1	1
LEGEND_COLORS	#377EB8	#E41A1C
LEGEND_LABELS	Cultivated	Wild

DATA
C01	#377EB8	Cultivated
C02	#377EB8	Cultivated
C08	#377EB8	Cultivated
C12	#377EB8	Cultivated
C14	#377EB8	Cultivated
C16	#377EB8	Cultivated
C17	#377EB8	Cultivated
C19	#377EB8	Cultivated
C24	#377EB8	Cultivated
C27	#377EB8	Cultivated
C30	#377EB8	Cultivated
C33	#377EB8	Cultivated
C34	#377EB8	Cultivated
C35	#377EB8	Cultivated
W01	#E41A1C	Wild
W02	#E41A1C	Wild
W03	#E41A1C	Wild
W04	#E41A1C	Wild
W05	#E41A1C	Wild
W06	#E41A1C	Wild
W07	#E41A1C	Wild
W08	#E41A1C	Wild
W09	#E41A1C	Wild
W10	#E41A1C	Wild
W11	#E41A1C	Wild
W12	#E41A1C	Wild
W13	#E41A1C	Wild
W14	#E41A1C	Wild
W15	#E41A1C	Wild
W16	#E41A1C	Wild
W17	#E41A1C	Wild
```

```
itol_branch_colors.txt ,åˆ†æ”¯æ ‡ç­¾é¢œè‰²
TREE_COLORS
SEPARATOR TAB
DATA
C01	branch	#377EB8	normal	2
C01	label	#377EB8	normal	1
C02	branch	#377EB8	normal	2
C02	label	#377EB8	normal	1
C08	branch	#377EB8	normal	2
C08	label	#377EB8	normal	1
C12	branch	#377EB8	normal	2
C12	label	#377EB8	normal	1
C14	branch	#377EB8	normal	2
C14	label	#377EB8	normal	1
C16	branch	#377EB8	normal	2
C16	label	#377EB8	normal	1
C17	branch	#377EB8	normal	2
C17	label	#377EB8	normal	1
C19	branch	#377EB8	normal	2
C19	label	#377EB8	normal	1
C24	branch	#377EB8	normal	2
C24	label	#377EB8	normal	1
C27	branch	#377EB8	normal	2
C27	label	#377EB8	normal	1
C30	branch	#377EB8	normal	2
C30	label	#377EB8	normal	1
C33	branch	#377EB8	normal	2
C33	label	#377EB8	normal	1
C34	branch	#377EB8	normal	2
C34	label	#377EB8	normal	1
C35	branch	#377EB8	normal	2
C35	label	#377EB8	normal	1
W01	branch	#E41A1C	normal	2
W01	label	#E41A1C	normal	1
W02	branch	#E41A1C	normal	2
W02	label	#E41A1C	normal	1
W03	branch	#E41A1C	normal	2
W03	label	#E41A1C	normal	1
W04	branch	#E41A1C	normal	2
W04	label	#E41A1C	normal	1
W05	branch	#E41A1C	normal	2
W05	label	#E41A1C	normal	1
W06	branch	#E41A1C	normal	2
W06	label	#E41A1C	normal	1
W07	branch	#E41A1C	normal	2
W07	label	#E41A1C	normal	1
W08	branch	#E41A1C	normal	2
W08	label	#E41A1C	normal	1
W09	branch	#E41A1C	normal	2
W09	label	#E41A1C	normal	1
W10	branch	#E41A1C	normal	2
W10	label	#E41A1C	normal	1
W11	branch	#E41A1C	normal	2
W11	label	#E41A1C	normal	1
W12	branch	#E41A1C	normal	2
W12	label	#E41A1C	normal	1
W13	branch	#E41A1C	normal	2
W13	label	#E41A1C	normal	1
W14	branch	#E41A1C	normal	2
W14	label	#E41A1C	normal	1
W15	branch	#E41A1C	normal	2
W15	label	#E41A1C	normal	1
W16	branch	#E41A1C	normal	2
W16	label	#E41A1C	normal	1
W17	branch	#E41A1C	normal	2
W17	label	#E41A1C	normal	1
```


### 3. PCA åˆ†æ

```bash
plink --vcf ../00.filter/all.LDfilter.vcf \
    --pca 10 \
    --out PCA_out \
    --allow-extra-chr \
    --set-missing-var-ids @:#
# Rç»˜å›¾
Rscript ../script/draw_PCA.R PCA_out.eigenvec 1 2 ../data/sample.pop PCA_out_figure
```

### 4. ç¾¤ä½“ç»“æ„åˆ†æ

```bash
# Step 0: vcfè½¬bedæ ¼å¼
plink --vcf ../00.filter/all.LDfilter.vcf \
    --make-bed \
    --out all \
    --allow-extra-chr \
    --keep-allele-order \
    --set-missing-var-ids @:#

# è½¬æ¢ä¸ºadmixtureæ ¼å¼ï¼ˆped,ä¸ä¼šå‡ºç°æŸ“è‰²ä½“å‘½åæŠ¥é”™ä¸­æ­¢ï¼‰
plink --vcf  all.LDfilter.vcf --recode 12 --out all.LDfilter.prune --allow-extra-chr

# Step 1: ç”ŸæˆAdmixtureè¿è¡Œè„šæœ¬
seq 2 4 | awk '{print "admixture --cv -j2 all.bed "$1" 1>admix."$1".log 2>&1"}' > admixture.sh

# pedç‰ˆæœ¬
seq 2 10 | awk '{print "admixture --cv -j4 all.LDfilter.prune.ped "$1" 1>admix."$1".log 2>&1"}' > admixture.sh
nohup parallel -j 10 < admixture.sh > admixture.log 2>&1 &
grep -h CV *.log | sort -t':' -k2 -n
# Step 2: è¿è¡ŒAdmixture
sh admixture.sh

# ä¹Ÿå¯å•ç‹¬è¿è¡Œï¼ˆè¦æ±‚æ–‡ä»¶ä¸­æŸ“è‰²ä½“å‘½ååªå«æ•°å­—ï¼‰
admixture --cv -j2 all.bed 2 1>admix.2.log 2>&1
admixture --cv -j2 all.bed 3 1>admix.3.log 2>&1
admixture --cv -j2 all.bed 4 1>admix.4.log 2>&1

# S2.1 é€‰æ‹©æœ€ä¼˜åˆ†ç¾¤
# æŸ¥çœ‹äº¤å‰éªŒè¯è¯¯å·®
grep -h CV log*.out | sort -t':' -k2 -n
# é€‰æ‹©CVè¯¯å·®æœ€å°çš„Kå€¼ï¼Œç»˜å›¾ä»£ç å¦ä½œæ•´ç†

# Step 3: ç»˜åˆ¶ç»“æ„å›¾
mkdir result
cp ./*.Q result/
Rscript ../script/draw_admixture.R result all.nosex structure
```
```r 
# admixtureåˆ†ç¾¤ç»˜å›¾
library(ggsci)
library(tidyverse)
library(ggthemes)
library(pophelper)
files <- paste0("all.LDfilter.prune.",2:10,".Q")
ID <- read.table("sample.list",header = F)
slist <- readQ(files = files)
for(i in 1:length(slist)){
  row.names(slist[[i]]) <- ID[,1]
} 
P1 <- plotQ(slist, imgoutput = "join",imgtype = "pdf",width = 20,height = 1.2,
            showindlab = T,useindlab = T,indlabhjust = 0.5,indlabangle = 90,
            indlabsize = 0.6,sortind = "all",sharedindlab = F,
            panelspacer = 0.02, indlabspacer = 0,indlabcol = "black",
            showtitle = T,titlelab = "Population Structure",titlecol = "black",
            titlehjust = 0.5,showsp = T,exportpath=getwd(),
            sppos = "left",splab = paste0("K=",2:10),splabangle = 180,dpi = 600,
            outputfilename = paste0("admixture_barplot"),units = "cm")


# å¼•å…¥åˆ†ç»„ä¿¡æ¯
group <- read.table("sample.list",header = F,row.names = 1) #ç¬¬ä¸€åˆ—ä¸ºæ ·å“åï¼Œç¬¬äºŒåˆ—ä¸ºç§ç¾¤åˆ†ç±»
P2 <- plotQ(slist,imgoutput = "join",
            imgtype = "png",  #è¾“å‡ºæ ¼å¼çš„å˜åŒ–
            height = 1.2,width = 20,
            showindlab = T,useindlab = T,sharedindlab = T,
            showtitle = T,titlelab = "Population Structure",titlecol = "black",
            showsp = T,sppos = "left",splab = paste0("K=",2:10),splabangle = 180,
            showlegend = F,ordergrp = T,grplab = group,   #æ·»åŠ åˆ†ç»„
            subsetgrp = LETTERS[1:31],  #æ·»åŠ ç»„å
            dpi = 600,
            outputfilename = paste0("structure_barplot_pop"),units = "cm",
            exportpath = getwd())

```

### 5. LD è¡°å‡åˆ†æï¼ˆä¸è¦åšLDä¿®å‰ªçš„vcfè¿›è¡Œè®¡ç®—ï¼‰

```bash
# Step 1: è®¡ç®—LD(å¯ä¿®æ”¹binå‚æ•°è°ƒæ•´æ›²çº¿å…‰æ»‘ç¨‹åº¦ï¼Œå¯ç»˜åˆ¶æ€»ç¾¤ä½“æˆ–è€…å•ä¸ªç¾¤ä½“)
# ä»¥ä¸‹ä¸ºè®¡ç®—å•ä¸ªç¾¤ä½“çš„LDè¡°å‡è·ç¦»ï¼Œäºšç¾¤æ–‡ä»¶ä¸ºä¸€åˆ—è¯¥ç¾¤çš„æ ·å“åã€‚e.g W01 W02ã€‚ã€‚ã€‚
/pub/software/PopLDdecay/bin/PopLDdecay \
    -InVCF ../data/all.vcf \
    -SubPop ../data/pop.SC.table \
    -MaxDist 500 \
    -OutStat pop.SC.stat

/pub/software/PopLDdecay/bin/PopLDdecay \
    -InVCF ../data/all.vcf \
    -SubPop ../data/pop.YZR.table \
    -MaxDist 500 \
    -OutStat pop.YZR.stat

# Step 2: ç»˜åˆ¶å•ç¾¤ä½“LDè¡°å‡å›¾
perl /pub/software/PopLDdecay/bin/Plot_OnePop.pl \
    -inFile pop.SC.stat.gz \
    -output pop.SC.ld

perl /pub/software/PopLDdecay/bin/Plot_OnePop.pl \
    -inFile pop.YZR.stat.gz \
    -output pop.YZR.ld

# Step 3: ç»˜åˆ¶å¤šç¾¤ä½“LDè¡°å‡å›¾
perl /pub/software/PopLDdecay/bin/Plot_MultiPop.pl \
    -inList ld_stat.list \
    -output ld_stat.multi
```

## é€‰æ‹©åˆ†æ

### 1. Ï€ å’Œ ROD åˆ†æ

```bash
vcf=../../data/all.vcf
window=20000
step=2000

# Step 1: è®¡ç®—piå€¼
vcftools --vcf $vcf \
    --window-pi $window \
    --window-pi-step $step \
    --keep ../../data/pop.SC.table \
    --out ./Pi.pop1

vcftools --vcf $vcf \
    --window-pi $window \
    --window-pi-step $step \
    --keep ../../data/pop.YZR.table \
    --out ./Pi.pop2

# Step 2: è®¡ç®—RODï¼ˆæ²¡æœ‰æä¾›è¯¥è„šæœ¬ï¼Œä¸‹é¢Rè„šæœ¬å·²ç»è®¡ç®—ï¼Œéœ€è¦ä¿å­˜ï¼‰
pi1=Pi.pop1.windowed.pi
pi2=Pi.pop2.windowed.pi
perl ../../script/merge2pi_ROD.pl $pi1 $pi2 > Pi.ROD
```
#### (è‡ªç”¨RODåˆ†æè„šæœ¬ï¼Œç½‘ä¸Šæ•°ä¿¡é™¢æä¾›ï¼‰
```R
setwd("~/workspace/soybean.project/select.place/")
library(ggplot2)
library(tidyr)
library(dplyr)
library(CMplot)

qtile <- 0.95  # åˆ†ä½æ•°é˜ˆå€¼ï¼Œä¾‹å¦‚0.95è¡¨ç¤º95%åˆ†ä½æ•°
prefix <- "outsample"# è¾“å‡ºæ–‡ä»¶å‰ç¼€

# è¯»å–æ•°æ®ï¼ˆæ›¿æ¢ä¸ºå®é™…æ–‡ä»¶è·¯å¾„ï¼‰
wd <- read.table("Pi.pop2_W.windowed.pi", header = TRUE)  # subpop1çš„Piæ•°æ®
cd <- read.table("Pi.pop1_C.windowed.pi", header = TRUE)  # subpop2çš„Piæ•°æ®

# æ•°æ®å¤„ç†ï¼šåˆå¹¶IDå¹¶è¿›è¡Œå†…è¿æ¥
# åŸå§‹æ•°æ®åˆ—ååº”ä¸ºï¼šCHROM   BIN_START BIN_END N_VARIANTS      PI
wd1 <- unite(wd, ID, CHROM, BIN_START, BIN_END, sep = ":", remove = TRUE)
cd1 <- unite(cd, ID, CHROM, BIN_START, BIN_END, sep = ":", remove = TRUE)
wild_cult <- inner_join(wd1, cd1, by = "ID")

# è®¡ç®—RODå’ŒPiæ¯”ç‡
wild_cult <- wild_cult %>%
  mutate(
    rod = 1 - PI.y / PI.x,  # ROD = 1 - (subpop2çš„Pi / subpop1çš„Pi)
    pi_ratio1 = PI.x / PI.y,  # subpop1çš„Pi / subpop2çš„Pi
    pi_ratio2 = PI.y / PI.x   # subpop2çš„Pi / subpop1çš„Pi
  ) %>%
  separate(ID, c("chr", "start", "end"), sep = ":") %>%
  unite(ID, chr, start, sep = ":", remove = FALSE)

# RODç»˜å›¾
{
  # è®¡ç®—RODçš„é˜ˆå€¼ï¼ˆæŒ‡å®šåˆ†ä½æ•°ï¼‰
  rod_cut <- quantile(wild_cult$rod, qtile, na.rm = TRUE)
  
  # ç­›é€‰ROD>0çš„æ•°æ®å¹¶ç»˜å›¾
  rod_data <- select(wild_cult, ID, chr, start, rod) %>% filter(rod > 0)
  
  # ä¿å­˜ä¸ºPDF
  pdf(paste(prefix, "ROD.pdf", sep = "."), width = 10, height = 3)
  CMplot(
    rod_data,
    type = "p",
    plot.type = "m",
    LOG10 = FALSE,
    col = c("blue4", "orange3"),
    cex = 0.2,
    band = 0.5,
    ylab.pos = 2.5,
    cex.axis = 0.8, # è¯¥å‚æ•°ç‰ˆæœ¬ä¸å¯¹åˆ™ä½¿ç”¨ä¸äº†ï¼Œä¼šæŠ¥é”™
    threshold = rod_cut,
    amplify = FALSE,
    ylab = expr(paste("ROD ( 1-", pi, "_subpop2/", pi, "_subpop1 )")),
    file.output = FALSE
  )
  dev.off()
  
  # ä¿å­˜ä¸ºPNG
  png(paste(prefix, "ROD.png", sep = "."), width = 1000, height = 300)
  CMplot(
    rod_data,
    type = "p",
    plot.type = "m",
    LOG10 = FALSE,
    col = c("blue4", "orange3"),
    cex = 0.2,
    band = 0.5,
    ylab.pos = 2.5,
    cex.axis = 0.8, # è¯¥å‚æ•°ç‰ˆæœ¬ä¸å¯¹åˆ™ä½¿ç”¨ä¸äº†ï¼Œä¼šæŠ¥é”™
    threshold = rod_cut,
    amplify = FALSE,
    ylab = expr(paste("ROD ( 1-", pi, "_subpop2/", pi, "_subpop1 )")),
    file.output = FALSE
  )
  dev.off()
  }

# Pi_ratio1ç»˜å›¾ï¼ˆsubpop1çš„Pi / subpop2çš„Piï¼‰
{
  ratio1_cut <- quantile(wild_cult$pi_ratio1, qtile, na.rm = TRUE)
  ratio1_data <- select(wild_cult, ID, chr, start, pi_ratio1)
  
  # PDFæ ¼å¼
  pdf(paste(prefix, "PIratio1.pdf", sep = "."), width = 10, height = 3)
  CMplot(
    ratio1_data,
    type = "p",
    plot.type = "m",
    LOG10 = FALSE,
    col = c("blue4", "orange3"),
    cex = 0.2,
    band = 0.5,
    ylab.pos = 2.5,
    cex.axis = 0.8,
    threshold = ratio1_cut,
    amplify = FALSE,
    ylab = expr(paste(pi, "_subpop1/", pi, "_subpop2")),
    file.output = FALSE
  )
  dev.off()
  
  # PNGæ ¼å¼
  png(paste(prefix, "PIratio1.png", sep = "."), width = 1000, height = 300)
  CMplot(
    ratio1_data,
    type = "p",
    plot.type = "m",
    LOG10 = FALSE,
    col = c("blue4", "orange3"),
    cex = 0.2,
    band = 0.5,
    ylab.pos = 2.5,
    cex.axis = 0.8,
    threshold = ratio1_cut,
    amplify = FALSE,
    ylab = expr(paste(pi, "_subpop1/", pi, "_subpop2")),
    file.output = FALSE
  )
  dev.off()
}

# Pi_ratio2ç»˜å›¾ï¼ˆsubpop2çš„Pi / subpop1çš„Piï¼‰
{
  ratio2_cut <- quantile(wild_cult$pi_ratio2, qtile, na.rm = TRUE)
  ratio2_data <- select(wild_cult, ID, chr, start, pi_ratio2)
  
  # PDFæ ¼å¼
  pdf(paste(prefix, "PIratio2.pdf", sep = "."), width = 10, height = 3)
  CMplot(
    ratio2_data,
    type = "p",
    plot.type = "m",
    LOG10 = FALSE,
    col = c("blue4", "orange3"),
    cex = 0.2,
    band = 0.5,
    ylab.pos = 2.5,
    cex.axis = 0.8,
    threshold = ratio2_cut,
    amplify = FALSE,
    ylab = expr(paste(pi, "_subpop2/", pi, "_subpop1")),
    file.output = FALSE
  )
  dev.off()
  
  # PNGæ ¼å¼
  png(paste(prefix, "PIratio2.png", sep = "."), width = 1000, height = 300)
  CMplot(
    ratio2_data,
    type = "p",
    plot.type = "m",
    LOG10 = FALSE,
    col = c("blue4", "orange3"),
    cex = 0.2,
    band = 0.5,
    ylab.pos = 2.5,
    cex.axis = 0.8,
    threshold = ratio2_cut,
    amplify = FALSE,
    ylab = expr(paste(pi, "_subpop2/", pi, "_subpop1")),
    file.output = FALSE
  )
  dev.off()
}

# è¾“å‡ºæ±‡æ€»è¡¨æ ¼
outtable <- select(wild_cult, ID, chr, start, end, PI.x, PI.y, rod, pi_ratio1, pi_ratio2)
outfile <- paste(prefix, "ROD_PiRatio.table", sep = ".")
write.table(outtable, file = outfile, sep = "\t", quote = FALSE, row.names = FALSE)

# è¾“å‡ºç»Ÿè®¡é˜ˆå€¼
cutfile <- paste(prefix, "ROD_PiRatio.cutoff", sep = ".")
outcut <- data.frame(
  Iterm = c("ROD(1-subpop2/subpop1)", "Piratio1(subpop1/subpop2)", "Piratio2(subpop2/subpop1)"),
  cutoff = c(rod_cut, ratio1_cut, ratio2_cut)
)
write.table(outcut, file = cutfile, sep = "\t", quote = FALSE, row.names = FALSE)
# ========== å•ç‹¬ä¿å­˜ ROD å€™é€‰åŒºåŸŸ ==========
rod_cut <- quantile(wild_cult$rod, qtile, na.rm = TRUE)

# ç­›é€‰ ROD è¶…è¿‡é˜ˆå€¼çš„åŒºåŸŸ
rod_candidates <- wild_cult %>%
  filter(rod > rod_cut) %>%
  select(chr, start, end, PI.x, PI.y, rod) %>%
  arrange(desc(rod))

# é‡å‘½ååˆ—
colnames(rod_candidates) <- c("Chr", "Start", "End", "Pi_wild", "Pi_cultivated", "ROD")

# ä¿å­˜
write.table(rod_candidates, 
            paste0(prefix, ".ROD_candidates.txt"),
            row.names = FALSE, quote = FALSE, sep = "\t")
```
### 2. Tajima's D åˆ†æ

```bash
vcf=../../data/all.vcf
window=20000

vcftools --vcf $vcf \
    --TajimaD $window \
    --keep ../../data/pop.SC.table \
    --out TajimaD.pop1

vcftools --vcf $vcf \
    --TajimaD $window \
    --keep ../../data/pop.YZR.table \
    --out TajimaD.pop2
```

### 3. Fst åˆ†æï¼ˆè‹¥è¾“å…¥vcf.gz,å‚æ•°ä¿®æ”¹ä¸º--gzvcfï¼‰

```bash
vcf=../../data/all.vcf
window=20000
step=2000

vcftools --vcf $vcf \
    --fst-window-size $window \
    --fst-window-step $step \
    --weir-fst-pop ../../data/pop.SC.table \
    --weir-fst-pop ../../data/pop.YZR.table \
    --out ./Fst.pop1.pop2
```
```r
library(ggplot2)
# ä½¿ç”¨çª—å£æ–‡ä»¶ç»˜å›¾
fst_data <- read.table("../select.place/Fst.popC.popW.windowed.weir.fst", header=TRUE)
head(fst_data)
# CHROM  BIN_START  BIN_END  N_VARIANTS  WEIGHTED_FST  MEAN_FST

ggplot(fst_data, aes(x=BIN_START/1000000, y=WEIGHTED_FST)) +
  geom_point(size=0.5, alpha=0.6, color="steelblue") +
  geom_hline(yintercept=0.15, linetype="dashed", color="red") +
  facet_wrap(~CHROM, scales="free_x", ncol=4) +
  labs(x="Position (Mb)", y="FST",
       title="Genome-wide FST (Cultivated vs Wild)") +
  theme_bw() +
  theme(strip.text = element_text(size=8))

ggsave("FST_plot.pdf", width=12, height=8)
ggsave("FST_plot.png", width=12, height=8, dpi=300)
```
### 4. ROD + Fst è”åˆåˆ†æ

```bash
fst=../03.Fst/Fst.pop1.pop2.windowed.weir.fst
ROD=../01.pi_ROD/Pi.ROD
gff=../../data/genome.gff

# æå–æ˜¾è‘—ä½ç‚¹
perl ../../script/top_Fst_ROD_S.pl $fst $ROD 0.05 0.05 Fst_ROD.table Fst_ROD.stat

# ç­›é€‰topä½ç‚¹
awk '$NF=="top"' Fst_ROD.table > Fst_ROD.table.top

# æå–åŸºå› æ³¨é‡Š
awk '$3=="gene"' $gff > gene.gff

# æ³¨é‡Šå€™é€‰åŸºå› 
bedtools intersect -wo -F 0.1 -a Fst_ROD.table.top -b gene.gff | \
    awk -F "\t" '{print $7"\t"$10"\t"$11"\t"$13"\t"$15}' | \
    sort -u > Fst_ROD.table.top.gene
```
### 5. Fst+tajimaD+Pi
```bash
# å‡†å¤‡Piã€TajimaD ã€Fst åˆ†æçš„ç»“æœæ–‡ä»¶
cat all.list
Fst Fst.subpop.windowed.weir.fst
Pi_subpop1   subpop1.windowed.pi
Pi_subpop2 subpop2.windowed.pi
TajimaD_sub1     TajimaD.subpop1.Tajima.D
TajimaD_sub2     TajimaD.subpop2.Tajima.D
```

```r
# ç»˜å›¾
# åŠ è½½æ‰€éœ€åº“
library(ggplot2)
library(tidyr)
library(dplyr)

# å®šä¹‰å‚æ•°
input <- "all.list"# æ›¿æ¢ä¸ºä½ çš„file.listæ–‡ä»¶è·¯å¾„
chrname <- "Chr01"    # æ›¿æ¢ä¸ºä½ è¦åˆ†æçš„æŸ“è‰²ä½“åç§°
p_start <- 2000000    # æ›¿æ¢ä¸ºèµ·å§‹ä½ç½®
p_end <- 4000000      # æ›¿æ¢ä¸ºç»“æŸä½ç½®
prefix <- "output"    # æ›¿æ¢ä¸ºè¾“å‡ºæ–‡ä»¶å‰ç¼€

# è¯»å–æ–‡ä»¶åˆ—è¡¨
flist <- read.table(input, header = FALSE)

# åˆå§‹åŒ–æ•°æ®æ¡†
alldata <- data.frame()

# å¾ªç¯è¯»å–å¹¶å¤„ç†æ¯ä¸ªæ–‡ä»¶
for(i in 1:nrow(flist)){
type <- flist[i, 1]
  data <- read.table(flist[i, 2], header = TRUE)


# æå–ç¬¬ä¸€åˆ—æŸ“è‰²ä½“ï¼Œç¬¬äºŒåˆ—èµ·å§‹ä½ç½®ï¼Œæœ€åä¸€åˆ—å€¼
  data_f <- data[, c(1, 2, ncol(data))] 

# ä¿®æ”¹åˆ—å
  colnames(data_f) <- c("chr", "start", "val")

# æŒ‰ä½ç½®ä¿¡æ¯è¿‡æ»¤æ•°æ®
  data_f <- data_f %>% 
    dplyr::filter(chr == chrname & start > p_start & start < p_end) 

# æ·»åŠ æ•°æ®æ¥æº
  data_f <- mutate(data_f, type = rep(type, nrow(data_f)))

# åˆå¹¶æˆä¸€ä¸ªè¡¨æ ¼
  alldata <- rbind(alldata, data_f)
}

# æ•°æ®å¤„ç†
alldata_f <- tidyr::separate(alldata, type, c("TP", "pop"), sep = "_") 
alldata_f$val[which(alldata_f$TP == "Fst" & alldata_f$val < 0)] <- 0 

# åˆ›å»ºç»˜å›¾
p <- ggplot(alldata_f, aes(x = start/1000000, y = val, color = pop)) +
  geom_point(cex = 1) +
  geom_smooth(method = "loess", span = 0.25, cex = 0.5, se = FALSE) +
  scale_x_continuous(expand = c(0, 0)) +
  theme_bw() +  # è®¾ç½®ä¸»é¢˜
  facet_wrap(~TP,
             scales = "free_y",
             ncol = 1,
             strip.position = "left") +
  theme(
    strip.background = element_blank(),  # å»é™¤æ ‡é¢˜èƒŒæ™¯
    strip.text.x = element_blank(),      # å»é™¤è½´title
    strip.placement = "outside",
    axis.title.x = element_blank(),      # å»é™¤æ¨ªè½´title
    axis.title.y = element_blank(),      # å»é™¤yè½´title
    panel.spacing.y = unit(0, "lines"),  # å»é™¤é—´è·
    legend.title = element_blank(),      # å»é™¤legend title
    panel.grid.major = element_blank(),  # å»é™¤ç½‘æ ¼
    panel.grid.minor = element_blank()   # å»é™¤ç½‘æ ¼
  )

print(p)

# ä¿å­˜PDFæ–‡ä»¶
pdf(paste(prefix, "combine_figure.pdf", sep = "."), width = 5, height = 4)
print(p)
dev.off()
```
### 6 Fst+Rod (å°è¯•å€ŸåŠ©aiå¼¥è¡¥ä¸Šå…å¾—Perlè„šæœ¬)
```perl
#!/usr/bin/perl
use strict;
use warnings;

# æ£€æŸ¥è¾“å…¥å‚æ•°
die "Usage: $0 <Fst_file> <ROD_file> <Fst_top_percent> <ROD_top_percent> <out_table> <out_stat>\n" unless @ARGV == 6;

# è§£æå‚æ•°
my ($fst_file, $rod_file, $fst_pct, $rod_pct, $out_table, $out_stat) = @ARGV;

# -------------------------- å¤„ç†Fstæ–‡ä»¶ --------------------------
my (%fst_data, @fst_values);
open my $fst_fh, '<', $fst_file or die "æ— æ³•æ‰“å¼€Fstæ–‡ä»¶: $!";
my $fst_header = <$fst_fh>; # è·³è¿‡è¡¨å¤´
while (my $line = <$fst_fh>) {
    chomp $line;
    my @cols = split /\t/, $line;
    my ($chr, $start, $end, $fst) = @cols[0,1,2,5]; # Fståœ¨weir.fstç¬¬6åˆ—ï¼ˆç´¢å¼•5ï¼‰
    
    # è¿‡æ»¤æ— æ•ˆå€¼
    next if $fst eq 'NA' || $fst =~ /NaN/ || $fst == -1 || $fst > 1;
    
    # ç»Ÿä¸€æŸ“è‰²ä½“å‘½åï¼ˆåŒ¹é…RODæ–‡ä»¶çš„chræ ¼å¼ï¼šglyma.Wm82.gnm6.Gm01ï¼‰
    $fst_data{"$chr:$start:$end"} = $fst;
    push @fst_values, $fst;
}
close $fst_fh;

# è®¡ç®—Fst topé˜ˆå€¼ï¼ˆé™åºæ’åºï¼‰
@fst_values = sort {$b <=> $a} @fst_values;
my $fst_top_idx = int(scalar(@fst_values) * $fst_pct);
my $fst_cutoff = $fst_values[$fst_top_idx] // 0;

# -------------------------- å¤„ç†RODæ–‡ä»¶ï¼ˆé€‚é…ä½ çš„æ ¼å¼ï¼‰ --------------------------
my (%rod_data, @rod_values);
open my $rod_fh, '<', $rod_file or die "æ— æ³•æ‰“å¼€RODæ–‡ä»¶: $!";
my $rod_header = <$rod_fh>; # è·³è¿‡è¡¨å¤´
while (my $line = <$rod_fh>) {
    chomp $line;
    my @cols = split /\t/, $line;
    # é€‚é…ä½ çš„æ–‡ä»¶ï¼šchr(1), start(2), end(3), rod(6)
    my ($chr, $start, $end, $rod) = @cols[1,2,3,6]; 
    
    # è¿‡æ»¤æ— æ•ˆ/å¼‚å¸¸å€¼ï¼ˆRODå€¼èŒƒå›´0~1ï¼‰
    next if $rod eq 'NA' || $rod =~ /NaN/ || $rod < 0 || $rod > 1;
    
    $rod_data{"$chr:$start:$end"} = $rod;
    push @rod_values, $rod;
}
close $rod_fh;

# è®¡ç®—ROD topé˜ˆå€¼ï¼ˆé™åºæ’åºï¼‰
@rod_values = sort {$b <=> $a} @rod_values;
my $rod_top_idx = int(scalar(@rod_values) * $rod_pct);
my $rod_cutoff = $rod_values[$rod_top_idx] // 0;

# -------------------------- è¾“å‡ºç»“æœè¡¨æ ¼ --------------------------
open my $table_fh, '>', $out_table or die "æ— æ³•å†™å…¥è¡¨æ ¼: $!";
print $table_fh join("\t", qw/Chr Start End Fst ROD Status/), "\n";

my ($total_top, $total_sites) = (0, 0);
foreach my $key (keys %fst_data) {
    my ($chr, $start, $end) = split /:/, $key;
    my $fst_val = $fst_data{$key};
    my $rod_val = $rod_data{$key} // 'NA';
    
    my $status = 'normal';
    # ä»…å½“ä¸¤ä¸ªå€¼éƒ½è¾¾æ ‡æ—¶æ ‡è®°ä¸ºtop
    if ($fst_val >= $fst_cutoff && $rod_val ne 'NA' && $rod_val >= $rod_cutoff) {
        $status = 'top';
        $total_top++;
    }
    
    print $table_fh join("\t", $chr, $start, $end, $fst_val, $rod_val, $status), "\n";
    $total_sites++;
}
close $table_fh;

# -------------------------- è¾“å‡ºç»Ÿè®¡æ–‡ä»¶ --------------------------
open my $stat_fh, '>', $out_stat or die "æ— æ³•å†™å…¥ç»Ÿè®¡æ–‡ä»¶: $!";
my $top_ratio = $total_sites > 0 ? sprintf("%.4f", $total_top/$total_sites) : "0.0000";
print $stat_fh <<"STAT";
# Fst + ROD è”åˆåˆ†æç»Ÿè®¡
Fstæ–‡ä»¶: $fst_file
RODæ–‡ä»¶: $rod_file
Fst top $fst_pct é˜ˆå€¼: $fst_cutoff
ROD top $rod_pct é˜ˆå€¼: $rod_cutoff
æ€»æœ‰æ•ˆä½ç‚¹æ•°é‡: $total_sites
Topæ˜¾è‘—ä½ç‚¹æ•°é‡: $total_top
Topä½ç‚¹å æ¯”: $top_ratio
STAT
close $stat_fh;

# è¿è¡Œæç¤º
print "åˆ†æå®Œæˆï¼\n";
print "ç»“æœè¡¨æ ¼: $out_table\nç»Ÿè®¡æ–‡ä»¶: $out_stat\n";
print "Fsté˜ˆå€¼: $fst_cutoff | RODé˜ˆå€¼: $rod_cutoff | Topä½ç‚¹: $total_top\n";
```
ROD_piæ–‡ä»¶æ ¼å¼ï¼Œè¦ä¸perlç´¢å¼•åˆ—å¯¹åº”
```text
ID	chr	start	end	PI.x	PI.y	rod	pi_ratio1	pi_ratio2
```
```bash
# è”åˆFST RODåˆ†æ
fst=Fst.popC.popW.windowed.weir.fst
ROD=outsample.ROD_PiRatio.table
gff=Wm82.gnm6.gff3

# æå–æ˜¾è‘—ä½ç‚¹
perl ~/pub/script/top_Fst_ROD_S.pl $fst $ROD 0.05 0.05 Fst_ROD.table Fst_ROD.stat

# ç­›é€‰topä½ç‚¹
awk '$NF=="top"' Fst_ROD.table > Fst_ROD.table.top

# æå–åŸºå› æ³¨é‡Š
awk '$3=="gene"' $gff > gene.gff

# æ³¨é‡Šå€™é€‰åŸºå› 
bedtools intersect -wo -F 0.1 -a Fst_ROD.table.top -b gene.gff | \
    awk -F "\t" '{print $7"\t"$10"\t"$11"\t"$13"\t"$15}' | \
    sort -u > Fst_ROD.table.top.gene
```

## GWAS åˆ†æ

### 1. åŸºå› å‹å¡«å……

```bash
beagle -Xmx4g -Djava.io.tmpdir=./TMP \
    gt=../data/all.vcf \
    out=all.impute \
    impute=true \
    window=10 \
    nthreads=2
```

### 2. æ ‡å‡† GWAS åˆ†æ

```bash
# Step 1: å‡†å¤‡VCFæ•°æ®
plink --vcf ../data/all.vcf \
    --maf 0.05 \
    --geno 0.1 \
    --recode12 \
    --output-missing-genotype 0 \
    --transpose \
    --out snp_filter \
    --set-missing-var-ids @:# \
    --allow-extra-chr

# å‡†å¤‡è¡¨å‹æ•°æ®
perl ../script/sort_pheno.pl snp_filter.tfam ../data/trait.table > trait.sort.txt

# Step 2: è®¡ç®—äº²ç¼˜å…³ç³»çŸ©é˜µ
/pub/software/emmax/emmax-kin-intel64 -v -d 10 -o ./pop.kinship snp_filter

# Step 3: è¿è¡ŒGWAS
/pub/software/emmax/emmax-intel64 -v -d 10 \
    -t snp_filter \
    -p trait.sort.txt \
    -k pop.kinship \
    -o emmax.out 1> emmax.log 2>emmax.err

# Step 4: ç»˜åˆ¶æ›¼å“ˆé¡¿å›¾
paste snp_filter.map emmax.out.ps | \
    awk 'BEGIN{print "SNP\tCHR\tBP\tP"}{if($2==$5){print $2"\t"$1"\t"$4"\t"$NF}}' \
    > emmax.out.ps.manht_input

Rscript ../script/manhattan.R emmax.out.ps.manht_input emmax.out.ps.manht_figure
```

### 3. GWAS-Q åˆ†æï¼ˆåŠ å…¥ç¾¤ä½“ç»“æ„åå˜é‡ï¼‰

```bash
# é“¾æ¥æ–‡ä»¶
ln -s ../01.GWAS/snp_filter.tped
ln -s ../01.GWAS/snp_filter.tfam
ln -s ../01.GWAS/snp_filter.nosex
ln -s ../01.GWAS/snp_filter.map
ln -s ../01.GWAS/trait.sort.txt
ln -s ../../02.population_genetics/03.structure/all.3.Q

# å‡†å¤‡QçŸ©é˜µ
paste snp_filter.nosex all.3.Q | awk '{print $1" "$1" 1 "$3" "$4}' > pop.Qmatrix

# è¿è¡ŒGWAS
/pub/software/emmax/emmax-intel64 -v -d 10 \
    -t snp_filter \
    -p trait.sort.txt \
    -k pop.kinship \
    -c pop.Qmatrix \
    -o emmax.out 1> emmax.log 2>emmax.err

# ç»˜åˆ¶æ›¼å“ˆé¡¿å›¾
paste snp_filter.map emmax.out.ps | \
    awk 'BEGIN{print "SNP\tCHR\tBP\tP"}{if($2==$5){print $2"\t"$1"\t"$4"\t"$NF}}' \
    > emmax.out.ps.manht_input

Rscript ../script/manhattan.R emmax.out.ps.manht_input emmax.out.ps.manht_figure_Q
```

### 4. çœŸå®æ•°æ® GWAS

```bash
# Step 1: å‡†å¤‡æ•°æ®
plink --vcf ../data_real/Sample215.M5M8H3.2allel.vcf \
    --maf 0.05 \
    --geno 0.1 \
    --recode12 \
    --output-missing-genotype 0 \
    --transpose \
    --out snp_filter \
    --set-missing-var-ids @:# \
    --allow-extra-chr

perl ../script/sort_pheno.pl snp_filter.tfam ../data_real/trait.C16_0.table > trait.sort.txt

# Step 2: è®¡ç®—äº²ç¼˜å…³ç³»çŸ©é˜µ
/pub/software/emmax/emmax-kin-intel64 -v -d 10 -o ./pop.kinship snp_filter

# Step 3: è¿è¡ŒGWAS
/pub/software/emmax/emmax-intel64 -v -d 10 \
    -t snp_filter \
    -p trait.sort.txt \
    -k pop.kinship \
    -o emmax.out 1> emmax.log 2>emmax.err

# Step 4: ç»˜åˆ¶æ›¼å“ˆé¡¿å›¾
paste snp_filter.map emmax.out.ps | \
    awk 'BEGIN{print "SNP\tCHR\tBP\tP"}{if($2==$5){print $2"\t"$1"\t"$4"\t"$NF}}' \
    > emmax.out.ps.manht_input

Rscript ../script/manhattan.R emmax.out.ps.manht_input emmax.out.ps.manht_figure_readData
```

## QTL-seq åˆ†æ

### 1. VCF è½¬æ¢ä¸º Table æ ¼å¼

```bash
ref=<reference.fasta>
vcf=<input.vcf>
outfile=<output.table>

gatk VariantsToTable \
    -R $ref \
    -V $vcf \
    -F CHROM -F POS -F REF -F ALT \
    -GF AD -GF DP -GF GQ -GF PL \
    -O $outfile
```



### 2. QTL-seq åˆ†æï¼ˆR è„šæœ¬ï¼‰

```R
library(QTLseqr)
library(ggplot2)

# è®¾ç½®å‚æ•°
HighBulk <- "SRR834931"  # é«˜å€¼æ··æ± åç§°
LowBulk <- "SRR834927"   # ä½å€¼æ··æ± åç§°
Chroms <- paste0(rep("Chr", 12), 1:12)  # æŸ“è‰²ä½“åˆ—è¡¨

# ==================== æ•°æ®è¯»å– ====================
df <- importFromGATK(
    file = "../data/Yang_et_al_2013.table",
    highBulk = HighBulk,
    lowBulk = LowBulk,
    chromList = Chroms
)

# ==================== è´¨é‡æ£€æŸ¥å›¾ ====================
# æ·±åº¦åˆ†å¸ƒå›¾
p1 <- ggplot(data = df) + 
    geom_histogram(aes(x = DP.HIGH + DP.LOW)) + 
    xlim(0, 800)
pdf(file = "SNP_depth.pdf")
p1
dev.off()

# REFç­‰ä½åŸºå› é¢‘ç‡åˆ†å¸ƒ
p2 <- ggplot(data = df) + 
    geom_histogram(aes(x = REF_FRQ))
pdf(file = "ref_allele_frequency.pdf")
p2
dev.off()

# é«˜å€¼æ··æ± SNP-indexåˆ†å¸ƒ
p3 <- ggplot(data = df) + 
    geom_histogram(aes(x = SNPindex.HIGH))
pdf(file = "SNPindex.HIGH.dis.pdf")
p3
dev.off()

# ä½å€¼æ··æ± SNP-indexåˆ†å¸ƒ
p4 <- ggplot(data = df) + 
    geom_histogram(aes(x = SNPindex.LOW))
pdf(file = "SNPindex.LOW.dis.pdf")
p4
dev.off()

# ==================== SNPè¿‡æ»¤ ====================
df_filt <- filterSNPs(
    SNPset = df,
    refAlleleFreq = 0.20,    # REFç­‰ä½åŸºå› é¢‘ç‡: 0.2~0.8
    minTotalDepth = 100,     # æœ€å°æ€»æ·±åº¦
    maxTotalDepth = 400,     # æœ€å¤§æ€»æ·±åº¦
    minSampleDepth = 40,     # å•æ ·å“æœ€å°æ·±åº¦
    minGQ = 99,              # åŸºå› å‹è´¨é‡é˜ˆå€¼
    verbose = TRUE
)

# ==================== deltaSNPindexè®¡ç®— ====================
df_filt <- runQTLseqAnalysis(
    df_filt,
    windowSize = 1e6,        # çª—å£å¤§å°
    popStruc = "F2",         # ç¾¤ä½“ç±»å‹
    bulkSize = c(385, 430),  # æ··æ± ä¸ªä½“æ•°
    replications = 10000,    # Bootstrapæ¬¡æ•°
    intervals = c(95, 99)    # ç½®ä¿¡åŒºé—´
)

# ==================== ç»“æœå¯è§†åŒ– ====================
# SNPæ²¿æŸ“è‰²ä½“åˆ†å¸ƒ
p5 <- plotQTLStats(SNPset = df_filt, var = "nSNPs")
pdf(file = "SNP_filter.window.pdf", width = 20, height = 4)
p5
dev.off()

# deltaSNPindexæ²¿æŸ“è‰²ä½“åˆ†å¸ƒ
p6 <- plotQTLStats(
    SNPset = df_filt, 
    var = "deltaSNP", 
    plotIntervals = TRUE
)
pdf(file = "deltaSNPindex.pdf", width = 20, height = 4)
p6
dev.off()

# ==================== æå–æ˜¾è‘—åŒºåŸŸ ====================
QTL <- getSigRegions(
    SNPset = df_filt, 
    method = "QTLseq", 
    interval = 95
)
write.table(QTL[[1]], "QTLseq_result.SigRegions.table", sep = "\t", quote = FALSE)

# ==================== å¯¼å‡ºå®Œæ•´ç»“æœ ====================
results99 <- getQTLTable(
    SNPset = df_filt,
    method = "QTLseq",
    interval = 99,
    export = FALSE
)
write.table(
    results99, 
    file = "QTLseq_result_deltaSNPindex_CI99.table", 
    sep = "\t", 
    quote = FALSE
)
```





## é™„å½•ï¼šåˆ†ææµç¨‹å›¾


```R
åŸå§‹æ•°æ® (FASTQ)
      â”‚
      â–¼
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚   è´¨é‡æ§åˆ¶ QC   â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
      â”‚
      â–¼
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚  æ¯”å¯¹ Mapping   â”‚ â”€â”€â–º BWA MEM
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
      â”‚
      â–¼
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚ å»é‡ MarkDup    â”‚ â”€â”€â–º Picard
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
      â”‚
      â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
      â–¼                      â–¼                      â–¼
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”          â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”          â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚ SNP/InDel â”‚          â”‚    SV     â”‚          â”‚    CNV    â”‚
â”‚   GATK    â”‚          â”‚  lumpy-sv â”‚          â”‚  CNVnator â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜          â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜          â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
      â”‚                      â”‚                      â”‚
      â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
                             â”‚
                             â–¼
                    â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
                    â”‚      æ³¨é‡Š       â”‚
                    â”‚ ANNOVAR/bedtoolsâ”‚
                    â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
                             â”‚
         â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
         â–¼                   â–¼                   â–¼
   â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”       â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”       â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
   â”‚ ç¾¤ä½“ç»“æ„  â”‚       â”‚ é€‰æ‹©åˆ†æ  â”‚       â”‚   GWAS    â”‚
   â”‚ PCA/Tree  â”‚       â”‚ Fst/Ï€/D   â”‚       â”‚   EMMAX   â”‚
   â”‚ Structure â”‚       â”‚           â”‚       â”‚           â”‚
   â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜       â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜       â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
```
### å¸¸è§çš„æ»‘åŠ¨çª—å£å¤§å°è®¾ç½®
```text
ä¸€ã€å¸¸ç”¨çª—å£å¤§å°
çª—å£å¤§å°	é€‚ç”¨åœºæ™¯
10 kb	ç²¾ç»†å®šä½ï¼Œå°åŸºå› ç»„
20 kb	è¾ƒå¸¸ç”¨ï¼Œé€‚ä¸­
50 kb	å¸¸ç”¨ï¼Œå¹³è¡¡ç²¾åº¦å’Œå™ªéŸ³
100 kb	æœ€å¸¸ç”¨ï¼Œå¤§åŸºå› ç»„ï¼ˆå¦‚å¤§è±†ï¼‰
200 kb	ç²—ç•¥æ‰«æï¼Œè¶…å¤§åŸºå› ç»„
äºŒã€çª—å£å¤§å°å–å†³äºä»€ä¹ˆï¼Ÿ
1. åŸºå› ç»„å¤§å°
ç‰©ç§	åŸºå› ç»„å¤§å°	å»ºè®®çª—å£
æ‹Ÿå—èŠ¥	~135 Mb	10-20 kb
æ°´ç¨»	~430 Mb	50-100 kb
å¤§è±†	~1.1 Gb	50-100 kb
ç‰ç±³	~2.3 Gb	100-200 kb
å°éº¦	~17 Gb	200-500 kb
ä¸‰ã€ä¸åŒåˆ†æçš„çª—å£å»ºè®®ï¼ˆå¤§è±†ï¼‰
åˆ†æ	å»ºè®®çª—å£	æ­¥é•¿
Tajima's D	100 kb	-
Ï€ï¼ˆæ ¸è‹·é…¸å¤šæ ·æ€§ï¼‰	100 kb	10 kb
Fst	100 kb	10 kb
ROD	100 kb	10 kb
é€‰æ‹©æ¶ˆé™¤ï¼ˆXP-CLRï¼‰	50-100 kb	10 kb
```
### é€‰æ‹©åˆ†æï¼ˆå¯é€‰ï¼‰
#### XP-CLR (è·¨ç¾¤ä½“å¤åˆä¼¼ç„¶æ¯”æ£€éªŒ)
```BASH
# å®‰è£…xpclr
conda create -n xpclr_env python=3.11 scipy=1.11.4 numpy -y
conda activate xpclr_env
mamba install xpclr
# è½¬æ¢vcf2xpclræ ¼å¼
vcftools --vcf filtered_all.vcf.gz \
         --freq \
         --out allele_freq

```
``` BASH
# æ–‡ä»¶å‡†å¤‡ï¼š
#1.vcfæ–‡ä»¶
bgzip -c all.missing_maf.vcf > all.missing_maf.vcf.gz
echo "Building index..."
tabix -p vcf all.missing_maf.vcf.gz
#2.ä¸¤ä¸ªè¦æ¯”è¾ƒç¾¤ä½“çš„æ ·æœ¬idçš„txtæ–‡ä»¶
#è®¡ç®—ï¼šåŸæ–¹æ³•è®¡ç®—å¤ªæ…¢
#!/bin/bash
RAW_VCF="all.missing_maf.vcf"
WILD="W.sample.list"
CULT="C.sample.list"
OUTDIR="./xpclr_results"
VCFDIR="./vcf_by_chr"

BCFTOOLS_THREADS=4
SPLIT_PARALLEL=10
XPCLR_PARALLEL=10

mkdir -p "$OUTDIR" "$VCFDIR"

# ===== ç¬¬é›¶æ­¥ï¼šè¿‡æ»¤åªä¿ç•™åŒç­‰ä½SNP + å‹ç¼©å»ºç´¢å¼• =====
SNP_VCF="all.snps_only.vcf.gz"

if [ ! -f "$SNP_VCF" ]; then
    echo "[Step 0] Filtering: keep biallelic SNPs only..."
    bcftools view "$RAW_VCF" \
        --types snps \
        -m2 -M2 \
        --threads $BCFTOOLS_THREADS \
        -Oz -o "$SNP_VCF"
    tabix -p vcf "$SNP_VCF"

    # ç»Ÿè®¡
    raw_count=$(bcftools view -H "$RAW_VCF" | wc -l)
    snp_count=$(bcftools view -H "$SNP_VCF" | wc -l)
    indel_count=$((raw_count - snp_count))
    echo "  è¿‡æ»¤å‰æ€»å˜å¼‚: $raw_count"
    echo "  å»æ‰ indel/å¤šç­‰ä½: $indel_count"
    echo "  ä¿ç•™ SNP: $snp_count"
else
    echo "[Step 0] $SNP_VCF already exists, skipping..."
fi
echo "============================================"

VCF="$SNP_VCF"

# ===== ç¬¬ä¸€æ­¥ï¼šè·å–æŸ“è‰²ä½“åˆ—è¡¨ =====
chromosomes=$(bcftools view -h "$VCF" | grep '^##contig' | sed 's/.*ID=\([^,>]*\).*/\1/')
echo "Chromosomes found:"
echo "$chromosomes"
echo "============================================"

# ===== ç¬¬äºŒæ­¥ï¼šå¹¶è¡Œæ‹†åˆ†VCF =====
echo "[Step 1] Splitting VCF by chromosome..."
echo "  bcftools threads per task: $BCFTOOLS_THREADS"
echo "  parallel tasks: $SPLIT_PARALLEL"
echo "============================================"

while IFS= read -r chr; do
    [ -z "$chr" ] && continue

    if [ -f "${VCFDIR}/${chr}.vcf.gz" ] && [ -f "${VCFDIR}/${chr}.vcf.gz.tbi" ]; then
        echo "  $chr already exists, skipping..."
        continue
    fi

    (
        echo "  Splitting $chr ..."
        bcftools view \
            -r "$chr" \
            --threads $BCFTOOLS_THREADS \
            "$VCF" \
            -Oz -o "${VCFDIR}/${chr}.vcf.gz"

        tabix -p vcf "${VCFDIR}/${chr}.vcf.gz"
        echo "  âœ“ $chr split done"
    ) &

    while [ $(jobs -rp | wc -l) -ge $SPLIT_PARALLEL ]; do
        sleep 2
    done
done <<< "$chromosomes"

wait
echo "[Step 1] All chromosomes split!"
echo "============================================"

# ===== éªŒè¯æ‹†åˆ†ç»“æœ =====
echo "Verifying split files..."
all_ok=true
while IFS= read -r chr; do
    [ -z "$chr" ] && continue
    if [ ! -f "${VCFDIR}/${chr}.vcf.gz" ]; then
        echo "  ERROR: ${chr}.vcf.gz missing!"
        all_ok=false
    else
        snp_count=$(bcftools view -H "${VCFDIR}/${chr}.vcf.gz" | wc -l)
        echo "  âœ“ $chr : $snp_count SNPs"
    fi
done <<< "$chromosomes"

if [ "$all_ok" = false ]; then
    echo "Some files missing! Check errors above."
    exit 1
fi
echo "============================================"

# ===== ç¬¬ä¸‰æ­¥ï¼šå¹¶è¡Œè¿è¡Œ xpclr =====
echo "[Step 2] Running XPCLR in parallel..."
echo "  parallel tasks: $XPCLR_PARALLEL"
echo "============================================"

while IFS= read -r chr; do
    [ -z "$chr" ] && continue

    if [ -f "${OUTDIR}/xpclr_${chr}.out" ] && [ -s "${OUTDIR}/xpclr_${chr}.out" ]; then
        echo "  $chr already completed, skipping..."
        continue
    fi

    (
        echo "  [$(date '+%H:%M:%S')] Starting $chr ..."
        xpclr \
            --out "${OUTDIR}/xpclr_${chr}.out" \
            --format vcf \
            --input "${VCFDIR}/${chr}.vcf.gz" \
            --samplesA "$WILD" \
            --samplesB "$CULT" \
            --size 100000 \
            --step 50000 \
            --chr "$chr" \
            2>"${OUTDIR}/xpclr_${chr}.log"
        echo "  [$(date '+%H:%M:%S')] âœ“ $chr completed"
    ) &

    while [ $(jobs -rp | wc -l) -ge $XPCLR_PARALLEL ]; do
        sleep 10
    done
done <<< "$chromosomes"

wait
echo "[Step 2] All XPCLR runs completed!"
echo "============================================"

# ===== ç¬¬å››æ­¥ï¼šåˆå¹¶ç»“æœ =====
echo "[Step 3] Merging results..."

first_chr=$(echo "$chromosomes" | head -1)
head -1 "${OUTDIR}/xpclr_${first_chr}.out" > "${OUTDIR}/chr_all_C_vs_W.txt"

while IFS= read -r chr; do
    [ -z "$chr" ] && continue
    if [ -f "${OUTDIR}/xpclr_${chr}.out" ]; then
        lines=$(tail -n +2 "${OUTDIR}/xpclr_${chr}.out" | wc -l)
        tail -n +2 "${OUTDIR}/xpclr_${chr}.out" >> "${OUTDIR}/chr_all_C_vs_W.txt"
        echo "  $chr : $lines windows"
    else
        echo "  WARNING: $chr result missing!"
    fi
done <<< "$chromosomes"

total=$(tail -n +2 "${OUTDIR}/chr_all_C_vs_W.txt" | wc -l)
echo "============================================"
echo "Done! Total windows: $total"
echo "Output: ${OUTDIR}/chr_all_C_vs_W.txt"
```

```r
# å¯è§†åŒ–ä»£ç ï¼š
library(dplyr)
library(ggplot2)

# ======== 1. è¯»å–åˆå¹¶åçš„å•ä¸ªç»“æœæ–‡ä»¶ ========
df <- read.delim("xpclr_results/chr_all_C_vs_W.txt", 
                 header = TRUE, stringsAsFactors = FALSE)

# ======== 2. æŸ¥çœ‹å®é™…åˆ—åï¼ˆé¦–æ¬¡è¿è¡Œæ—¶å–æ¶ˆæ³¨é‡Šç¡®è®¤ï¼‰ ========
cat("åˆ—åï¼š", colnames(df), "\n")
head(df)

# ======== 3. ç»Ÿä¸€åˆ—åï¼ˆxpclr è¾“å‡ºåˆ—åé€šå¸¸æ˜¯ chrom/start/stop/xpclrï¼‰ ========
# å¦‚æœä½ çš„åˆ—åä¸åŒï¼Œæ ¹æ®ä¸Šé¢ cat è¾“å‡ºä¿®æ”¹è¿™é‡Œ
if ("chrom" %in% colnames(df))    df <- df %>% rename(chr = chrom)
if ("start" %in% colnames(df))    df <- df %>% rename(pos_start = start)
if ("stop"  %in% colnames(df))    df <- df %>% rename(pos_stop = stop)
# å¦‚æœ xpclr åˆ—åæ˜¯ xpclr_scoreï¼š
if ("xpclr_score" %in% colnames(df)) df <- df %>% rename(xpclr = xpclr_score)

# ======== 4. æ•°æ®æ¸…æ´— ========
df <- df %>%
  mutate(xpclr = ifelse(is.na(xpclr) | xpclr <= 0, NA, xpclr)) %>%
  filter(!is.na(xpclr)) %>%
  mutate(pos_center = (pos_start + pos_stop) / 2)

# ======== 5. æŸ“è‰²ä½“æ’åºï¼ˆè‡ªåŠ¨ä»æ•°æ®ä¸­è·å–ï¼Œè€Œéç¡¬ç¼–ç ï¼‰ ========
# æå–æŸ“è‰²ä½“ç¼–å·ä¸­çš„æ•°å­—éƒ¨åˆ†è¿›è¡Œè‡ªç„¶æ’åº
chr_levels <- df %>%
  distinct(chr) %>%
  mutate(chr_num = as.numeric(gsub("\\D", "", chr))) %>%
  arrange(chr_num) %>%
  pull(chr)

df$chr <- factor(df$chr, levels = chr_levels)

# ======== 6. è®¡ç®—ç´¯ç§¯åæ ‡ ========
data_cum <- df %>% 
  group_by(chr) %>% 
  summarise(chr_len = max(pos_center)) %>% 
  mutate(tot = cumsum(as.numeric(chr_len)) - chr_len) %>% 
  select(-chr_len)

df <- left_join(df, data_cum, by = "chr") %>%
  mutate(pos_cum = pos_center + tot)

# ======== 7. é˜ˆå€¼ ========
threshold <- quantile(df$xpclr, 0.95, na.rm = TRUE)

# ======== 8. åæ ‡è½´æ ‡ç­¾ ========
axis_df <- df %>% 
  group_by(chr) %>% 
  summarize(center = (max(pos_cum) + min(pos_cum)) / 2)

# ======== 9. é¢œè‰²ï¼ˆæ ¹æ®å®é™…æŸ“è‰²ä½“æ•°é‡è‡ªåŠ¨ç”Ÿæˆï¼‰ ========
n_chr <- length(chr_levels)
chr_colors <- rep(c("#d7191c", "#2c7bb6"), length.out = n_chr)

# ======== 10. ç»˜å›¾ ========
p <- ggplot(df, aes(x = pos_cum, y = xpclr, color = chr)) +
  geom_line(linewidth = 0.5) +    # æ³¨æ„ï¼šæ–°ç‰ˆggplot2ç”¨linewidthæ›¿ä»£size
  scale_color_manual(values = chr_colors) +
  geom_hline(yintercept = threshold, linetype = "dashed", 
             color = "black", alpha = 0.7) +
  scale_x_continuous(label = axis_df$chr, breaks = axis_df$center) +
  labs(
    x = "Chromosome",
    y = "XP-CLR Score",
    subtitle = paste0("Dashed line: 95th percentile threshold (", 
                      round(threshold, 2), ")")
  ) +
  theme_minimal(base_size = 12) +
  theme(
    legend.position = "none",
    plot.title = element_text(hjust = 0.5, size = 14, face = "bold"),
    panel.grid.minor = element_blank(),
    panel.grid.major.x = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1)  # Xè½´æ ‡ç­¾æ—‹è½¬
  )

# ======== 11. ä¿å­˜ ========
ggsave("xpclr_manhattan.pdf", p, width = 14, height = 5)
ggsave("xpclr_manhattan.png", p, width = 14, height = 5, dpi = 300)

print(p)
```
```bash
# xpclr æºç çš„å·²çŸ¥bugï¼Œéœ€è¦ç›´æ¥ä¿®æ”¹æºç 
--- Logging error ---
Traceback (most recent call last):
File "/home2/xtc/.conda/envs/xpclr_env/lib/python3.11/logging/init.py", line 1110, in emit
msg = self.format(record)
^^^^^^^^^^^^^^^^^^^
File "/home2/xtc/.conda/envs/xpclr_env/lib/python3.11/logging/init.py", line 953, in format
return fmt.format(record)
^^^^^^^^^^^^^^^^^^
File "/home2/xtc/.conda/envs/xpclr_env/lib/python3.11/logging/init.py", line 687, in format
record.message = record.getMessage()
^^^^^^^^^^^^^^^^^^^
File "/home2/xtc/.conda/envs/xpclr_env/lib/python3.11/logging/init.py", line 377, in getMessage
msg = msg % self.args
~~~~^~~~~~~~~~~
TypeError: not all arguments converted during string formatting
Call stack:
File "/home2/xtc/.conda/envs/xpclr_env/bin/xpclr", line 196, in <module>
main()
File "/home2/xtc/.conda/envs/xpclr_env/bin/xpclr", line 182, in main
model_lik, null_lik, selcoef, nsnps, navail, snpedges = xpclr.methods.xpclr_scan(
File "/home2/xtc/.conda/envs/xpclr_env/lib/python3.11/site-packages/xpclr/methods.py", line 333, in xpclr_scan
li_data[i] = compute_xpclr(window_data, sel_coefs)
File "/home2/xtc/.conda/envs/xpclr_env/lib/python3.11/site-packages/xpclr/methods.py", line 225, in compute_xpclr
ll = calculate_cl_romberg(s_coef, dat)
File "/home2/xtc/.conda/envs/xpclr_env/lib/python3.11/site-packages/xpclr/methods.py", line 204, in calculate_cl_romberg
cl = calculate_likelihood(values=(xj, nj, c, p2freq, var))
File "/home2/xtc/.conda/envs/xpclr_env/lib/python3.11/site-packages/xpclr/methods.py", line 137, in chen_likelihood
logger.warning(w[-1].message, i_likl, i_base)
Message: DeprecationWarning('Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)')
Arguments: ((0.050147255182536286, 4.2684261388273304e-07, {'neval': 63, 'last': 2, 'iord': array([ 1, 2, 5, 4, 3,ã€‚ã€‚ã€‚ã€‚
# ä¸€é”®ä¿®å¤
# æ‰¾åˆ°æ–‡ä»¶
METHODS_FILE="/home2/xtc/.conda/envs/xpclr_env/lib/python3.11/site-packages/xpclr/methods.py"

# å¤‡ä»½
cp "$METHODS_FILE" "${METHODS_FILE}.bak"

# ä¿®å¤ç¬¬137è¡Œ
sed -i 's/logger.warning(w\[-1\].message, i_likl, i_base)/logger.warning(f"{w[-1].message} | likl={i_likl}, base={i_base}")/' "$METHODS_FILE"
éªŒè¯æ˜¯å¦ä¿®æ”¹æˆåŠŸ

grep -n "logger.warning" "$METHODS_FILE"

#åº”è¯¥çœ‹åˆ°ï¼š
137:                    logger.warning(f"{w[-1].message} | likl={i_likl}, base={i_base}")
# åŸå› è§£é‡Š

# åŸä»£ç ï¼ˆæœ‰bugï¼‰ï¼š
logger.warning(w[-1].message, i_likl, i_base)
#              â†‘ è¿™ä¸ªå­—ç¬¦ä¸²é‡Œæ²¡æœ‰ %s å ä½ç¬¦ï¼Œå´ä¼ äº†ä¸¤ä¸ªå‚æ•° â†’ TypeError

# ä¿®å¤åï¼š
logger.warning(f"{w[-1].message} | likl={i_likl}, base={i_base}")
#              â†‘ ç”¨ f-string æŠŠæ‰€æœ‰å‚æ•°æ‹¼è¿›ä¸€ä¸ªå­—ç¬¦ä¸²ï¼Œä¸å†æœ‰å¤šä½™å‚æ•°

```
#### XP-EHHï¼ˆæ›´åŠ å¿«é€Ÿç®€ä¾¿çš„æ–¹æ³•åœ¨8000é‡æµ‹åºæ–‡ç« ä¸­æœ‰ï¼‰
```bash
# æ‹†åˆ†vcfæ–‡ä»¶ä¸ºä¸åŒç¾¤ä½“
cat ../chr.list | while read chr; do echo "vcftools --gzvcf ${chr}.vcf.gz --keep ../W.sample.list --recode --recode-INFO-all --out W.${chr}.missmaf" >> split.chr.vcf.sh; done
# åŸºå› å‹å¡«å……
cat ../chr.list | while read chr; do echo "java -Xmx10g -jar /home2/xtc/.conda/envs/xpehh/share/beagle-5.5_27Feb25.75f-0/beagle.jar nthreads=4 ne=30 gt=W.${chr}.missmaf.recode.vcf out=W_${chr}.phasing.recode" >> W.beagle.sh; done
cat ../chr.list | while read chr; do echo "java -Xmx10g -jar /home2/xtc/.conda/envs/xpehh/share/beagle-5.5_27Feb25.75f-0/beagle.jar nthreads=4 ne=30 gt=C.${chr}.missmaf.recode.vcf out=C_${chr}.phasing.recode" >> C.beagle.sh; done

nohup parallel -j 5 --joblog Wbeagle_progress.log < W.beagle.sh > beagle_W.log 2>&1 &
nohup parallel -j 5 --joblog Cbeagle_progress.log < C.beagle.sh > beagle_C.log 2>&1 &

# è½¬æ¢æ ¼å¼
mkdir -p ./xpehh_result
for i in {01..20}; do
vcftools --gzvcf "W_glyma.Wm82.gnm6.Gm${i}.phasing.recode.vcf.gz" --plink --out ./xpehh_result/W.Gm${i}
done 

for i in {01..20}; do vcftools --gzvcf "C_glyma.Wm82.gnm6.Gm${i}.phasing.recode.vcf.gz" --plink --out ./xpehh_result/C.Gm${i}; done 
```
è®¡ç®—XP-EHHéœ€è¦è®¡ç®—é—ä¼ è·ç¦»ï¼Œä¸€èˆ¬æ¥è¯´é»˜è®¤ä½¿ç”¨1cM=1Mb,å°†ç¬¬å››åˆ—çš„ç‰©ç†ä½ç½®ä¹˜ä»¥0.000001å³å¯æ¢ç®—æˆé—ä¼ è·ç¦»
```bash
for i in {01..20}; do
  awk '{print $1, $2, $4/1000000, $4}' C.Gm${i}.map > C.Gm${i}.genetic.map
done
for i in {01..20}; do
  awk '{print $1, $2, $4/1000000, $4}' W.Gm${i}.map > W.Gm${i}.genetic.map
done
```
```bash
# è®¡ç®—
mkdir -p results
for i in {01..20}; do
selscan --xpehh --vcf ../C_glyma.Wm82.gnm6.Gm${i}.phasing.recode.vcf.gz --vcf-ref ../W_glyma.Wm82.gnm6.Gm${i}.phasing.recode.vcf.gz --map W.Gm${i}.genetic.map --threads 10 --out ./results/C_W_Gm${i}
done
```

```bash
# è„šæœ¬å½’ä¸€åŒ– https://github.com/Fmaizer/sliding-window-for-xpehh-results ï¼ˆpandas å’Œ tqdmï¼‰
for i in {01..20}; do python norm.py --input C_W_Gm${i}.xpehh.out --window 100000 --step 50000 --chr $i; done
# åˆå¹¶æ‰€æœ‰æŸ“è‰²ä½“ç»“æœ
awk 'FNR==1 && NR!=1{next;}{print}' C_W_Gm*.xpehh.out.norm.100kb.windows > C_W_all.xpehh.out.norm.100kb.windows
```

```r
# ç»˜å›¾ï¼ˆå„æŸ“è‰²ä½“æ•£ç‚¹å›¾ï¼‰
library(ggplot2)

# è¯»å–çª—å£æ±‡æ€»æ•°æ®
xpehh <- read.table("C_W_all.xpehh.out.norm.100kb.windows", header=TRUE)

# æŸ¥çœ‹åˆ—åï¼Œç¡®è®¤å®é™…åç§°
colnames(xpehh)

# è®¡ç®—çª—å£ä¸­ç‚¹ä½ç½®ï¼ˆç”¨äºç»˜å›¾ï¼‰
xpehh$mid <- (xpehh$win_start + xpehh$win_end) / 2

# ç»˜åˆ¶ XP-EHH æ‰«æå›¾ï¼ˆä½¿ç”¨ mean_normxpehhï¼‰
ggplot(xpehh, aes(x=mid/1000000, y=mean_normxpehh)) +
  geom_point(size=0.5, alpha=0.6, color="steelblue") +
  geom_hline(yintercept=0, linetype="dashed", color="gray50") +
  geom_hline(yintercept=2, linetype="dashed", color="red") +
  geom_hline(yintercept=-2, linetype="dashed", color="blue") +
  facet_wrap(~chr, scales="free_x", ncol=4) +
  labs(x="Position (Mb)", y="Mean Normalized XP-EHH",
       title="XP-EHH Scan (100kb Windows)") +
  theme_bw() +
  theme(axis.text.x = element_text(angle=45, hjust=1))
ggsave("chr.W_C.xpehh.pot.png",width = 10,height = 7,dpi = 300)
# è¯†åˆ«æ˜¾è‘—ä¿¡å·ï¼ˆ|mean_normxpehh| > 2ï¼‰
candidates_pos <- xpehh[xpehh$mean_normxpehh > 2, ]
candidates_neg <- xpehh[xpehh$mean_normxpehh < -2, ]

# ä¿å­˜å€™é€‰åŒºåŸŸ
write.table(candidates_pos, "candidates_positive.txt", 
            row.names=FALSE, quote=FALSE, sep="\t")
write.table(candidates_neg, "candidates_negative.txt", 
            row.names=FALSE, quote=FALSE, sep="\t")

# æŸ¥çœ‹å€™é€‰åŒºåŸŸæ•°é‡
cat("æ­£é€‰æ‹©ä¿¡å·çª—å£æ•°:", nrow(candidates_pos), "\n")
cat("è´Ÿé€‰æ‹©ä¿¡å·çª—å£æ•°:", nrow(candidates_neg), "\n")

XP_EHH<-read.table("C_W_all.xpehh.out.norm.100kb.windows",
                   header = TRUE, stringsAsFactors = FALSE)
XP_EHH$mid<-(XP_EHH$win_start+XP_EHH$win_end)/2
ggplot(XP_EHH, aes(x = mid, y = mean_normxpehh)) +
  geom_line(linewidth = 0.3,color="skyblue") +
  theme_minimal()+
  theme(
    text             = element_text(family = "serif", color = "black"),
    axis.text.x      = element_text(angle = 0, color = "black", size = 10),
    axis.text.y      = element_text(color = "black", size = 10),
    axis.title.y     = element_text(color = "black", size = 12),
    axis.ticks       = element_line(size = 0.4, colour = "black"),
    panel.grid.minor = element_blank(),
    panel.grid.major = element_blank(),
    panel.border     = element_rect(fill = NA, color = "black", size = 0.6),
    plot.title       = element_text(hjust = 0.5, size = 14, face = "bold"),
    plot.margin      = margin(10, 10, 10, 10),
    # å°†å›¾ä¾‹ç§»å…¥å›¾å†…ï¼Œå³ä¸‹è§’
    legend.position = c(0.95, 0.05),
    legend.justification = c("right", "bottom"),
    legend.background = element_rect(fill = alpha("white", 0.6), color = NA),
    legend.key = element_blank())
ggsave("all.chr.W_C.xpehh.line.png",width = 10,height = 7,dpi = 300)

```