+++
date = '2026-02-03T12:08:55+08:00'
draft = false
title = 'äº‘å—å†œä¸šå¤§å­¦é‡æµ‹åºé¡¹ç›®æ•™ç¨‹'
# æ·»åŠ åˆ†ç±»
categories:
  - é‡æµ‹åºåˆ†æ


# æ·»åŠ æ ‡ç­¾
tags:
  - ç¾¤ä½“ç»“æ„åˆ†æ
  - å˜å¼‚æ£€æµ‹
  - é‡æµ‹åº
  - GWAS
  - QTL-seq
+++
äº‘å—å†œä¸šå¤§å­¦ äº‘å—ç”Ÿç‰©èµ„æºä¿æŠ¤ä¸åˆ©ç”¨å›½å®¶é‡ç‚¹å®éªŒå®¤

æè¯¦ | 2019 å¹´ 12 æœˆ 31 æ—¥

## ç›®å½•

- [è½¯ä»¶å®‰è£…](#%E8%BD%AF%E4%BB%B6%E5%AE%89%E8%A3%85)
- [å˜å¼‚æ£€æµ‹](#%E5%8F%98%E5%BC%82%E6%A3%80%E6%B5%8B)
- [ç¾¤ä½“ç»“æ„åˆ†æ](#%E7%BE%A4%E4%BD%93%E7%BB%93%E6%9E%84%E5%88%86%E6%9E%90)
- [é€‰æ‹©åˆ†æ](#%E9%80%89%E6%8B%A9%E5%88%86%E6%9E%90)
- [GWAS åˆ†æ](#gwas%E5%88%86%E6%9E%90)
- [QTL-seq åˆ†æ](#qtl-seq%E5%88%86%E6%9E%90)
- [é™„å½•ï¼šåˆ†ææµç¨‹å›¾](#%E9%99%84%E5%BD%95%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B%E5%9B%BE)



## è½¯ä»¶å®‰è£…



> âš ï¸ æ³¨æ„ï¼šè¯¥éƒ¨åˆ†è½¯ä»¶å¾ˆéš¾ç”¨ conda ç›´æ¥å®‰è£…ï¼Œå®‰è£…æ­¥éª¤æ¯”è¾ƒç‰¹æ®Šã€‚

### 1. lumpy-sv

ğŸ’¡ lumpy-sv ä¾èµ– Python2.7ï¼Œéœ€åˆ‡æ¢åˆ° py27 ç¯å¢ƒä¸‹è¿›è¡Œå®‰è£…ã€‚

```bash
conda activate py27
git clone --recursive https://github.com/arq5x/lumpy-sv.git
cd lumpy-sv
make
```

### 2. svtyper

```bash
conda install svtyper
```

### 3. cnvnator

ğŸ’¡ cnvnator ä¾èµ–äº root è½¯ä»¶åŒ…åŠ samtools è½¯ä»¶åŒ…ï¼ˆåŒ…å« HTSlibï¼‰ï¼Œå› æ­¤å…ˆå®‰è£…ä¾èµ–ã€‚

#### 3.1 å®‰è£… samtools

```bash
wget https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2
tar -xvf samtools-1.9.tar.bz2
cd samtools-1.9/
./configure
make
```

#### 3.2 å®‰è£… root è½¯ä»¶

```bash
# ç›´æ¥ä¸‹è½½åè§£å‹å³å¯
wget https://root.cern/download/root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz
tar -zxvf root_v6.18.04.Linux-ubuntu18-x86_64-gcc7.4.tar.gz
```

#### 3.3 å®‰è£… cnvnator

```bash
git clone https://github.com/abyzovlab/CNVnator.git
cd CNVnator

# é“¾æ¥samtoolsè½¯ä»¶ç›®å½•åˆ°cnvnatorç›®å½•
ln -s ../samtools-1.9 ./samtools

# åŠ è½½rootè½¯ä»¶åˆ°ç¯å¢ƒå˜é‡
export ROOTSYS=/pub/software/root
export PATH=/pub/software/root/bin:$PATH
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ROOTSYS/lib

# å®‰è£…cnvnator
make LIBS="-lcrypto"
```



## å˜å¼‚æ£€æµ‹



### 1. æ„å»ºåŸºå› ç»„ index

```bash
# åˆ›å»ºè½¯é“¾æ¥
ln -s ../data/genome.fasta ./genome.fasta

# for gatk
samtools faidx genome.fasta

# for bwa
bwa index genome.fasta

# for picard
picard CreateSequenceDictionary R=genome.fasta
```

### 2. æ¯”å¯¹æ’åºå»é‡

#### 2.1 æ¯”å¯¹æ’åº

```bash
bwa mem -t 2 -R '@RG\tID:S1\tSM:S1\tPL:illumina' \
    ../01.ref/genome.fasta \
    ../data/S1_1.fq.gz ../data/S1_2.fq.gz | \
    /pub/software/samtools/samtools sort -@ 2 -m 1G -o S1.sort.bam -

bwa mem -t 2 -R '@RG\tID:S2\tSM:S2\tPL:illumina' \
    ../01.ref/genome.fasta \
    ../data/S2_1.fq.gz ../data/S2_2.fq.gz | \
    /pub/software/samtools/samtools sort -@ 2 -m 1G -o S2.sort.bam -
```

#### 2.2 å»é™¤é‡å¤

```bash
picard -Xmx4g MarkDuplicates \
    I=S1.sort.bam \
    O=S1.sort.rmdup.bam \
    CREATE_INDEX=true \
    REMOVE_DUPLICATES=true \
    M=S1.marked_dup_metrics.txt

picard -Xmx4g MarkDuplicates \
    I=S2.sort.bam \
    O=S2.sort.rmdup.bam \
    CREATE_INDEX=true \
    REMOVE_DUPLICATES=true \
    M=S2.marked_dup_metrics.txt
```

#### 2.3 æ¯”å¯¹ç‡ç»Ÿè®¡

```bash
/pub/software/samtools/samtools flagstat S1.sort.bam > S1.sort.bam.flagstat
/pub/software/samtools/samtools flagstat S2.sort.bam > S2.sort.bam.flagstat
```

#### 2.4 æ‰¹é‡æ¯”å¯¹è„šæœ¬ç”Ÿæˆ

```bash
less ../data/sample.list | awk '{ 
    print "bwa mem -t 2 -R '\''@RG\\tID:"$1"\\tSM:"$1"\\tPL:illumina'\'' \
    ../01.ref/genome.fasta ../data/"$1"_1.fq.gz ../data/"$1"_2.fq.gz | \
    /pub/software/samtools/samtools sort -@ 2 -m 1G -o "$1".sort.bam -" 
}'
```

### 3. SNPã€Indel æ£€æµ‹

#### 3.1 HaplotypeCaller

```bash
gatk --java-options "-Xmx10g -Djava.io.tmpdir=./tmp" HaplotypeCaller \
    -R ../01.ref/genome.fasta \
    -I ../02.mapping/S1.sort.rmdup.bam \
    -ERC GVCF \
    -O S1.g.vcf 1>S1.HC.log 2>&1

gatk --java-options "-Xmx10g -Djava.io.tmpdir=./tmp" HaplotypeCaller \
    -R ../01.ref/genome.fasta \
    -I ../02.mapping/S2.sort.rmdup.bam \
    -ERC GVCF \
    -O S2.g.vcf 1>S2.HC.log 2>&1
```

#### 3.2 CombineGVCFs

```bash
ls ./S*.g.vcf > gvcf.list

gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" CombineGVCFs \
    -R ../01.ref/genome.fasta \
    -V gvcf.list \
    -O all.merge.g.vcf
```

#### 3.3GenotpeGVCFs

```bash
gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" GenotypeGVCFs \
    -R ../01.ref/genome.fasta \
    --variant all.merge.g.vcf \
    -O all.merge_raw.vcf
```

#### 3.4 SNP è¿‡æ»¤

```bash
# æå–SNP
gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" SelectVariants \
    -R ../01.ref/genome.fasta \
    -V all.merge_raw.vcf \
    --select-type SNP \
    -O all.raw.snp.vcf

# æ ‡è®°è¿‡æ»¤
gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" VariantFiltration \
    -R ../01.ref/genome.fasta \
    -V all.raw.snp.vcf \
    --filter-expression "QD < 2.0 || MQ < 40.0 || FS > 60.0 || SOR > 3.0 || MQRankSum < -12.5 || ReadPosRankSum < -8.0" \
    --filter-name 'SNP_filter' \
    -O all.filter.snp.vcf

# æå–è¿‡æ»¤åçš„SNP
gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" SelectVariants \
    -R ../01.ref/genome.fasta \
    -V all.filter.snp.vcf \
    --exclude-filtered \
    -O all.filtered.snp.vcf
```

#### 3.5 InDel è¿‡æ»¤

```bash
# æå–InDel
gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" SelectVariants \
    -R ../01.ref/genome.fasta \
    -V all.merge_raw.vcf \
    --select-type INDEL \
    -O all.raw.indel.vcf

# æ ‡è®°è¿‡æ»¤
gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" VariantFiltration \
    -R ../01.ref/genome.fasta \
    -V all.raw.indel.vcf \
    --filter-expression "QD < 2.0 || FS > 200.0 || SOR > 10.0 || MQRankSum < -12.5 || ReadPosRankSum < -8.0" \
    --filter-name 'INDEL_filter' \
    -O all.filter.indel.vcf

# æå–è¿‡æ»¤åçš„InDel
gatk --java-options "-Xmx4g -Djava.io.tmpdir=./tmp" SelectVariants \
    -R ../01.ref/genome.fasta \
    -V all.filter.indel.vcf \
    --exclude-filtered \
    -O all.filtered.indel.vcf
```

### 4. SV æ£€æµ‹

```bash
# Step 1: æå– discordants reads
/pub/software/samtools/samtools view -b -F 1294 -@ 12 \
    ../02.mapping/S1.sort.rmdup.bam > S1.discordants.bam
/pub/software/samtools/samtools view -b -F 1294 -@ 12 \
    ../02.mapping/S2.sort.rmdup.bam > S2.discordants.bam

# Step 2: æå– split reads
/pub/software/samtools/samtools view -h ../02.mapping/S1.sort.rmdup.bam | \
    /pub/software/lumpy-sv/scripts/extractSplitReads_BwaMem -i stdin | \
    /pub/software/samtools/samtools sort - > S1.splitters.bam

/pub/software/samtools/samtools view -h ../02.mapping/S2.sort.rmdup.bam | \
    /pub/software/lumpy-sv/scripts/extractSplitReads_BwaMem -i stdin | \
    /pub/software/samtools/samtools sort - > S2.splitters.bam

# Step 3: è¿è¡Œ lumpy
/pub/software/lumpy-sv/bin/lumpyexpress \
    -B ../02.mapping/S1.sort.rmdup.bam,../02.mapping/S2.sort.rmdup.bam \
    -S S1.splitters.bam,S2.splitters.bam \
    -D S1.discordants.bam,S2.discordants.bam \
    -o all.sv.lumpy.vcf

# Step 4: ä¸ªä½“åŸºå› å‹åˆ†å‹
vcftools --vcf all.sv.lumpy.vcf --indv S1 --recode --recode-INFO-all --out S1 && \
    svtyper -i S1.recode.vcf -B ../02.mapping/S1.sort.rmdup.bam -o S1.genotype.vcf

vcftools --vcf all.sv.lumpy.vcf --indv S2 --recode --recode-INFO-all --out S2 && \
    svtyper -i S2.recode.vcf -B ../02.mapping/S2.sort.rmdup.bam -o S2.genotype.vcf
```

### 5. CNV æ£€æµ‹

#### 5.1 å‡†å¤‡åŸºå› ç»„

```bash
# å°†åŸºå› ç»„æ‹†åˆ†ä¸ºå•æ¡æŸ“è‰²ä½“
ln -s ../data/genome.fasta ./genome.fa
seqkit split -i ./genome.fa -O split

# é‡å‘½åæ–‡ä»¶
ls ./split/genome.id_*.fa | sed 's/./split/genome.id_//' | \
    awk '{print "mv ./split/genome.id_"$aa" ./split/"$aa}' > mv_name.sh
sh mv_name.sh
```

#### 5.2 è¿è¡Œ cnvnator

```bash
# è®¾ç½®ç¯å¢ƒå˜é‡
export ROOTSYS=/pub/software/root
export PATH=/pub/software/root/bin:$PATH
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$ROOTSYS/lib

# é“¾æ¥æ–‡ä»¶
ln -s ../02.mapping/S1.sort.rmdup.bam

# æå–æ¯”å¯¹ç»“æœ
/pub/software/CNVnator/cnvnator -genome genome.fa -root S1.root -tree S1.sort.rmdup.bam

# ç”Ÿæˆæ·±åº¦åˆ†å¸ƒ
/pub/software/CNVnator/cnvnator -genome genome.fa -root S1.root -his 500 -d split

# è¿›è¡Œç»Ÿè®¡è®¡ç®—
/pub/software/CNVnator/cnvnator -root S1.root -stat 500

# æ£€æŸ¥bin sizeæ˜¯å¦åˆé€‚
/pub/software/CNVnator/cnvnator -root S1.root -eval 500 > S1.eval.ratio

# RDä¿¡å·åˆ†å‰²
/pub/software/CNVnator/cnvnator -root S1.root -partition 500

# è¿›è¡ŒCNVæ£€æµ‹
/pub/software/CNVnator/cnvnator -root S1.root -call 500 > S1.cnv
```

### 6. SNP/InDel æ³¨é‡Š

#### 6.1 å‡†å¤‡æ³¨é‡Šæ•°æ®åº“

```bash
ln -s ../data/genome.gtf
ln -s ../data/genome.fasta

gtfToGenePred -genePredExt genome.gtf genome_refGene.txt

/pub/software/annovar/retrieve_seq_from_fasta.pl \
    --format refGene \
    --seqfile genome.fasta \
    genome_refGene.txt \
    --out genome_refGeneMrna.fa
```

#### 6.2 æ³¨é‡Š VCF æ–‡ä»¶

```bash
ln -s ../03.SNP_indel/all.filtered.snp.vcf

# æ ¼å¼è½¬æ¢
/pub/software/annovar/convert2annovar.pl \
    -format vcf4old \
    all.filtered.snp.vcf > all.snp.vcf.annovar.input

# è¿›è¡Œæ³¨é‡Š
/pub/software/annovar/annotate_variation.pl \
    -geneanno \
    --neargene 2000 \
    -buildver genome \
    -dbtype refGene \
    -outfile all.anno \
    -exonsort \
    all.snp.vcf.annovar.input ./
```

### 7. CNV å’Œ SV æ³¨é‡Š

```bash
# å¤„ç†SVæ–‡ä»¶
less -S ../04.SV/all.sv.lumpy.vcf | \
    awk -F ";" '{ if($1!~/#/){print $1"\t"$2"\t"$3"\t"$4} }' | \
    awk '{print $1"\t"$2"\t"$11"\t"$3"\t"$8}' | \
    sed 's/END=//' | sed 's/SVTYPE=//' > SV.bed

# è¿‡æ»¤BNDç±»å‹
awk '$5 !~ /BND/' SV.bed > SV_filter.bed

# SVæ³¨é‡Š
bedtools intersect -wo -a SV_filter.bed -b gene.gtf > SV_filter.Anno

# å¤„ç†CNVæ–‡ä»¶
less -S ../05.CNV/S1.cnv | \
    awk '{print $2 "\t" $1}' | \
    sed 's/:/\t/' | sed 's/-/\t/' > CNV.bed

# CNVæ³¨é‡Š
bedtools intersect -wo -a CNV.bed -b gene.gtf > CNV.Anno
```

## ç¾¤ä½“ç»“æ„åˆ†æ

### 1. æ•°æ®è¿‡æ»¤

```bash
vcf=../data/all.vcf

# è¿‡æ»¤ç¼ºå¤±ç‡å’Œæœ€å°ç­‰ä½åŸºå› é¢‘ç‡
plink --vcf $vcf \
    --geno 0.1 \
    --maf 0.01 \
    --out all.missing_maf \
    --recode vcf-iid \
    --allow-extra-chr \
    --set-missing-var-ids @:# \
    --keep-allele-order

# è¿‡æ»¤è¿é”ä¸å¹³è¡¡
plink --vcf all.missing_maf.vcf \
    --indep-pairwise 50 10 0.2 \
    --out tmp.ld \
    --allow-extra-chr \
    --set-missing-var-ids @:#

plink --vcf all.missing_maf.vcf \
    --make-bed \
    --extract tmp.ld.prune.in \
    --out all.LDfilter \
    --recode vcf-iid \
    --keep-allele-order \
    --allow-extra-chr \
    --set-missing-var-ids @:#
```

### 2. å»ºæ ‘åˆ†æ

#### 2.1 NJ æ ‘

```bash
# æ ¼å¼è½¬æ¢
perl ../../script/vcf2phy.pl ../../00.filter/all.LDfilter.vcf > sequences.dat

# è®¡ç®—é—ä¼ è·ç¦»
echo -e "sequences.dat\nY" > dnadist.cfg
dnadist < dnadist.cfg > dnadist.log
mv outfile infile.dist

# æ„å»ºNJæ ‘
echo -e "infile.dist\nY" > neighbor.cfg
neighbor < neighbor.cfg > nj.log

# æ ¼å¼æ•´ç†
less infile.dist | tr '\n' '|' | sed 's/| / /g' | tr '|' '\n' > infile.dist.table
less outtree | tr '\n' ' ' | sed 's/ //g' > outtree.nwk
```

#### 2.2 SNPhylo å»ºæ ‘

```bash
/pub/software/SNPhylo/snphylo.sh \
    -v ../../00.filter/all.LDfilter.vcf \
    -r \
    -l 1 \
    -m 0.01 \
    -o GA0001 \
    -b \
    -B 2 \
    -P tree_snphylo
```



### 3. PCA åˆ†æ

```bash
plink --vcf ../00.filter/all.LDfilter.vcf \
    --pca 10 \
    --out PCA_out \
    --allow-extra-chr \
    --set-missing-var-ids @:#

Rscript ../script/draw_PCA.R PCA_out.eigenvec 1 2 ../data/sample.pop PCA_out_figure
```

### 4. ç¾¤ä½“ç»“æ„åˆ†æ

```bash
# Step 0: vcfè½¬bedæ ¼å¼
plink --vcf ../00.filter/all.LDfilter.vcf \
    --make-bed \
    --out all \
    --allow-extra-chr \
    --keep-allele-order \
    --set-missing-var-ids @:#

# Step 1: ç”ŸæˆAdmixtureè¿è¡Œè„šæœ¬
seq 2 4 | awk '{print "admixture --cv -j2 all.bed "$1" 1>admix."$1".log 2>&1"}' > admixture.sh

# Step 2: è¿è¡ŒAdmixture
sh admixture.sh

# ä¹Ÿå¯å•ç‹¬è¿è¡Œ
admixture --cv -j2 all.bed 2 1>admix.2.log 2>&1
admixture --cv -j2 all.bed 3 1>admix.3.log 2>&1
admixture --cv -j2 all.bed 4 1>admix.4.log 2>&1

# Step 3: ç»˜åˆ¶ç»“æ„å›¾
mkdir result
cp ./*.Q result/
Rscript ../script/draw_admixture.R result all.nosex structure
```

### 5. LD è¡°å‡åˆ†æ

```bash
# Step 1: è®¡ç®—LD
/pub/software/PopLDdecay/bin/PopLDdecay \
    -InVCF ../data/all.vcf \
    -SubPop ../data/pop.SC.table \
    -MaxDist 500 \
    -OutStat pop.SC.stat

/pub/software/PopLDdecay/bin/PopLDdecay \
    -InVCF ../data/all.vcf \
    -SubPop ../data/pop.YZR.table \
    -MaxDist 500 \
    -OutStat pop.YZR.stat

# Step 2: ç»˜åˆ¶å•ç¾¤ä½“LDè¡°å‡å›¾
perl /pub/software/PopLDdecay/bin/Plot_OnePop.pl \
    -inFile pop.SC.stat.gz \
    -output pop.SC.ld

perl /pub/software/PopLDdecay/bin/Plot_OnePop.pl \
    -inFile pop.YZR.stat.gz \
    -output pop.YZR.ld

# Step 3: ç»˜åˆ¶å¤šç¾¤ä½“LDè¡°å‡å›¾
perl /pub/software/PopLDdecay/bin/Plot_MultiPop.pl \
    -inList ld_stat.list \
    -output ld_stat.multi
```

## é€‰æ‹©åˆ†æ

### 1. Ï€ å’Œ ROD åˆ†æ

```bash
vcf=../../data/all.vcf
window=20000
step=2000

# Step 1: è®¡ç®—piå€¼
vcftools --vcf $vcf \
    --window-pi $window \
    --window-pi-step $step \
    --keep ../../data/pop.SC.table \
    --out ./Pi.pop1

vcftools --vcf $vcf \
    --window-pi $window \
    --window-pi-step $step \
    --keep ../../data/pop.YZR.table \
    --out ./Pi.pop2

# Step 2: è®¡ç®—ROD
pi1=Pi.pop1.windowed.pi
pi2=Pi.pop2.windowed.pi
perl ../../script/merge2pi_ROD.pl $pi1 $pi2 > Pi.ROD
```

### 2. Tajima's D åˆ†æ

```bash
vcf=../../data/all.vcf
window=20000

vcftools --vcf $vcf \
    --TajimaD $window \
    --keep ../../data/pop.SC.table \
    --out TajimaD.pop1

vcftools --vcf $vcf \
    --TajimaD $window \
    --keep ../../data/pop.YZR.table \
    --out TajimaD.pop2
```

### 3. Fst åˆ†æ

```bash
vcf=../../data/all.vcf
window=20000
step=2000

vcftools --vcf $vcf \
    --fst-window-size $window \
    --fst-window-step $step \
    --weir-fst-pop ../../data/pop.SC.table \
    --weir-fst-pop ../../data/pop.YZR.table \
    --out ./Fst.pop1.pop2
```

### 4. ROD + Fst è”åˆåˆ†æ

```bash
fst=../03.Fst/Fst.pop1.pop2.windowed.weir.fst
ROD=../01.pi_ROD/Pi.ROD
gff=../../data/genome.gff

# æå–æ˜¾è‘—ä½ç‚¹
perl ../../script/top_Fst_ROD_S.pl $fst $ROD 0.05 0.05 Fst_ROD.table Fst_ROD.stat

# ç­›é€‰topä½ç‚¹
awk '$NF=="top"' Fst_ROD.table > Fst_ROD.table.top

# æå–åŸºå› æ³¨é‡Š
awk '$3=="gene"' $gff > gene.gff

# æ³¨é‡Šå€™é€‰åŸºå› 
bedtools intersect -wo -F 0.1 -a Fst_ROD.table.top -b gene.gff | \
    awk -F "\t" '{print $7"\t"$10"\t"$11"\t"$13"\t"$15}' | \
    sort -u > Fst_ROD.table.top.gene
```



## GWAS åˆ†æ

### 1. åŸºå› å‹å¡«å……

```bash
beagle -Xmx4g -Djava.io.tmpdir=./TMP \
    gt=../data/all.vcf \
    out=all.impute \
    impute=true \
    window=10 \
    nthreads=2
```

### 2. æ ‡å‡† GWAS åˆ†æ

```bash
# Step 1: å‡†å¤‡VCFæ•°æ®
plink --vcf ../data/all.vcf \
    --maf 0.05 \
    --geno 0.1 \
    --recode12 \
    --output-missing-genotype 0 \
    --transpose \
    --out snp_filter \
    --set-missing-var-ids @:# \
    --allow-extra-chr

# å‡†å¤‡è¡¨å‹æ•°æ®
perl ../script/sort_pheno.pl snp_filter.tfam ../data/trait.table > trait.sort.txt

# Step 2: è®¡ç®—äº²ç¼˜å…³ç³»çŸ©é˜µ
/pub/software/emmax/emmax-kin-intel64 -v -d 10 -o ./pop.kinship snp_filter

# Step 3: è¿è¡ŒGWAS
/pub/software/emmax/emmax-intel64 -v -d 10 \
    -t snp_filter \
    -p trait.sort.txt \
    -k pop.kinship \
    -o emmax.out 1> emmax.log 2>emmax.err

# Step 4: ç»˜åˆ¶æ›¼å“ˆé¡¿å›¾
paste snp_filter.map emmax.out.ps | \
    awk 'BEGIN{print "SNP\tCHR\tBP\tP"}{if($2==$5){print $2"\t"$1"\t"$4"\t"$NF}}' \
    > emmax.out.ps.manht_input

Rscript ../script/manhattan.R emmax.out.ps.manht_input emmax.out.ps.manht_figure
```

### 3. GWAS-Q åˆ†æï¼ˆåŠ å…¥ç¾¤ä½“ç»“æ„åå˜é‡ï¼‰

```bash
# é“¾æ¥æ–‡ä»¶
ln -s ../01.GWAS/snp_filter.tped
ln -s ../01.GWAS/snp_filter.tfam
ln -s ../01.GWAS/snp_filter.nosex
ln -s ../01.GWAS/snp_filter.map
ln -s ../01.GWAS/trait.sort.txt
ln -s ../../02.population_genetics/03.structure/all.3.Q

# å‡†å¤‡QçŸ©é˜µ
paste snp_filter.nosex all.3.Q | awk '{print $1" "$1" 1 "$3" "$4}' > pop.Qmatrix

# è¿è¡ŒGWAS
/pub/software/emmax/emmax-intel64 -v -d 10 \
    -t snp_filter \
    -p trait.sort.txt \
    -k pop.kinship \
    -c pop.Qmatrix \
    -o emmax.out 1> emmax.log 2>emmax.err

# ç»˜åˆ¶æ›¼å“ˆé¡¿å›¾
paste snp_filter.map emmax.out.ps | \
    awk 'BEGIN{print "SNP\tCHR\tBP\tP"}{if($2==$5){print $2"\t"$1"\t"$4"\t"$NF}}' \
    > emmax.out.ps.manht_input

Rscript ../script/manhattan.R emmax.out.ps.manht_input emmax.out.ps.manht_figure_Q
```

### 4. çœŸå®æ•°æ® GWAS

```bash
# Step 1: å‡†å¤‡æ•°æ®
plink --vcf ../data_real/Sample215.M5M8H3.2allel.vcf \
    --maf 0.05 \
    --geno 0.1 \
    --recode12 \
    --output-missing-genotype 0 \
    --transpose \
    --out snp_filter \
    --set-missing-var-ids @:# \
    --allow-extra-chr

perl ../script/sort_pheno.pl snp_filter.tfam ../data_real/trait.C16_0.table > trait.sort.txt

# Step 2: è®¡ç®—äº²ç¼˜å…³ç³»çŸ©é˜µ
/pub/software/emmax/emmax-kin-intel64 -v -d 10 -o ./pop.kinship snp_filter

# Step 3: è¿è¡ŒGWAS
/pub/software/emmax/emmax-intel64 -v -d 10 \
    -t snp_filter \
    -p trait.sort.txt \
    -k pop.kinship \
    -o emmax.out 1> emmax.log 2>emmax.err

# Step 4: ç»˜åˆ¶æ›¼å“ˆé¡¿å›¾
paste snp_filter.map emmax.out.ps | \
    awk 'BEGIN{print "SNP\tCHR\tBP\tP"}{if($2==$5){print $2"\t"$1"\t"$4"\t"$NF}}' \
    > emmax.out.ps.manht_input

Rscript ../script/manhattan.R emmax.out.ps.manht_input emmax.out.ps.manht_figure_readData
```

## QTL-seq åˆ†æ

### 1. VCF è½¬æ¢ä¸º Table æ ¼å¼

```bash
ref=<reference.fasta>
vcf=<input.vcf>
outfile=<output.table>

gatk VariantsToTable \
    -R $ref \
    -V $vcf \
    -F CHROM -F POS -F REF -F ALT \
    -GF AD -GF DP -GF GQ -GF PL \
    -O $outfile
```



### 2. QTL-seq åˆ†æï¼ˆR è„šæœ¬ï¼‰

```R
library(QTLseqr)
library(ggplot2)

# è®¾ç½®å‚æ•°
HighBulk <- "SRR834931"  # é«˜å€¼æ··æ± åç§°
LowBulk <- "SRR834927"   # ä½å€¼æ··æ± åç§°
Chroms <- paste0(rep("Chr", 12), 1:12)  # æŸ“è‰²ä½“åˆ—è¡¨

# ==================== æ•°æ®è¯»å– ====================
df <- importFromGATK(
    file = "../data/Yang_et_al_2013.table",
    highBulk = HighBulk,
    lowBulk = LowBulk,
    chromList = Chroms
)

# ==================== è´¨é‡æ£€æŸ¥å›¾ ====================
# æ·±åº¦åˆ†å¸ƒå›¾
p1 <- ggplot(data = df) + 
    geom_histogram(aes(x = DP.HIGH + DP.LOW)) + 
    xlim(0, 800)
pdf(file = "SNP_depth.pdf")
p1
dev.off()

# REFç­‰ä½åŸºå› é¢‘ç‡åˆ†å¸ƒ
p2 <- ggplot(data = df) + 
    geom_histogram(aes(x = REF_FRQ))
pdf(file = "ref_allele_frequency.pdf")
p2
dev.off()

# é«˜å€¼æ··æ± SNP-indexåˆ†å¸ƒ
p3 <- ggplot(data = df) + 
    geom_histogram(aes(x = SNPindex.HIGH))
pdf(file = "SNPindex.HIGH.dis.pdf")
p3
dev.off()

# ä½å€¼æ··æ± SNP-indexåˆ†å¸ƒ
p4 <- ggplot(data = df) + 
    geom_histogram(aes(x = SNPindex.LOW))
pdf(file = "SNPindex.LOW.dis.pdf")
p4
dev.off()

# ==================== SNPè¿‡æ»¤ ====================
df_filt <- filterSNPs(
    SNPset = df,
    refAlleleFreq = 0.20,    # REFç­‰ä½åŸºå› é¢‘ç‡: 0.2~0.8
    minTotalDepth = 100,     # æœ€å°æ€»æ·±åº¦
    maxTotalDepth = 400,     # æœ€å¤§æ€»æ·±åº¦
    minSampleDepth = 40,     # å•æ ·å“æœ€å°æ·±åº¦
    minGQ = 99,              # åŸºå› å‹è´¨é‡é˜ˆå€¼
    verbose = TRUE
)

# ==================== deltaSNPindexè®¡ç®— ====================
df_filt <- runQTLseqAnalysis(
    df_filt,
    windowSize = 1e6,        # çª—å£å¤§å°
    popStruc = "F2",         # ç¾¤ä½“ç±»å‹
    bulkSize = c(385, 430),  # æ··æ± ä¸ªä½“æ•°
    replications = 10000,    # Bootstrapæ¬¡æ•°
    intervals = c(95, 99)    # ç½®ä¿¡åŒºé—´
)

# ==================== ç»“æœå¯è§†åŒ– ====================
# SNPæ²¿æŸ“è‰²ä½“åˆ†å¸ƒ
p5 <- plotQTLStats(SNPset = df_filt, var = "nSNPs")
pdf(file = "SNP_filter.window.pdf", width = 20, height = 4)
p5
dev.off()

# deltaSNPindexæ²¿æŸ“è‰²ä½“åˆ†å¸ƒ
p6 <- plotQTLStats(
    SNPset = df_filt, 
    var = "deltaSNP", 
    plotIntervals = TRUE
)
pdf(file = "deltaSNPindex.pdf", width = 20, height = 4)
p6
dev.off()

# ==================== æå–æ˜¾è‘—åŒºåŸŸ ====================
QTL <- getSigRegions(
    SNPset = df_filt, 
    method = "QTLseq", 
    interval = 95
)
write.table(QTL[[1]], "QTLseq_result.SigRegions.table", sep = "\t", quote = FALSE)

# ==================== å¯¼å‡ºå®Œæ•´ç»“æœ ====================
results99 <- getQTLTable(
    SNPset = df_filt,
    method = "QTLseq",
    interval = 99,
    export = FALSE
)
write.table(
    results99, 
    file = "QTLseq_result_deltaSNPindex_CI99.table", 
    sep = "\t", 
    quote = FALSE
)
```





## é™„å½•ï¼šåˆ†ææµç¨‹å›¾

plaintext

```R
åŸå§‹æ•°æ® (FASTQ)
      â”‚
      â–¼
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚   è´¨é‡æ§åˆ¶ QC   â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
      â”‚
      â–¼
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚  æ¯”å¯¹ Mapping   â”‚ â”€â”€â–º BWA MEM
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
      â”‚
      â–¼
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚ å»é‡ MarkDup    â”‚ â”€â”€â–º Picard
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
      â”‚
      â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
      â–¼                      â–¼                      â–¼
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”          â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”          â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚ SNP/InDel â”‚          â”‚    SV     â”‚          â”‚    CNV    â”‚
â”‚   GATK    â”‚          â”‚  lumpy-sv â”‚          â”‚  CNVnator â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜          â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜          â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
      â”‚                      â”‚                      â”‚
      â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
                             â”‚
                             â–¼
                    â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
                    â”‚      æ³¨é‡Š       â”‚
                    â”‚ ANNOVAR/bedtoolsâ”‚
                    â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
                             â”‚
         â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
         â–¼                   â–¼                   â–¼
   â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”       â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”       â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
   â”‚ ç¾¤ä½“ç»“æ„  â”‚       â”‚ é€‰æ‹©åˆ†æ  â”‚       â”‚   GWAS    â”‚
   â”‚ PCA/Tree  â”‚       â”‚ Fst/Ï€/D   â”‚       â”‚   EMMAX   â”‚
   â”‚ Structure â”‚       â”‚           â”‚       â”‚           â”‚
   â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜       â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜       â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
```

